// ornamentum/esm5/data-table-resource/services/data-table-websocket-data-fetch.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
import { webSocket } from 'rxjs/webSocket';
/**
 * Data table web socket data fetch service.
 * @template T
 */
var /**
 * Data table web socket data fetch service.
 * @template T
 */
DataTableWebsocketDataFetchService = /** @class */ (function () {
    function DataTableWebsocketDataFetchService() {
    }
    /**
     * Initialize web socket connection.
     * @param config Socket configuration parameters.
     */
    /**
     * Initialize web socket connection.
     * @param {?=} config Socket configuration parameters.
     * @return {?}
     */
    DataTableWebsocketDataFetchService.prototype.init = /**
     * Initialize web socket connection.
     * @param {?=} config Socket configuration parameters.
     * @return {?}
     */
    function (config) {
        this.socket = webSocket(config);
        this.subject = new Subject();
    };
    Object.defineProperty(DataTableWebsocketDataFetchService.prototype, "socketStream", {
        /**
         * Get socket stream reference.
         */
        get: /**
         * Get socket stream reference.
         * @return {?}
         */
        function () {
            return this.socket;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Web socket data bind event handler.
     * Must call init prior to calling this function.
     * @param mapper Response data mapper callback. map source stream format to data table expected stream or apply additional formatting.
     * @return Data table bind event handler.
     */
    /**
     * Web socket data bind event handler.
     * Must call init prior to calling this function.
     * @param {?=} mapper Response data mapper callback. map source stream format to data table expected stream or apply additional formatting.
     * @return {?} Data table bind event handler.
     */
    DataTableWebsocketDataFetchService.prototype.onDataBind = /**
     * Web socket data bind event handler.
     * Must call init prior to calling this function.
     * @param {?=} mapper Response data mapper callback. map source stream format to data table expected stream or apply additional formatting.
     * @return {?} Data table bind event handler.
     */
    function (mapper) {
        var _this = this;
        if (!this.socket) {
            throw Error('Initialize socket data source before data bind.');
        }
        this.socketSubscription = this.socket.subscribe(this.subject);
        return (/**
         * @param {?=} params
         * @return {?}
         */
        function (params) {
            if (params) {
                _this.socket.next((/** @type {?} */ ({
                    type: 'data-fetch',
                    offset: params.offset,
                    limit: params.limit,
                    fields: params.fields
                })));
                if (mapper) {
                    return mapper(_this.subject);
                }
                return _this.subject;
            }
        });
    };
    /**
     * Dispose web socket connection.
     */
    /**
     * Dispose web socket connection.
     * @return {?}
     */
    DataTableWebsocketDataFetchService.prototype.dispose = /**
     * Dispose web socket connection.
     * @return {?}
     */
    function () {
        if (this.socket) {
            this.socket.complete();
        }
        if (this.socketSubscription) {
            this.socketSubscription.unsubscribe();
        }
        if (this.subject) {
            this.subject.unsubscribe();
        }
    };
    return DataTableWebsocketDataFetchService;
}());
/**
 * Data table web socket data fetch service.
 * @template T
 */
export { DataTableWebsocketDataFetchService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTableWebsocketDataFetchService.prototype.socket;
    /**
     * @type {?}
     * @private
     */
    DataTableWebsocketDataFetchService.prototype.subject;
    /**
     * @type {?}
     * @private
     */
    DataTableWebsocketDataFetchService.prototype.socketSubscription;
}
