// ornamentum/esm5/data-table/components/data-table-no-data-body/data-table-no-data-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef } from '@angular/core';
import { DataFetchMode } from '../../models/data-fetch-mode.enum';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableEventStateService } from '../../services/data-table-event.service';
/**
 * No data body component. Used to display no data message when data not available.
 */
var DataTableNoDataBodyComponent = /** @class */ (function () {
    function DataTableNoDataBodyComponent(config, eventStateService) {
        this.config = config;
        this.eventStateService = eventStateService;
    }
    /**
     * Reset all applied filters.
     */
    /**
     * Reset all applied filters.
     * @return {?}
     */
    DataTableNoDataBodyComponent.prototype.resetFilters = /**
     * Reset all applied filters.
     * @return {?}
     */
    function () {
        this.eventStateService.dataFetchStream.emit(DataFetchMode.HARD_RELOAD);
    };
    Object.defineProperty(DataTableNoDataBodyComponent.prototype, "showDefaultNoDataTemplate", {
        /**
         * Get default no data template display status.
         * @return True if default no data template should be displayed.
         */
        get: /**
         * Get default no data template display status.
         * @return {?} True if default no data template should be displayed.
         */
        function () {
            return !!(!this.noRecordsTemplate && this.config.translations.noDataMessage);
        },
        enumerable: true,
        configurable: true
    });
    DataTableNoDataBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-data-table-no-data-body',
                    template: "<div class=\"ng-data-table-no-records-message\" *ngIf=\"showDefaultNoDataTemplate\">\n  <h1 class=\"ng-data-table-no-records-main-heading-message\">{{ config.translations.noDataMessage.header }}</h1>\n  <h2 class=\"ng-data-table-no-records-sub-heading-message\">{{ config.translations.noDataMessage.body }}</h2>\n  <button (click)=\"resetFilters()\" class=\"ng-data-table-no-records-message-button\">Reload Data</button>\n</div>\n<ng-template *ngIf=\"noRecordsTemplate\" [ngTemplateOutlet]=\"noRecordsTemplate\"></ng-template>\n"
                }] }
    ];
    /** @nocollapse */
    DataTableNoDataBodyComponent.ctorParameters = function () { return [
        { type: DataTableConfigService },
        { type: DataTableEventStateService }
    ]; };
    DataTableNoDataBodyComponent.propDecorators = {
        noRecordsTemplate: [{ type: Input }]
    };
    return DataTableNoDataBodyComponent;
}());
export { DataTableNoDataBodyComponent };
if (false) {
    /** @type {?} */
    DataTableNoDataBodyComponent.prototype.noRecordsTemplate;
    /** @type {?} */
    DataTableNoDataBodyComponent.prototype.config;
    /** @type {?} */
    DataTableNoDataBodyComponent.prototype.eventStateService;
}
