// ornamentum/esm5/data-table-resource/data-table-resource.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ResourceUtilityModule } from '../resource-utility/resource-utility.module';
import { DataTableResourceConfigService } from './services/data-table-resource-config.service';
import { DataTableHttpResourceFactoryService } from './services/data-table-http-resource-factory.service';
import { DataTableWebsocketResourceFactoryService } from './services/data-table-websocket-resource-factory.service';
/** @type {?} */
var PROVIDERS = [DataTableHttpResourceFactoryService, DataTableWebsocketResourceFactoryService];
/**
 * Data table data source module.
 */
var DataTableResourceModule = /** @class */ (function () {
    function DataTableResourceModule() {
    }
    /**
     * Set module root configuration overrides
     * @return Module with custom providers
     */
    /**
     * Set module root configuration overrides
     * @return {?} Module with custom providers
     */
    DataTableResourceModule.forRoot = /**
     * Set module root configuration overrides
     * @return {?} Module with custom providers
     */
    function () {
        return {
            ngModule: DataTableResourceModule,
            providers: [DataTableResourceConfigService]
        };
    };
    DataTableResourceModule.decorators = [
        { type: NgModule, args: [{
                    imports: [ResourceUtilityModule],
                    providers: tslib_1.__spread(PROVIDERS)
                },] }
    ];
    return DataTableResourceModule;
}());
export { DataTableResourceModule };
export { DataTableHttpDataFetchService } from './services/data-table-http-data-fetch.service';
export { DataTableHttpResourceFactoryService };
export { DataTableWebsocketDataFetchService } from './services/data-table-websocket-data-fetch.service';
export { DataTableWebsocketResourceFactoryService };
