// ornamentum/esm5/utility/utility.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { InitDirective } from './directives/init.directive';
import { FocusDirective } from './directives/focus.directive';
import { PixelConverterPipe } from './pipes/pixel-converter.pipe';
import { UtilityConfigService } from './services/utility-config.service';
import { DragAndDropService } from './services/drag-and-drop.service';
import { GlobalRefService } from './services/global-ref.service';
import { PopoverComponentLoaderFactoryService } from './services/popover-component-loader-factory.service';
import { ResizeService } from './services/resize.service';
import { ValidatorService } from './services/validator.service';
/** @type {?} */
var DIRECTIVES = [InitDirective, FocusDirective];
/** @type {?} */
var PIPES = [PixelConverterPipe];
/** @type {?} */
var PROVIDERS = [DragAndDropService, GlobalRefService, PopoverComponentLoaderFactoryService, ResizeService, ValidatorService];
/**
 * Module representing utility providers, directives and pipes.
 */
var UtilityModule = /** @class */ (function () {
    function UtilityModule() {
    }
    /**
     * Set module root configuration overrides.
     * @return Module with custom providers.
     */
    /**
     * Set module root configuration overrides.
     * @return {?} Module with custom providers.
     */
    UtilityModule.forRoot = /**
     * Set module root configuration overrides.
     * @return {?} Module with custom providers.
     */
    function () {
        return {
            ngModule: UtilityModule,
            providers: [UtilityConfigService]
        };
    };
    UtilityModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: tslib_1.__spread(DIRECTIVES, PIPES),
                    providers: tslib_1.__spread(PROVIDERS),
                    exports: tslib_1.__spread(DIRECTIVES, PIPES)
                },] }
    ];
    return UtilityModule;
}());
export { UtilityModule };
export { DragAndDropService } from './services/drag-and-drop.service';
export { GlobalRefService } from './services/global-ref.service';
export { UtilityConfigService } from './services/utility-config.service';
export { PopoverComponentLoaderFactoryService } from './services/popover-component-loader-factory.service';
export { PopoverComponentLoader } from './services/popover-component-loader.class';
