// ornamentum/esm2015/data-table/components/data-table/data-table.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, ContentChildren, EventEmitter, Input, Output, QueryList, TemplateRef, forwardRef, ElementRef, ViewChild, NgZone } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { Observable, of } from 'rxjs';
import { catchError, debounceTime, switchMap } from 'rxjs/operators';
import { get } from '../../../utility/services/object-utility.class';
import { DataFetchMode } from '../../models/data-fetch-mode.enum';
import { DataTableColumnComponent } from '../data-table-column/data-table-column.component';
import { DragAndDropService, GlobalRefService } from '../../../utility/utility.module';
import { ValidatorService } from '../../../utility/services/validator.service';
import { DataTableEventStateService } from '../../services/data-table-event.service';
import { DataTableDataStateService } from '../../services/data-table-data-state.service';
import { DataTablePersistenceService } from '../../services/data-table-persistence.service';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableScrollPositionService } from '../../services/data-table-scroll-position.service';
import { DataTableResourceService } from '../../services/data-table-resource.service';
/**
 * Data table component; Data table entry component.
 */
export class DataTableComponent {
    /**
     * @param {?} dragAndDropService
     * @param {?} dataTableStateService
     * @param {?} globalRefService
     * @param {?} eventStateService
     * @param {?} dataTableResourceService
     * @param {?} zone
     * @param {?} dataStateService
     * @param {?} scrollPositionService
     * @param {?} config
     */
    constructor(dragAndDropService, dataTableStateService, globalRefService, eventStateService, dataTableResourceService, zone, dataStateService, scrollPositionService, config) {
        this.dragAndDropService = dragAndDropService;
        this.dataTableStateService = dataTableStateService;
        this.globalRefService = globalRefService;
        this.eventStateService = eventStateService;
        this.dataTableResourceService = dataTableResourceService;
        this.zone = zone;
        this.dataStateService = dataStateService;
        this.scrollPositionService = scrollPositionService;
        this.config = config;
        this.storageMode = config.storageMode;
        this.headerClick = this.eventStateService.headerClickStream;
        this.allRowSelectChange = this.eventStateService.allRowSelectChangeStream;
        this.rowBind = this.eventStateService.rowBindStream;
        this.rowClick = this.eventStateService.rowClickStream;
        this.rowDoubleClick = this.eventStateService.rowDoubleClickStream;
        this.rowSelectChange = this.eventStateService.rowSelectChangeStream;
        this.cellBind = this.eventStateService.cellBindStream;
        this.cellClick = this.eventStateService.cellClickStream;
        this.init = this.eventStateService.initStream;
        this.dataBound = this.eventStateService.dataBoundStream;
        this.columnBind = this.eventStateService.columnBind;
    }
    // Input Events
    /**
     * Set data bind event handler callback. This handler is fired on each data fetch request.
     * @param {?} value
     * @return {?}
     */
    set onDataBind(value) {
        this.dataStateService.onDataBind = value;
    }
    /**
     * Set filter value extract event handler callback. Used to extract filter value collection dynamically when
     * explicit data bind functionality is used with onDataBind callback.
     * @param {?} value
     * @return {?}
     */
    set onFilterValueExtract(value) {
        this.dataStateService.onFilterValueExtract = value;
    }
    /**
     * Set on dynamic row span extract event handler callback.
     * @param {?} value
     * @return {?}
     */
    set onDynamicRowSpanExtract(value) {
        this.dataStateService.onDynamicRowSpanExtract = value;
    }
    // Input parameters
    /**
     * Set static data items collection. No need to set data source when static items collection is provided.
     * @param {?} value
     * @return {?}
     */
    set items(value) {
        if (!value) {
            return;
        }
        this.eventStateService.staticDataSourceStream.next(value);
    }
    /**
     * Set data source observable. This observable is used to retrieve row data for binding.
     * @param {?} source
     * @return {?}
     */
    set dataSource(source) {
        this.initDataSource(source);
    }
    /**
     * Set data table unique identifier.
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        if (!ValidatorService.idPatternValidatorExpression.test(value)) {
            throw Error('Invalid [id] input value. Unique identifier parameter only accept string begin with a letter ([A-Za-z]) and may be followed by any number of letters, digits ([0-9]), hyphens ("-"), underscores ("_").');
        }
        this.dataStateService.id = value;
    }
    /**
     * Set persist table state on provided storage mode if true. Depends on storageMode property.
     * @param {?} value
     * @return {?}
     */
    set persistTableState(value) {
        this.config.persistTableState = value;
    }
    /**
     * Set storage mode to persist table state. Only applicable when persistTableState is true.
     * @param {?} value
     * @return {?}
     */
    set storageMode(value) {
        this.dataTableStateService.storageMode = value;
    }
    /**
     * Set multiple column sortable if true. Only applicable for sortable true columns.
     * @param {?} value
     * @return {?}
     */
    set multiColumnSortable(value) {
        this.config.multiColumnSortable = value;
    }
    /**
     * Set table header bar visible if true.
     * @param {?} value
     * @return {?}
     */
    set showHeader(value) {
        this.config.showHeader = value;
    }
    /**
     * Set title to be shown in the header. Only applicable when showHeader is true.
     * @param {?} value
     * @return {?}
     */
    set title(value) {
        this.config.title = value;
    }
    /**
     * Set width value in pixels. Can be used to set the width of teh table (responsive if not set).
     * @param {?} value
     * @return {?}
     */
    set width(value) {
        this.config.width = value;
    }
    /**
     * Set minimum table content height value in pixels. Can be used to set the minimum height of the table content area.
     * @param {?} value
     * @return {?}
     */
    set minContentHeight(value) {
        this.config.minContentHeight = value;
    }
    /**
     * Minimum table content width value in pixels. Can be used to set the minimum width of the table content area.
     * @param {?} value
     * @return {?}
     */
    set minContentWidth(value) {
        this.config.minContentWidth = value;
    }
    /**
     * Table content height value in pixels. This configuration can be used to enable table content vertical
     * scrolling for responsive design.
     * @param {?} value
     * @return {?}
     */
    set contentHeight(value) {
        this.config.contentHeight = value;
    }
    /**
     * Show pagination bar if true. Depends on offset and limit values. Trigger dataLoad event with offset
     * and limit values.
     * @param {?} value
     * @return {?}
     */
    set pageable(value) {
        this.config.pageable = value;
    }
    /**
     * Enable scrolling based on-demand data loading functionality if true. Trigger dataLoad event with offset
     * and limit values when scroll to bottom until data source exhaust.
     * @param {?} value
     * @return {?}
     */
    set loadOnScroll(value) {
        this.config.loadOnScroll = value;
    }
    /**
     * Set view height distance ratio to trigger data fetch on scroll. Applicable only when load on scroll mode is enabled.
     * @param {?} value
     * @return {?}
     */
    set loadViewDistanceRatio(value) {
        this.config.loadViewDistanceRatio = value;
    }
    /**
     * Set auto generated index column with row numbering if true.
     * @param {?} value
     * @return {?}
     */
    set showIndexColumn(value) {
        this.config.showIndexColumn = value;
    }
    /**
     * Set index column header title. Applicable when showIndexColumn is true.
     * @param {?} value
     * @return {?}
     */
    set indexColumnTitle(value) {
        this.config.indexColumnTitle = value;
    }
    /**
     * Set row select checkbox and select state if true.
     * @param {?} value
     * @return {?}
     */
    set rowSelectable(value) {
        this.config.rowSelectable = value;
    }
    /**
     * Data table row select mode. Applicable only when rowSelectable is true.
     * @param {?} value
     * @return {?}
     */
    set selectMode(value) {
        this.config.selectMode = value;
    }
    /**
     * Set select all row checkbox on column header visible if true.
     * Only applicable when showRowSelectCheckbox, rowSelectable is true & item selectMode is  multi.
     * @param {?} value
     * @return {?}
     */
    set showRowSelectCheckbox(value) {
        this.config.showRowSelectCheckbox = value;
    }
    /**
     * Set select all row checkbox on column header visible if true.
     * Only applicable when showRowSelectCheckbox, rowSelectable is true & item selectMode is  multi.
     * @param {?} value
     * @return {?}
     */
    set showRowSelectAllCheckbox(value) {
        this.config.showRowSelectAllCheckbox = value;
    }
    /**
     * Set substitute rows visible if true. Fill with empty rows when row count < limit.
     * @param {?} value
     * @return {?}
     */
    set showSubstituteRows(value) {
        this.config.showSubstituteRows = value;
    }
    /**
     * Set row expander visible if true. Render ngDataTableExpand template on expand click.
     * @param {?} value
     * @return {?}
     */
    set expandableRows(value) {
        this.config.expandableRows = value;
    }
    /**
     * Set trigger row select on click event if true. Applicable only when rowSelectable is true.
     * @param {?} value
     * @return {?}
     */
    set selectOnRowClick(value) {
        this.config.selectOnRowClick = value;
    }
    /**
     * Set expand and render expand template on row click if true. Only applicable when expandableRows is true.
     * @param {?} value
     * @return {?}
     */
    set expandOnRowClick(value) {
        this.config.expandOnRowClick = value;
    }
    /**
     * Auto trigger dataLoad event on initialization if true.
     * @param {?} value
     * @return {?}
     */
    set autoFetch(value) {
        this.config.autoFetch = value;
    }
    /**
     * Set loading spinner visible if true. Show loading spinner when data fetch operation is triggered.
     * @param {?} value
     * @return {?}
     */
    set showLoadingSpinner(value) {
        this.config.showLoadingSpinner = value;
    }
    /**
     * Set select option track by field path which is used to uniquely identify row for selection tracking.
     * This field support object paths expressions 'root[0].nest'.
     * @param {?} value
     * @return {?}
     */
    set selectTrackBy(value) {
        this.config.selectTrackBy = value;
    }
    /**
     * Set selected row identifier. Select specified row on initial load.
     * Applicable when row select mode is SINGLE or SINGLE_TOGGLE.
     * @param {?} value
     * @return {?}
     */
    set selectedRow(value) {
        this.dataStateService.selectedRow = value;
        this.eventStateService.rowSelectChangeStream.emit(this.dataStateService.selectedRow);
    }
    /**
     * Set selected row identifiers collection. Select specified rows on initial load.
     * Applicable when selectMode is SINGLE or SINGLE_TOGGLE true.
     * @param {?} value
     * @return {?}
     */
    set selectedRows(value) {
        this.dataStateService.selectedRows = value || [];
        this.eventStateService.rowSelectChangeStream.emit(this.dataStateService.selectedRows);
    }
    /**
     * Set filter debounce time in milliseconds. Applicable only when filterDebounce is true.
     * @param {?} value
     * @return {?}
     */
    set filterDebounceTime(value) {
        this.config.filterDebounceTime = value;
    }
    /**
     * Set filter data debounce enabled state with provided filterDebounceTime if true.
     * @param {?} value
     * @return {?}
     */
    set filterDebounce(value) {
        this.config.filterDebounce = value;
    }
    /**
     * Set refresh button visible if true. Only applicable when showHeader is true.
     * @param {?} value
     * @return {?}
     */
    set showRefreshButton(value) {
        this.config.showRefreshButton = value;
    }
    /**
     * Row selector column width in pixels. Applicable only when showColumnSelector is true.
     * @param {?} value
     * @return {?}
     */
    set showColumnSelector(value) {
        this.config.showColumnSelector = value;
    }
    /**
     * Set column selector dropdown width in pixels. Only applicable when showColumnSelector is true.
     * @param {?} value
     * @return {?}
     */
    set columnSelectorWidth(value) {
        this.config.columnSelectorWidth = value;
    }
    /**
     * Set expander column width in pixels. Applicable only when expandableRows is true.
     * @param {?} value
     * @return {?}
     */
    set expanderColumnWidth(value) {
        this.config.expanderColumnWidth = value;
    }
    /**
     * Set index column width in pixels. Applicable only when showIndexColumn is true.
     * @param {?} value
     * @return {?}
     */
    set indexColumnWidth(value) {
        this.config.indexColumnWidth = value;
    }
    /**
     * Set row selector column width in pixels. Applicable only when showColumnSelector is true.
     * @param {?} value
     * @return {?}
     */
    set selectionColumnWidth(value) {
        this.config.selectionColumnWidth = value;
    }
    /**
     * Set translation data object. Used to localize table static label text.
     * @param {?} data
     * @return {?}
     */
    set translations(data) {
        this.config.translations = data;
    }
    /**
     * Set row expand loading spinner visible if true. Applicable only when row expand is enabled.
     * @param {?} value
     * @return {?}
     */
    set showRowExpandLoadingSpinner(value) {
        this.config.showRowExpandLoadingSpinner = value;
    }
    /**
     * Set data offset value (start offset index). Applicable only when pageable is true.
     * @param {?} value
     * @return {?}
     */
    set offset(value) {
        this.config.offset = value;
        this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
    }
    /**
     * Set data limit value (page size). Applicable only when pageable is true.
     * @param {?} value
     * @return {?}
     */
    set limit(value) {
        this.config.limit = value;
        this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
    }
    /**
     * Set current page number. Auto calculate offset depending on page number,
     * do not explicitly set offset when page is used.
     * @param {?} value
     * @return {?}
     */
    set page(value) {
        this.offset = (value - 1) * this.config.limit;
    }
    /**
     * Get current page number.
     * @return {?}
     */
    get page() {
        return Math.floor(this.config.offset / this.config.limit) + 1;
    }
    /**
     * Get data table header padding in pixels.
     * @return {?}
     */
    get headerPadding() {
        return this.config.contentHeight ? this.globalRefService.scrollbarWidth : 0;
    }
    /**
     * Get data loading status.
     * @return {?}
     */
    get isLoading() {
        return !(this.config.loadOnScroll && this.dataStateService.dataRows.length)
            && this.config.showLoadingSpinner && this.dataStateService.dataLoading;
    }
    /**
     * On after data bind event handler
     * @private
     * @param {?} queryResult Query result object
     * @return {?}
     */
    onAfterDataBind(queryResult) {
        this.dataStateService.itemCount = queryResult.count;
        this.setDataRows(queryResult.items);
        if (this.dataStateService.heardReload) {
            this.eventStateService.fetchFilterOptionsStream.next(false);
            this.dataStateService.heardReload = false;
        }
        this.dataStateService.dataLoading = false;
        this.eventStateService.dataBoundStream.emit();
    }
    /**
     * Get data item selected state
     * @private
     * @param {?} item Data item object
     * @return {?} True if item is selected
     */
    getSelectedState(item) {
        /** @type {?} */
        const id = get(item, this.config.selectTrackBy);
        if (id === undefined) {
            return false;
        }
        if (this.config.selectMode === 'multi') {
            return this.dataStateService.selectedRows.indexOf(id) > -1;
        }
        return this.dataStateService.selectedRow === id;
    }
    /**
     * Set data table item collection
     * @private
     * @param {?} items Data table item collection
     * @return {?}
     */
    setDataRows(items) {
        /** @type {?} */
        const mappedItems = items.map((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        (item, index) => {
            /** @type {?} */
            let currentIndex;
            if (this.config.loadOnScroll || this.config.pageable) {
                currentIndex = this.config.offset + index + 1;
            }
            else {
                currentIndex = index + 1;
            }
            return {
                dataLoaded: false,
                expanded: false,
                disabled: false,
                color: '',
                cssClass: '',
                tooltip: '',
                index: currentIndex,
                item,
                selected: this.getSelectedState(item)
            };
        }));
        if (this.config.loadOnScroll) {
            this.dataStateService.dataRows = [...this.dataStateService.dataRows, ...mappedItems];
        }
        else {
            this.dataStateService.dataRows = mappedItems;
        }
        if (this.config.selectMode === 'multi') {
            this.dataStateService.allRowSelected =
                this.dataStateService.dataRows.length !== 0 &&
                    this.dataStateService.dataRows.every((/**
                     * @param {?} dataRow
                     * @return {?}
                     */
                    (dataRow) => {
                        return dataRow.selected;
                    }));
        }
        if (!this.config.loadOnScroll) {
            /** @type {?} */
            const substituteRowCount = this.config.limit - this.dataStateService.dataRows.length;
            this.dataStateService.substituteRows = Array.from({
                length: substituteRowCount
            });
        }
    }
    /**
     * Initialize data fetch event
     * @private
     * @return {?}
     */
    initDataFetchEvent() {
        /** @type {?} */
        const noop = {
            items: [],
            count: 0
        };
        this.dataFetchStreamSubscription = this.eventStateService.dataFetchStream
            .pipe(debounceTime(20), switchMap((/**
         * @param {?} fetchMode
         * @return {?}
         */
        (fetchMode) => this.mapDataBind(fetchMode))), catchError((/**
         * @return {?}
         */
        () => {
            return of(noop);
        })))
            .subscribe((/**
         * @param {?} queryResult
         * @return {?}
         */
        (queryResult) => {
            this.onAfterDataBind(queryResult);
        }), (/**
         * @return {?}
         */
        () => {
            this.onAfterDataBind(noop);
        }));
    }
    /**
     * Re-map data binding data
     * @private
     * @param {?} fetchMode Data fetch mode
     * @return {?} Data table query result stream
     */
    mapDataBind(fetchMode) {
        this.dataStateService.dataLoading = true;
        if (fetchMode === DataFetchMode.HARD_RELOAD) {
            this.clearRowSelectState();
            this.clearColumnState();
            this.dataStateService.heardReload = true;
            this.config.offset = 0;
        }
        /** @type {?} */
        const params = {
            loadData: fetchMode === DataFetchMode.HARD_RELOAD || fetchMode === DataFetchMode.SOFT_RELOAD
        };
        if (this.columns) {
            params.fields = this.columns
                .filter((/**
             * @param {?} column
             * @return {?}
             */
            column => {
                return column.sortable || column.filterable;
            }))
                .reduce((/**
             * @param {?} acc
             * @param {?} column
             * @return {?}
             */
            (acc, column) => {
                if (column.sortField || column.filterField) {
                    acc.push({
                        field: column.sortField || column.filterField,
                        column
                    });
                }
                else {
                    acc.push({
                        field: column.field,
                        column
                    });
                }
                return acc;
            }), [])
                .map((/**
             * @param {?} uniqueField
             * @return {?}
             */
            (uniqueField) => {
                /** @type {?} */
                let filter;
                if (uniqueField.column.showDropdownFilter) {
                    if (uniqueField.column.dropdownFilterSelectMode === 'multi') {
                        filter = uniqueField.column.filter && uniqueField.column.filter.map((/**
                         * @param {?} filterValue
                         * @return {?}
                         */
                        filterValue => filterValue.key));
                    }
                    else {
                        filter = uniqueField.column.filter && uniqueField.column.filter.key;
                    }
                }
                else {
                    filter = uniqueField.column.filter;
                }
                return {
                    field: uniqueField.field,
                    sortable: uniqueField.column.sortable,
                    sortOrder: uniqueField.column.sortOrder,
                    sortPriority: uniqueField.column.sortPriority || (uniqueField.column.sortOrder ? 1 : 0),
                    filterable: uniqueField.column.filterable,
                    filterValue: filter,
                    filterExpression: uniqueField.column.filterExpression,
                };
            }));
        }
        if (this.config.pageable || this.config.loadOnScroll) {
            params.offset = this.config.offset;
            params.limit = this.config.limit;
        }
        if (this.config.persistTableState) {
            this.dataTableStateService.setState(this.dataStateService.id, params);
        }
        return this.dataStateService.onDataBind(params);
    }
    /**
     * Initialize data table state via previous state snapshot; Applicable only when persist table state is enabled
     * @private
     * @return {?}
     */
    initDataTableState() {
        if (this.config.persistTableState) {
            /** @type {?} */
            const dataTableState = this.dataTableStateService.getState(this.dataStateService.id);
            if (dataTableState) {
                this.columns.forEach((/**
                 * @param {?} column
                 * @return {?}
                 */
                column => {
                    /** @type {?} */
                    const field = dataTableState.fields.find((/**
                     * @param {?} col
                     * @return {?}
                     */
                    col => {
                        return col.field === column.field;
                    }));
                    if (field) {
                        if (column.filterable && field.filterable) {
                            if (column.showDropdownFilter) {
                                if (field.filterValue) {
                                    if (column.dropdownFilterSelectMode === 'multi') {
                                        column.filter = field.filterValue.map((/**
                                         * @param {?} filterValue
                                         * @return {?}
                                         */
                                        (filterValue) => {
                                            return {
                                                key: filterValue,
                                                value: filterValue
                                            };
                                        }));
                                    }
                                    else {
                                        column.filter = {
                                            key: field.filterValue,
                                            value: field.filterValue
                                        };
                                    }
                                }
                            }
                            else {
                                column.filter = field.filterValue;
                            }
                        }
                        if (column.sortable && field.sortable) {
                            column.sortOrder = field.sortOrder;
                        }
                    }
                }));
                this.config.limit = dataTableState.limit;
                this.config.offset = dataTableState.offset;
            }
        }
    }
    /**
     * After component initialize lifecycle event handler
     * @return {?}
     */
    ngAfterContentInit() {
        this.dataStateService.relativeParentElement = this.dataTableElement.nativeElement;
        if (!this.dataStateService.onDataBind) {
            this.dataSource = this.eventStateService.staticDataSourceStream;
        }
        this.initDataTableState();
        this.initDataFetchEvent();
        if (this.config.autoFetch) {
            this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
        }
        this.eventStateService.fetchFilterOptionsStream.next(true);
        this.eventStateService.initStream.emit(this);
        if (this.config.loadOnScroll) {
            this.scrollPositionSubscription = this.scrollPositionService.scrollPositionStream.subscribe((/**
             * @param {?} pos
             * @return {?}
             */
            (pos) => {
                /** @type {?} */
                const roundingPixel = 1;
                /** @type {?} */
                const gutterPixel = 1;
                if (pos.isVertical
                    && pos.scrollTop >= pos.scrollHeight - (1 + this.config.loadViewDistanceRatio) * pos.clientHeight - roundingPixel - gutterPixel
                    && (this.config.offset + this.config.limit) < this.dataStateService.itemCount
                    && !this.dataStateService.dataLoading) {
                    this.dataStateService.dataLoading = true;
                    this.zone.run((/**
                     * @return {?}
                     */
                    () => {
                        this.offset = this.config.offset + this.config.limit;
                    }));
                }
            }));
        }
    }
    /**
     * Reset column sort and filter state
     * @private
     * @return {?}
     */
    clearColumnState() {
        this.columns.forEach((/**
         * @param {?} column
         * @return {?}
         */
        (column) => {
            column.resetSortOrder();
            column.filter = undefined;
        }));
    }
    /**
     * Clear selected data row status
     * @private
     * @return {?}
     */
    clearRowSelectState() {
        this.dataStateService.selectedRow = undefined;
        this.dataStateService.selectedRows = [];
        this.dataStateService.allRowSelected = false;
    }
    /**
     * Fetch data from data source
     * @param {?=} fetchMode Data fetch mode
     * @return {?}
     */
    fetchData(fetchMode = DataFetchMode.SOFT_RELOAD) {
        this.eventStateService.dataFetchStream.next(fetchMode);
    }
    /**
     * Initialize data source
     * @param {?} source Data source stream
     * @return {?}
     */
    initDataSource(source) {
        this.dataTableResourceService.setDataSource(source);
        this.onDataBind = (/**
         * @param {?} params
         * @return {?}
         */
        (params) => {
            if (params.loadData) {
                this.dataTableResourceService.setDataSource(source);
            }
            return this.dataTableResourceService.query(params);
        });
        this.onFilterValueExtract = (/**
         * @param {?} column
         * @return {?}
         */
        (column) => {
            return this.dataTableResourceService.extractFilterOptions(column);
        });
    }
    /**
     * Component value write event handler; Form control support implementation
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        if (this.config.selectMode === 'multi') {
            this.selectedRows = value;
        }
        else {
            this.selectedRow = value;
        }
    }
    /**
     * Register select change event handler; Form control support implementation
     * @param {?} onSelectChange Select change event handler callback
     * @return {?}
     */
    registerOnChange(onSelectChange) {
        this.rowSelectChangeSubscription = this.eventStateService.rowSelectChangeStream.subscribe((/**
         * @param {?} selectedIds
         * @return {?}
         */
        (selectedIds) => {
            onSelectChange(selectedIds);
        }));
    }
    /**
     * Register on touch event handler; Form control support implementation
     * @param {?} fn Touch event callback handler
     * @return {?}
     */
    registerOnTouched(fn) { }
    /**
     * Get table width in pixels
     * @return {?}
     */
    get tableWidth() {
        return this.config.width || this.dataStateService.tableWidth;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.dataStateService.id) {
            throw Error('Missing required parameter value for [id] input.');
        }
        if (this.config.loadOnScroll) {
            if (!this.config.minContentHeight) {
                throw Error('[minContentHeight] is required when [infiniteScrollable] is enabled.');
            }
            if (this.config.pageable) {
                throw Error('[pageable] and [infiniteScrollable] cannot be enabled at the same time.');
            }
        }
    }
    /**
     * Component destroy lifecycle event handler
     * @return {?}
     */
    ngOnDestroy() {
        if (this.dataFetchStreamSubscription) {
            this.dataFetchStreamSubscription.unsubscribe();
        }
        if (this.rowSelectChangeSubscription) {
            this.rowSelectChangeSubscription.unsubscribe();
        }
        if (this.scrollPositionSubscription) {
            this.scrollPositionSubscription.unsubscribe();
        }
        this.dataTableResourceService.dispose();
    }
}
DataTableComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-data-table',
                template: "<div class=\"ng-data-table-container\" #dataTableElement [class.scrollable]=\"config.contentHeight\">\n  <ng-data-table-header *ngIf=\"config.showHeader\" [columns]=\"columns\"> </ng-data-table-header>\n  <div class=\"ng-data-table-content-container\" [ngScrollPosition]=\"scrollPositionService.scrollPositionStream\">\n    <div class=\"ng-data-table-content-header-container\" [style.paddingRight]=\"headerPadding | ngPx\">\n      <div class=\"ng-data-table-content-header\" [ngScrollElement]=\"scrollPositionService.scrollPositionStream\">\n        <table class=\"ng-data-table\" [style.width]=\"tableWidth | ngPx\" [style.minWidth]=\"config.minContentWidth | ngPx\">\n          <colgroup ngDataTableColGroup [columns]=\"columns\"></colgroup>\n          <thead ngDataTableHead [columns]=\"columns\"></thead>\n        </table>\n      </div>\n    </div>\n    <div class=\"ng-data-table-content\" [ngScrollPosition]=\"scrollPositionService.scrollPositionStream\"\n         [style.height]=\"config.contentHeight | ngPx\" [style.minHeight]=\"config.minContentHeight | ngPx\">\n      <ng-data-table-loading-spinner [loadingSpinnerTemplate]=\"loadingSpinnerTemplate\" [isLoading]=\"isLoading\" [showOverlay]=\"true\">\n      </ng-data-table-loading-spinner>\n      <table class=\"ng-data-table\" [style.width]=\"tableWidth | ngPx\" [style.minWidth]=\"config.minContentWidth | ngPx\" [hidden]=\"isLoading\">\n        <colgroup ngDataTableColGroup [columns]=\"columns\"></colgroup>\n        <tbody\n          ngDataTableBody\n          [columns]=\"columns\"\n          [rowExpandTemplate]=\"rowExpandTemplate\"\n          [rowExpandLoadingSpinnerTemplate]=\"rowExpandLoadingSpinnerTemplate\"\n        ></tbody>\n      </table>\n      <ng-data-table-no-data-body\n        class=\"ng-data-table-no-records-message-container\"\n        *ngIf=\"dataStateService.showNoDataOverlay\"\n        [noRecordsTemplate]=\"noRecordsTemplate\"\n      >\n      </ng-data-table-no-data-body>\n    </div>\n  </div>\n  <ng-data-table-pagination *ngIf=\"config.pageable && dataStateService.itemCount\"></ng-data-table-pagination>\n</div>\n",
                providers: [
                    DataTableConfigService,
                    DataTableEventStateService,
                    DataTablePersistenceService,
                    DataTableDataStateService,
                    DataTableScrollPositionService,
                    DataTableResourceService,
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => DataTableComponent)),
                        multi: true
                    }
                ]
            }] }
];
/** @nocollapse */
DataTableComponent.ctorParameters = () => [
    { type: DragAndDropService },
    { type: DataTablePersistenceService },
    { type: GlobalRefService },
    { type: DataTableEventStateService },
    { type: DataTableResourceService },
    { type: NgZone },
    { type: DataTableDataStateService },
    { type: DataTableScrollPositionService },
    { type: DataTableConfigService }
];
DataTableComponent.propDecorators = {
    columns: [{ type: ContentChildren, args: [DataTableColumnComponent,] }],
    rowExpandTemplate: [{ type: ContentChild, args: ['ngDataTableRowExpand', { static: true },] }],
    noRecordsTemplate: [{ type: ContentChild, args: ['ngDataTableNoRecords', { static: true },] }],
    loadingSpinnerTemplate: [{ type: ContentChild, args: ['ngDataTableLoadingSpinner', { static: true },] }],
    rowExpandLoadingSpinnerTemplate: [{ type: ContentChild, args: ['ngDataTableRowExpandLoadingSpinner', { static: true },] }],
    dataTableElement: [{ type: ViewChild, args: ['dataTableElement', { static: true },] }],
    init: [{ type: Output }],
    rowSelectChange: [{ type: Output }],
    rowClick: [{ type: Output }],
    rowDoubleClick: [{ type: Output }],
    headerClick: [{ type: Output }],
    allRowSelectChange: [{ type: Output }],
    cellClick: [{ type: Output }],
    dataBound: [{ type: Output }],
    rowBind: [{ type: Output }],
    columnBind: [{ type: Output }],
    cellBind: [{ type: Output }],
    onDataBind: [{ type: Input }],
    onFilterValueExtract: [{ type: Input }],
    onDynamicRowSpanExtract: [{ type: Input }],
    items: [{ type: Input }],
    dataSource: [{ type: Input }],
    id: [{ type: Input }],
    persistTableState: [{ type: Input }],
    storageMode: [{ type: Input }],
    multiColumnSortable: [{ type: Input }],
    showHeader: [{ type: Input }],
    title: [{ type: Input }],
    width: [{ type: Input }],
    minContentHeight: [{ type: Input }],
    minContentWidth: [{ type: Input }],
    contentHeight: [{ type: Input }],
    pageable: [{ type: Input }],
    loadOnScroll: [{ type: Input }],
    loadViewDistanceRatio: [{ type: Input }],
    showIndexColumn: [{ type: Input }],
    indexColumnTitle: [{ type: Input }],
    rowSelectable: [{ type: Input }],
    selectMode: [{ type: Input }],
    showRowSelectCheckbox: [{ type: Input }],
    showRowSelectAllCheckbox: [{ type: Input }],
    showSubstituteRows: [{ type: Input }],
    expandableRows: [{ type: Input }],
    selectOnRowClick: [{ type: Input }],
    expandOnRowClick: [{ type: Input }],
    autoFetch: [{ type: Input }],
    showLoadingSpinner: [{ type: Input }],
    selectTrackBy: [{ type: Input }],
    selectedRow: [{ type: Input }],
    selectedRows: [{ type: Input }],
    filterDebounceTime: [{ type: Input }],
    filterDebounce: [{ type: Input }],
    showRefreshButton: [{ type: Input }],
    showColumnSelector: [{ type: Input }],
    columnSelectorWidth: [{ type: Input }],
    expanderColumnWidth: [{ type: Input }],
    indexColumnWidth: [{ type: Input }],
    selectionColumnWidth: [{ type: Input }],
    translations: [{ type: Input }],
    showRowExpandLoadingSpinner: [{ type: Input }],
    offset: [{ type: Input }],
    limit: [{ type: Input }],
    page: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.rowSelectChangeSubscription;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.dataFetchStreamSubscription;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.scrollPositionSubscription;
    /**
     * Data table column component collection.
     * @type {?}
     */
    DataTableComponent.prototype.columns;
    /**
     * Template to display when data row is expanded for detail view.
     * @type {?}
     */
    DataTableComponent.prototype.rowExpandTemplate;
    /**
     * Template to display when data set is empty.
     * @type {?}
     */
    DataTableComponent.prototype.noRecordsTemplate;
    /**
     * Template to display while loading data.
     * @type {?}
     */
    DataTableComponent.prototype.loadingSpinnerTemplate;
    /**
     * Template to display while row is expanding to load detail view.
     * @type {?}
     */
    DataTableComponent.prototype.rowExpandLoadingSpinnerTemplate;
    /**
     * Data table self DOM element reference.
     * @type {?}
     */
    DataTableComponent.prototype.dataTableElement;
    /**
     * Data table initialize event handler.
     * Triggered after data table initialize.
     * @type {?}
     */
    DataTableComponent.prototype.init;
    /**
     * Row selected state change event handler.
     * Triggered when table row selected state change.
     * @type {?}
     */
    DataTableComponent.prototype.rowSelectChange;
    /**
     * Row click event handler.
     * Triggered when data row is clicked.
     * @type {?}
     */
    DataTableComponent.prototype.rowClick;
    /**
     * Row double click event handler.
     * Triggered when data row is double clicked.
     * @type {?}
     */
    DataTableComponent.prototype.rowDoubleClick;
    /**
     * Header click event handler.
     * Triggered when header column clicked.
     * @type {?}
     */
    DataTableComponent.prototype.headerClick;
    /**
     * All row select change event handler.
     * Triggered when all row select state changed.
     * @type {?}
     */
    DataTableComponent.prototype.allRowSelectChange;
    /**
     * Cell click event handler.
     * Triggered when clicked on a cell.
     * @type {?}
     */
    DataTableComponent.prototype.cellClick;
    /**
     * Data bound event handler.
     * Triggered after data bind.
     * @type {?}
     */
    DataTableComponent.prototype.dataBound;
    /**
     * Row bind event handler.
     * Trigger on each row data bind.
     * @type {?}
     */
    DataTableComponent.prototype.rowBind;
    /**
     * Column bind event handler.
     * Triggered after column data bind.
     * @type {?}
     */
    DataTableComponent.prototype.columnBind;
    /**
     * Cell bind event handler.
     * Triggered after data table cell data bind.
     * @type {?}
     */
    DataTableComponent.prototype.cellBind;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.dragAndDropService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.dataTableStateService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.globalRefService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.eventStateService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.dataTableResourceService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.zone;
    /** @type {?} */
    DataTableComponent.prototype.dataStateService;
    /** @type {?} */
    DataTableComponent.prototype.scrollPositionService;
    /** @type {?} */
    DataTableComponent.prototype.config;
}
