// ornamentum/esm5/data-table/components/data-table-column-title-header/data-table-column-title-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { get } from '../../../utility/services/object-utility.class';
import { DataFetchMode } from '../../models/data-fetch-mode.enum';
import { DragAndDropService } from '../../../utility/utility.module';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableEventStateService } from '../../services/data-table-event.service';
import { DataTableDataStateService } from '../../services/data-table-data-state.service';
/**
 * Column title header component. Render data table column headers.
 */
var DataTableColumnTitleHeaderComponent = /** @class */ (function () {
    function DataTableColumnTitleHeaderComponent(dragAndDropService, eventStateService, dataStateService, config) {
        this.dragAndDropService = dragAndDropService;
        this.eventStateService = eventStateService;
        this.dataStateService = dataStateService;
        this.config = config;
        this.resizeInProgress = false;
    }
    /**
     * Header click event handler.
     * @param column Data table column component reference.
     * @param event Mouse event arguments object.
     */
    /**
     * Header click event handler.
     * @param {?} column Data table column component reference.
     * @param {?} event Mouse event arguments object.
     * @return {?}
     */
    DataTableColumnTitleHeaderComponent.prototype.onHeaderClick = /**
     * Header click event handler.
     * @param {?} column Data table column component reference.
     * @param {?} event Mouse event arguments object.
     * @return {?}
     */
    function (column, event) {
        if (!this.resizeInProgress) {
            this.sortData(column);
            this.eventStateService.headerClickStream.emit({ column: column, event: event });
        }
        else {
            this.resizeInProgress = false; // this is because I can't prevent click from mousup of the drag end
        }
    };
    /**
     * Sort data event handler.
     * @param column Data table column component reference.
     */
    /**
     * Sort data event handler.
     * @private
     * @param {?} column Data table column component reference.
     * @return {?}
     */
    DataTableColumnTitleHeaderComponent.prototype.sortData = /**
     * Sort data event handler.
     * @private
     * @param {?} column Data table column component reference.
     * @return {?}
     */
    function (column) {
        if (column.sortable) {
            /** @type {?} */
            var prevSortOrder = column.sortOrder;
            if (column.sortOrder) {
                column.sortOrder = column.getNewSortOrder();
            }
            else {
                if (!this.config.multiColumnSortable) {
                    /** @type {?} */
                    var sortColumns = this.columns.filter((/**
                     * @param {?} item
                     * @return {?}
                     */
                    function (item) { return item.sortable; }));
                    sortColumns.forEach((/**
                     * @param {?} sortColumn
                     * @return {?}
                     */
                    function (sortColumn) {
                        if (sortColumn !== column) {
                            sortColumn.sortOrder = '';
                        }
                    }));
                }
                column.sortOrder = 'asc';
            }
            if (this.config.multiColumnSortable) {
                if (column.sortOrder === '') {
                    /** @type {?} */
                    var sortColumns = this.columns.filter((/**
                     * @param {?} item
                     * @return {?}
                     */
                    function (item) { return item.sortable; }));
                    sortColumns.forEach((/**
                     * @param {?} sortColumn
                     * @return {?}
                     */
                    function (sortColumn) {
                        if (sortColumn !== column && sortColumn.sortPriority > column.sortPriority) {
                            --sortColumn.sortPriority;
                        }
                    }));
                    column.sortPriority = undefined;
                    --this.dataStateService.currentSortPriority;
                }
                else {
                    if (!prevSortOrder) {
                        column.sortPriority = ++this.dataStateService.currentSortPriority;
                    }
                }
            }
            this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
        }
    };
    /**
     * Set column width.
     * @param width Width value in pixels.
     * @param column Data table column component reference.
     */
    /**
     * Set column width.
     * @param {?} width Width value in pixels.
     * @param {?} column Data table column component reference.
     * @return {?}
     */
    DataTableColumnTitleHeaderComponent.prototype.setColumnWidth = /**
     * Set column width.
     * @param {?} width Width value in pixels.
     * @param {?} column Data table column component reference.
     * @return {?}
     */
    function (width, column) {
        column.actualWidth = width;
    };
    /**
     * Column resize event handler.
     * @param event Resize mouse event.
     * @param column Data table column component.
     * @param columnElement Table header cell element DOM reference.
     */
    /**
     * Column resize event handler.
     * @param {?} event Resize mouse event.
     * @param {?} column Data table column component.
     * @param {?} columnElement Table header cell element DOM reference.
     * @return {?}
     */
    DataTableColumnTitleHeaderComponent.prototype.onColumnResize = /**
     * Column resize event handler.
     * @param {?} event Resize mouse event.
     * @param {?} column Data table column component.
     * @param {?} columnElement Table header cell element DOM reference.
     * @return {?}
     */
    function (event, column, columnElement) {
        var _this = this;
        this.resizeInProgress = true;
        this.dragAndDropService.drag(event, {
            move: (/**
             * @param {?} moveEvent
             * @param {?} dx
             * @return {?}
             */
            function (moveEvent, dx) {
                /** @type {?} */
                var newWidth = columnElement.offsetWidth + dx;
                if (column.resizeMinLimit !== undefined && newWidth < column.resizeMinLimit) {
                    return;
                }
                column.actualWidth = newWidth;
                /** @type {?} */
                var totalWidth = 0;
                _this.columns.forEach((/**
                 * @param {?} col
                 * @return {?}
                 */
                function (col) {
                    col.width = col.actualWidth;
                    totalWidth += col.width;
                }));
                _this.dataStateService.tableWidth = totalWidth;
            })
        });
    };
    Object.defineProperty(DataTableColumnTitleHeaderComponent.prototype, "allRowSelected", {
        /**
         * Get all row selected state.
         */
        get: /**
         * Get all row selected state.
         * @return {?}
         */
        function () {
            return this.dataStateService.allRowSelected;
        },
        /**
         * Set all row selected state.
         * @param value All row selected status.
         */
        set: /**
         * Set all row selected state.
         * @param {?} value All row selected status.
         * @return {?}
         */
        function (value) {
            this.dataStateService.allRowSelected = value;
            this.allRowSelectedChanged(this.dataStateService.allRowSelected);
            this.eventStateService.allRowSelectChangeStream.emit(this.dataStateService.allRowSelected);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * All row select change event handler.
     * @param selectedState Row selected status.
     */
    /**
     * All row select change event handler.
     * @private
     * @param {?} selectedState Row selected status.
     * @return {?}
     */
    DataTableColumnTitleHeaderComponent.prototype.allRowSelectedChanged = /**
     * All row select change event handler.
     * @private
     * @param {?} selectedState Row selected status.
     * @return {?}
     */
    function (selectedState) {
        var _this = this;
        this.dataStateService.dataRows.forEach((/**
         * @param {?} row
         * @return {?}
         */
        function (row) {
            /** @type {?} */
            var id = get(row.item, _this.config.selectTrackBy);
            /** @type {?} */
            var index = _this.dataStateService.selectedRows.indexOf(id);
            if (selectedState && index < 0) {
                _this.dataStateService.selectedRows.push(id);
            }
            else if (!selectedState && index > -1) {
                _this.dataStateService.selectedRows.splice(index, 1);
            }
            row.selected = selectedState;
        }));
        this.eventStateService.rowSelectChangeStream.emit(this.dataStateService.selectedRows);
    };
    Object.defineProperty(DataTableColumnTitleHeaderComponent.prototype, "showAllRowSelectCheckbox", {
        /**
         * Get all row select checkbox display status.
         * @return True if all row select checkbox should be displayed.
         */
        get: /**
         * Get all row select checkbox display status.
         * @return {?} True if all row select checkbox should be displayed.
         */
        function () {
            return this.config.selectMode === 'multi' && this.config.showRowSelectAllCheckbox;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} column
     * @return {?}
     */
    DataTableColumnTitleHeaderComponent.prototype.showSortPriorityLabel = /**
     * @param {?} column
     * @return {?}
     */
    function (column) {
        return !!(this.config.multiColumnSortable && this.dataStateService.currentSortPriority > 1 && column.sortPriority);
    };
    DataTableColumnTitleHeaderComponent.decorators = [
        { type: Component, args: [{
                    exportAs: 'ngDataTableColumnTitleHeader',
                    // tslint:disable-next-line
                    selector: '[ngDataTableColumnTitleHeader]',
                    template: "<th class=\"ng-data-table-expand-column-header\" *ngIf=\"config.expandableRows\"></th>\n<th class=\"ng-data-table-index-column-header\" *ngIf=\"config.showIndexColumn\">\n  <span [textContent]=\"config.indexColumnTitle\"></span>\n</th>\n<th class=\"ng-data-table-select-column-header\" *ngIf=\"config.showRowSelectCheckboxColumn\">\n  <div class=\"ng-data-table-checkbox-container\" *ngIf=\"showAllRowSelectCheckbox\">\n    <input class=\"ng-data-table-checkbox-input\" type=\"checkbox\"\n           [id]=\"dataStateService.getUniqueId('rsa', 0)\" [(ngModel)]=\"allRowSelected\"/>\n    <label [for]=\"dataStateService.getUniqueId('rsa', 0)\"></label>\n  </div>\n</th>\n<ng-container *ngFor=\"let column of columns\">\n  <th class=\"ng-data-table-column-header\"\n    #columnHeader\n    *ngIf=\"column.visible\"\n    (click)=\"onHeaderClick(column, $event)\"\n    (ngElementWidth)=\"setColumnWidth($event, column)\"\n    [class.sortable]=\"column.sortable\"\n    [class.resizable]=\"column.resizable\"\n    [ngClass]=\"column.cssClass\">\n    <span class=\"ng-data-table-column-sort-priority\" *ngIf=\"showSortPriorityLabel(column)\"><small>{{column.sortPriority}}</small></span>\n    <span *ngIf=\"!column.headerTemplate\" [textContent]=\"column.title\" class=\"ng-data-table-column-header-label\"></span>\n    <ng-container *ngIf=\"column.headerTemplate\" [ngTemplateOutlet]=\"column.headerTemplate\" [ngTemplateOutletContext]=\"{ column: column }\">\n    </ng-container>\n    <span class=\"ng-data-table-column-sort-icon\" [hidden]=\"!column.sortable\" [ngClass]=\"column.getSortIconClass()\"></span>\n    <span class=\"ng-data-table-column-resize-handle\" *ngIf=\"column.resizable\" (mousedown)=\"onColumnResize($event, column, columnHeader)\"> </span>\n  </th>\n</ng-container>\n"
                }] }
    ];
    /** @nocollapse */
    DataTableColumnTitleHeaderComponent.ctorParameters = function () { return [
        { type: DragAndDropService },
        { type: DataTableEventStateService },
        { type: DataTableDataStateService },
        { type: DataTableConfigService }
    ]; };
    DataTableColumnTitleHeaderComponent.propDecorators = {
        columns: [{ type: Input }]
    };
    return DataTableColumnTitleHeaderComponent;
}());
export { DataTableColumnTitleHeaderComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTableColumnTitleHeaderComponent.prototype.resizeInProgress;
    /** @type {?} */
    DataTableColumnTitleHeaderComponent.prototype.columns;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnTitleHeaderComponent.prototype.dragAndDropService;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnTitleHeaderComponent.prototype.eventStateService;
    /** @type {?} */
    DataTableColumnTitleHeaderComponent.prototype.dataStateService;
    /** @type {?} */
    DataTableColumnTitleHeaderComponent.prototype.config;
}
