// ornamentum/esm2015/data-table-resource/data-table-resource.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ResourceUtilityModule } from '../resource-utility/resource-utility.module';
import { DataTableResourceConfigService } from './services/data-table-resource-config.service';
import { DataTableHttpResourceFactoryService } from './services/data-table-http-resource-factory.service';
import { DataTableWebsocketResourceFactoryService } from './services/data-table-websocket-resource-factory.service';
/** @type {?} */
const PROVIDERS = [DataTableHttpResourceFactoryService, DataTableWebsocketResourceFactoryService];
/**
 * Data table data source module.
 */
export class DataTableResourceModule {
    /**
     * Set module root configuration overrides
     * @return {?} Module with custom providers
     */
    static forRoot() {
        return {
            ngModule: DataTableResourceModule,
            providers: [DataTableResourceConfigService]
        };
    }
}
DataTableResourceModule.decorators = [
    { type: NgModule, args: [{
                imports: [ResourceUtilityModule],
                providers: [...PROVIDERS]
            },] }
];
export { DataTableHttpDataFetchService } from './services/data-table-http-data-fetch.service';
export { DataTableHttpResourceFactoryService };
export { DataTableWebsocketDataFetchService } from './services/data-table-websocket-data-fetch.service';
export { DataTableWebsocketResourceFactoryService };
