// ornamentum/esm5/data-table/components/data-table/data-table.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ContentChild, ContentChildren, EventEmitter, Input, Output, QueryList, TemplateRef, forwardRef, ElementRef, ViewChild, NgZone } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { Observable, of } from 'rxjs';
import { catchError, debounceTime, switchMap } from 'rxjs/operators';
import { get } from '../../../utility/services/object-utility.class';
import { DataFetchMode } from '../../models/data-fetch-mode.enum';
import { DataTableColumnComponent } from '../data-table-column/data-table-column.component';
import { DragAndDropService, GlobalRefService } from '../../../utility/utility.module';
import { ValidatorService } from '../../../utility/services/validator.service';
import { DataTableEventStateService } from '../../services/data-table-event.service';
import { DataTableDataStateService } from '../../services/data-table-data-state.service';
import { DataTablePersistenceService } from '../../services/data-table-persistence.service';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableScrollPositionService } from '../../services/data-table-scroll-position.service';
import { DataTableResourceService } from '../../services/data-table-resource.service';
/**
 * Data table component; Data table entry component.
 */
var DataTableComponent = /** @class */ (function () {
    function DataTableComponent(dragAndDropService, dataTableStateService, globalRefService, eventStateService, dataTableResourceService, zone, dataStateService, scrollPositionService, config) {
        this.dragAndDropService = dragAndDropService;
        this.dataTableStateService = dataTableStateService;
        this.globalRefService = globalRefService;
        this.eventStateService = eventStateService;
        this.dataTableResourceService = dataTableResourceService;
        this.zone = zone;
        this.dataStateService = dataStateService;
        this.scrollPositionService = scrollPositionService;
        this.config = config;
        this.storageMode = config.storageMode;
        this.headerClick = this.eventStateService.headerClickStream;
        this.allRowSelectChange = this.eventStateService.allRowSelectChangeStream;
        this.rowBind = this.eventStateService.rowBindStream;
        this.rowClick = this.eventStateService.rowClickStream;
        this.rowDoubleClick = this.eventStateService.rowDoubleClickStream;
        this.rowSelectChange = this.eventStateService.rowSelectChangeStream;
        this.cellBind = this.eventStateService.cellBindStream;
        this.cellClick = this.eventStateService.cellClickStream;
        this.init = this.eventStateService.initStream;
        this.dataBound = this.eventStateService.dataBoundStream;
        this.columnBind = this.eventStateService.columnBind;
    }
    Object.defineProperty(DataTableComponent.prototype, "onDataBind", {
        // Input Events
        /**
         * Set data bind event handler callback. This handler is fired on each data fetch request.
         */
        set: 
        // Input Events
        /**
         * Set data bind event handler callback. This handler is fired on each data fetch request.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.onDataBind = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "onFilterValueExtract", {
        /**
         * Set filter value extract event handler callback. Used to extract filter value collection dynamically when
         * explicit data bind functionality is used with onDataBind callback.
         */
        set: /**
         * Set filter value extract event handler callback. Used to extract filter value collection dynamically when
         * explicit data bind functionality is used with onDataBind callback.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.onFilterValueExtract = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "onDynamicRowSpanExtract", {
        /**
         * Set on dynamic row span extract event handler callback.
         */
        set: /**
         * Set on dynamic row span extract event handler callback.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.onDynamicRowSpanExtract = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "items", {
        // Input parameters
        /**
         * Set static data items collection. No need to set data source when static items collection is provided.
         */
        set: 
        // Input parameters
        /**
         * Set static data items collection. No need to set data source when static items collection is provided.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!value) {
                return;
            }
            this.eventStateService.staticDataSourceStream.next(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "dataSource", {
        /**
         * Set data source observable. This observable is used to retrieve row data for binding.
         */
        set: /**
         * Set data source observable. This observable is used to retrieve row data for binding.
         * @param {?} source
         * @return {?}
         */
        function (source) {
            this.initDataSource(source);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "id", {
        /**
         * Set data table unique identifier.
         */
        set: /**
         * Set data table unique identifier.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!ValidatorService.idPatternValidatorExpression.test(value)) {
                throw Error('Invalid [id] input value. Unique identifier parameter only accept string begin with a letter ([A-Za-z]) and may be followed by any number of letters, digits ([0-9]), hyphens ("-"), underscores ("_").');
            }
            this.dataStateService.id = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "persistTableState", {
        /**
         * Set persist table state on provided storage mode if true. Depends on storageMode property.
         */
        set: /**
         * Set persist table state on provided storage mode if true. Depends on storageMode property.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.persistTableState = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "storageMode", {
        /**
         * Set storage mode to persist table state. Only applicable when persistTableState is true.
         */
        set: /**
         * Set storage mode to persist table state. Only applicable when persistTableState is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataTableStateService.storageMode = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "multiColumnSortable", {
        /**
         * Set multiple column sortable if true. Only applicable for sortable true columns.
         */
        set: /**
         * Set multiple column sortable if true. Only applicable for sortable true columns.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.multiColumnSortable = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showHeader", {
        /**
         * Set table header bar visible if true.
         */
        set: /**
         * Set table header bar visible if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showHeader = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "title", {
        /**
         * Set title to be shown in the header. Only applicable when showHeader is true.
         */
        set: /**
         * Set title to be shown in the header. Only applicable when showHeader is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "width", {
        /**
         * Set width value in pixels. Can be used to set the width of teh table (responsive if not set).
         */
        set: /**
         * Set width value in pixels. Can be used to set the width of teh table (responsive if not set).
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.width = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "minContentHeight", {
        /**
         * Set minimum table content height value in pixels. Can be used to set the minimum height of the table content area.
         */
        set: /**
         * Set minimum table content height value in pixels. Can be used to set the minimum height of the table content area.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.minContentHeight = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "minContentWidth", {
        /**
         * Minimum table content width value in pixels. Can be used to set the minimum width of the table content area.
         */
        set: /**
         * Minimum table content width value in pixels. Can be used to set the minimum width of the table content area.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.minContentWidth = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "contentHeight", {
        /**
         * Table content height value in pixels. This configuration can be used to enable table content vertical
         * scrolling for responsive design.
         */
        set: /**
         * Table content height value in pixels. This configuration can be used to enable table content vertical
         * scrolling for responsive design.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.contentHeight = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "pageable", {
        /**
         * Show pagination bar if true. Depends on offset and limit values. Trigger dataLoad event with offset
         * and limit values.
         */
        set: /**
         * Show pagination bar if true. Depends on offset and limit values. Trigger dataLoad event with offset
         * and limit values.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.pageable = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "loadOnScroll", {
        /**
         * Enable scrolling based on-demand data loading functionality if true. Trigger dataLoad event with offset
         * and limit values when scroll to bottom until data source exhaust.
         */
        set: /**
         * Enable scrolling based on-demand data loading functionality if true. Trigger dataLoad event with offset
         * and limit values when scroll to bottom until data source exhaust.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.loadOnScroll = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "loadViewDistanceRatio", {
        /**
         * Set view height distance ratio to trigger data fetch on scroll. Applicable only when load on scroll mode is enabled.
         */
        set: /**
         * Set view height distance ratio to trigger data fetch on scroll. Applicable only when load on scroll mode is enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.loadViewDistanceRatio = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showIndexColumn", {
        /**
         * Set auto generated index column with row numbering if true.
         */
        set: /**
         * Set auto generated index column with row numbering if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showIndexColumn = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "indexColumnTitle", {
        /**
         * Set index column header title. Applicable when showIndexColumn is true.
         */
        set: /**
         * Set index column header title. Applicable when showIndexColumn is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.indexColumnTitle = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "rowSelectable", {
        /**
         * Set row select checkbox and select state if true.
         */
        set: /**
         * Set row select checkbox and select state if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.rowSelectable = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "selectMode", {
        /**
         * Data table row select mode. Applicable only when rowSelectable is true.
         */
        set: /**
         * Data table row select mode. Applicable only when rowSelectable is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.selectMode = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showRowSelectCheckbox", {
        /**
         * Set select all row checkbox on column header visible if true.
         * Only applicable when showRowSelectCheckbox, rowSelectable is true & item selectMode is  multi.
         */
        set: /**
         * Set select all row checkbox on column header visible if true.
         * Only applicable when showRowSelectCheckbox, rowSelectable is true & item selectMode is  multi.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showRowSelectCheckbox = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showRowSelectAllCheckbox", {
        /**
         * Set select all row checkbox on column header visible if true.
         * Only applicable when showRowSelectCheckbox, rowSelectable is true & item selectMode is  multi.
         */
        set: /**
         * Set select all row checkbox on column header visible if true.
         * Only applicable when showRowSelectCheckbox, rowSelectable is true & item selectMode is  multi.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showRowSelectAllCheckbox = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showSubstituteRows", {
        /**
         * Set substitute rows visible if true. Fill with empty rows when row count < limit.
         */
        set: /**
         * Set substitute rows visible if true. Fill with empty rows when row count < limit.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showSubstituteRows = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "expandableRows", {
        /**
         * Set row expander visible if true. Render ngDataTableExpand template on expand click.
         */
        set: /**
         * Set row expander visible if true. Render ngDataTableExpand template on expand click.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.expandableRows = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "selectOnRowClick", {
        /**
         * Set trigger row select on click event if true. Applicable only when rowSelectable is true.
         */
        set: /**
         * Set trigger row select on click event if true. Applicable only when rowSelectable is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.selectOnRowClick = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "expandOnRowClick", {
        /**
         * Set expand and render expand template on row click if true. Only applicable when expandableRows is true.
         */
        set: /**
         * Set expand and render expand template on row click if true. Only applicable when expandableRows is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.expandOnRowClick = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "autoFetch", {
        /**
         * Auto trigger dataLoad event on initialization if true.
         */
        set: /**
         * Auto trigger dataLoad event on initialization if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.autoFetch = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showLoadingSpinner", {
        /**
         * Set loading spinner visible if true. Show loading spinner when data fetch operation is triggered.
         */
        set: /**
         * Set loading spinner visible if true. Show loading spinner when data fetch operation is triggered.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showLoadingSpinner = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "selectTrackBy", {
        /**
         * Set select option track by field path which is used to uniquely identify row for selection tracking.
         * This field support object paths expressions 'root[0].nest'.
         */
        set: /**
         * Set select option track by field path which is used to uniquely identify row for selection tracking.
         * This field support object paths expressions 'root[0].nest'.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.selectTrackBy = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "selectedRow", {
        /**
         * Set selected row identifier. Select specified row on initial load.
         * Applicable when row select mode is SINGLE or SINGLE_TOGGLE.
         */
        set: /**
         * Set selected row identifier. Select specified row on initial load.
         * Applicable when row select mode is SINGLE or SINGLE_TOGGLE.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.selectedRow = value;
            this.eventStateService.rowSelectChangeStream.emit(this.dataStateService.selectedRow);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "selectedRows", {
        /**
         * Set selected row identifiers collection. Select specified rows on initial load.
         * Applicable when selectMode is SINGLE or SINGLE_TOGGLE true.
         */
        set: /**
         * Set selected row identifiers collection. Select specified rows on initial load.
         * Applicable when selectMode is SINGLE or SINGLE_TOGGLE true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.selectedRows = value || [];
            this.eventStateService.rowSelectChangeStream.emit(this.dataStateService.selectedRows);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "filterDebounceTime", {
        /**
         * Set filter debounce time in milliseconds. Applicable only when filterDebounce is true.
         */
        set: /**
         * Set filter debounce time in milliseconds. Applicable only when filterDebounce is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.filterDebounceTime = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "filterDebounce", {
        /**
         * Set filter data debounce enabled state with provided filterDebounceTime if true.
         */
        set: /**
         * Set filter data debounce enabled state with provided filterDebounceTime if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.filterDebounce = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showRefreshButton", {
        /**
         * Set refresh button visible if true. Only applicable when showHeader is true.
         */
        set: /**
         * Set refresh button visible if true. Only applicable when showHeader is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showRefreshButton = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showColumnSelector", {
        /**
         * Row selector column width in pixels. Applicable only when showColumnSelector is true.
         */
        set: /**
         * Row selector column width in pixels. Applicable only when showColumnSelector is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showColumnSelector = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "columnSelectorWidth", {
        /**
         * Set column selector dropdown width in pixels. Only applicable when showColumnSelector is true.
         */
        set: /**
         * Set column selector dropdown width in pixels. Only applicable when showColumnSelector is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.columnSelectorWidth = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "expanderColumnWidth", {
        /**
         * Set expander column width in pixels. Applicable only when expandableRows is true.
         */
        set: /**
         * Set expander column width in pixels. Applicable only when expandableRows is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.expanderColumnWidth = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "indexColumnWidth", {
        /**
         * Set index column width in pixels. Applicable only when showIndexColumn is true.
         */
        set: /**
         * Set index column width in pixels. Applicable only when showIndexColumn is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.indexColumnWidth = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "selectionColumnWidth", {
        /**
         * Set row selector column width in pixels. Applicable only when showColumnSelector is true.
         */
        set: /**
         * Set row selector column width in pixels. Applicable only when showColumnSelector is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.selectionColumnWidth = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "translations", {
        /**
         * Set translation data object. Used to localize table static label text.
         */
        set: /**
         * Set translation data object. Used to localize table static label text.
         * @param {?} data
         * @return {?}
         */
        function (data) {
            this.config.translations = data;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "showRowExpandLoadingSpinner", {
        /**
         * Set row expand loading spinner visible if true. Applicable only when row expand is enabled.
         */
        set: /**
         * Set row expand loading spinner visible if true. Applicable only when row expand is enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showRowExpandLoadingSpinner = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "offset", {
        /**
         * Set data offset value (start offset index). Applicable only when pageable is true.
         */
        set: /**
         * Set data offset value (start offset index). Applicable only when pageable is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.offset = value;
            this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "limit", {
        /**
         * Set data limit value (page size). Applicable only when pageable is true.
         */
        set: /**
         * Set data limit value (page size). Applicable only when pageable is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.limit = value;
            this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "page", {
        /**
         * Get current page number.
         */
        get: /**
         * Get current page number.
         * @return {?}
         */
        function () {
            return Math.floor(this.config.offset / this.config.limit) + 1;
        },
        /**
         * Set current page number. Auto calculate offset depending on page number,
         * do not explicitly set offset when page is used.
         */
        set: /**
         * Set current page number. Auto calculate offset depending on page number,
         * do not explicitly set offset when page is used.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.offset = (value - 1) * this.config.limit;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "headerPadding", {
        /**
         * Get data table header padding in pixels.
         */
        get: /**
         * Get data table header padding in pixels.
         * @return {?}
         */
        function () {
            return this.config.contentHeight ? this.globalRefService.scrollbarWidth : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableComponent.prototype, "isLoading", {
        /**
         * Get data loading status.
         */
        get: /**
         * Get data loading status.
         * @return {?}
         */
        function () {
            return !(this.config.loadOnScroll && this.dataStateService.dataRows.length)
                && this.config.showLoadingSpinner && this.dataStateService.dataLoading;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * On after data bind event handler
     * @param queryResult Query result object
     */
    /**
     * On after data bind event handler
     * @private
     * @param {?} queryResult Query result object
     * @return {?}
     */
    DataTableComponent.prototype.onAfterDataBind = /**
     * On after data bind event handler
     * @private
     * @param {?} queryResult Query result object
     * @return {?}
     */
    function (queryResult) {
        this.dataStateService.itemCount = queryResult.count;
        this.setDataRows(queryResult.items);
        if (this.dataStateService.heardReload) {
            this.eventStateService.fetchFilterOptionsStream.next(false);
            this.dataStateService.heardReload = false;
        }
        this.dataStateService.dataLoading = false;
        this.eventStateService.dataBoundStream.emit();
    };
    /**
     * Get data item selected state
     * @param item Data item object
     * @return True if item is selected
     */
    /**
     * Get data item selected state
     * @private
     * @param {?} item Data item object
     * @return {?} True if item is selected
     */
    DataTableComponent.prototype.getSelectedState = /**
     * Get data item selected state
     * @private
     * @param {?} item Data item object
     * @return {?} True if item is selected
     */
    function (item) {
        /** @type {?} */
        var id = get(item, this.config.selectTrackBy);
        if (id === undefined) {
            return false;
        }
        if (this.config.selectMode === 'multi') {
            return this.dataStateService.selectedRows.indexOf(id) > -1;
        }
        return this.dataStateService.selectedRow === id;
    };
    /**
     * Set data table item collection
     * @param items Data table item collection
     */
    /**
     * Set data table item collection
     * @private
     * @param {?} items Data table item collection
     * @return {?}
     */
    DataTableComponent.prototype.setDataRows = /**
     * Set data table item collection
     * @private
     * @param {?} items Data table item collection
     * @return {?}
     */
    function (items) {
        var _this = this;
        /** @type {?} */
        var mappedItems = items.map((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        function (item, index) {
            /** @type {?} */
            var currentIndex;
            if (_this.config.loadOnScroll || _this.config.pageable) {
                currentIndex = _this.config.offset + index + 1;
            }
            else {
                currentIndex = index + 1;
            }
            return {
                dataLoaded: false,
                expanded: false,
                disabled: false,
                color: '',
                cssClass: '',
                tooltip: '',
                index: currentIndex,
                item: item,
                selected: _this.getSelectedState(item)
            };
        }));
        if (this.config.loadOnScroll) {
            this.dataStateService.dataRows = tslib_1.__spread(this.dataStateService.dataRows, mappedItems);
        }
        else {
            this.dataStateService.dataRows = mappedItems;
        }
        if (this.config.selectMode === 'multi') {
            this.dataStateService.allRowSelected =
                this.dataStateService.dataRows.length !== 0 &&
                    this.dataStateService.dataRows.every((/**
                     * @param {?} dataRow
                     * @return {?}
                     */
                    function (dataRow) {
                        return dataRow.selected;
                    }));
        }
        if (!this.config.loadOnScroll) {
            /** @type {?} */
            var substituteRowCount = this.config.limit - this.dataStateService.dataRows.length;
            this.dataStateService.substituteRows = Array.from({
                length: substituteRowCount
            });
        }
    };
    /**
     * Initialize data fetch event
     */
    /**
     * Initialize data fetch event
     * @private
     * @return {?}
     */
    DataTableComponent.prototype.initDataFetchEvent = /**
     * Initialize data fetch event
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var noop = {
            items: [],
            count: 0
        };
        this.dataFetchStreamSubscription = this.eventStateService.dataFetchStream
            .pipe(debounceTime(20), switchMap((/**
         * @param {?} fetchMode
         * @return {?}
         */
        function (fetchMode) { return _this.mapDataBind(fetchMode); })), catchError((/**
         * @return {?}
         */
        function () {
            return of(noop);
        })))
            .subscribe((/**
         * @param {?} queryResult
         * @return {?}
         */
        function (queryResult) {
            _this.onAfterDataBind(queryResult);
        }), (/**
         * @return {?}
         */
        function () {
            _this.onAfterDataBind(noop);
        }));
    };
    /**
     * Re-map data binding data
     * @param fetchMode Data fetch mode
     * @return Data table query result stream
     */
    /**
     * Re-map data binding data
     * @private
     * @param {?} fetchMode Data fetch mode
     * @return {?} Data table query result stream
     */
    DataTableComponent.prototype.mapDataBind = /**
     * Re-map data binding data
     * @private
     * @param {?} fetchMode Data fetch mode
     * @return {?} Data table query result stream
     */
    function (fetchMode) {
        this.dataStateService.dataLoading = true;
        if (fetchMode === DataFetchMode.HARD_RELOAD) {
            this.clearRowSelectState();
            this.clearColumnState();
            this.dataStateService.heardReload = true;
            this.config.offset = 0;
        }
        /** @type {?} */
        var params = {
            loadData: fetchMode === DataFetchMode.HARD_RELOAD || fetchMode === DataFetchMode.SOFT_RELOAD
        };
        if (this.columns) {
            params.fields = this.columns
                .filter((/**
             * @param {?} column
             * @return {?}
             */
            function (column) {
                return column.sortable || column.filterable;
            }))
                .reduce((/**
             * @param {?} acc
             * @param {?} column
             * @return {?}
             */
            function (acc, column) {
                if (column.sortField || column.filterField) {
                    acc.push({
                        field: column.sortField || column.filterField,
                        column: column
                    });
                }
                else {
                    acc.push({
                        field: column.field,
                        column: column
                    });
                }
                return acc;
            }), [])
                .map((/**
             * @param {?} uniqueField
             * @return {?}
             */
            function (uniqueField) {
                /** @type {?} */
                var filter;
                if (uniqueField.column.showDropdownFilter) {
                    if (uniqueField.column.dropdownFilterSelectMode === 'multi') {
                        filter = uniqueField.column.filter && uniqueField.column.filter.map((/**
                         * @param {?} filterValue
                         * @return {?}
                         */
                        function (filterValue) { return filterValue.key; }));
                    }
                    else {
                        filter = uniqueField.column.filter && uniqueField.column.filter.key;
                    }
                }
                else {
                    filter = uniqueField.column.filter;
                }
                return {
                    field: uniqueField.field,
                    sortable: uniqueField.column.sortable,
                    sortOrder: uniqueField.column.sortOrder,
                    sortPriority: uniqueField.column.sortPriority || (uniqueField.column.sortOrder ? 1 : 0),
                    filterable: uniqueField.column.filterable,
                    filterValue: filter,
                    filterExpression: uniqueField.column.filterExpression,
                };
            }));
        }
        if (this.config.pageable || this.config.loadOnScroll) {
            params.offset = this.config.offset;
            params.limit = this.config.limit;
        }
        if (this.config.persistTableState) {
            this.dataTableStateService.setState(this.dataStateService.id, params);
        }
        return this.dataStateService.onDataBind(params);
    };
    /**
     * Initialize data table state via previous state snapshot; Applicable only when persist table state is enabled
     */
    /**
     * Initialize data table state via previous state snapshot; Applicable only when persist table state is enabled
     * @private
     * @return {?}
     */
    DataTableComponent.prototype.initDataTableState = /**
     * Initialize data table state via previous state snapshot; Applicable only when persist table state is enabled
     * @private
     * @return {?}
     */
    function () {
        if (this.config.persistTableState) {
            /** @type {?} */
            var dataTableState_1 = this.dataTableStateService.getState(this.dataStateService.id);
            if (dataTableState_1) {
                this.columns.forEach((/**
                 * @param {?} column
                 * @return {?}
                 */
                function (column) {
                    /** @type {?} */
                    var field = dataTableState_1.fields.find((/**
                     * @param {?} col
                     * @return {?}
                     */
                    function (col) {
                        return col.field === column.field;
                    }));
                    if (field) {
                        if (column.filterable && field.filterable) {
                            if (column.showDropdownFilter) {
                                if (field.filterValue) {
                                    if (column.dropdownFilterSelectMode === 'multi') {
                                        column.filter = field.filterValue.map((/**
                                         * @param {?} filterValue
                                         * @return {?}
                                         */
                                        function (filterValue) {
                                            return {
                                                key: filterValue,
                                                value: filterValue
                                            };
                                        }));
                                    }
                                    else {
                                        column.filter = {
                                            key: field.filterValue,
                                            value: field.filterValue
                                        };
                                    }
                                }
                            }
                            else {
                                column.filter = field.filterValue;
                            }
                        }
                        if (column.sortable && field.sortable) {
                            column.sortOrder = field.sortOrder;
                        }
                    }
                }));
                this.config.limit = dataTableState_1.limit;
                this.config.offset = dataTableState_1.offset;
            }
        }
    };
    /**
     * After component initialize lifecycle event handler
     */
    /**
     * After component initialize lifecycle event handler
     * @return {?}
     */
    DataTableComponent.prototype.ngAfterContentInit = /**
     * After component initialize lifecycle event handler
     * @return {?}
     */
    function () {
        var _this = this;
        this.dataStateService.relativeParentElement = this.dataTableElement.nativeElement;
        if (!this.dataStateService.onDataBind) {
            this.dataSource = this.eventStateService.staticDataSourceStream;
        }
        this.initDataTableState();
        this.initDataFetchEvent();
        if (this.config.autoFetch) {
            this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
        }
        this.eventStateService.fetchFilterOptionsStream.next(true);
        this.eventStateService.initStream.emit(this);
        if (this.config.loadOnScroll) {
            this.scrollPositionSubscription = this.scrollPositionService.scrollPositionStream.subscribe((/**
             * @param {?} pos
             * @return {?}
             */
            function (pos) {
                /** @type {?} */
                var roundingPixel = 1;
                /** @type {?} */
                var gutterPixel = 1;
                if (pos.isVertical
                    && pos.scrollTop >= pos.scrollHeight - (1 + _this.config.loadViewDistanceRatio) * pos.clientHeight - roundingPixel - gutterPixel
                    && (_this.config.offset + _this.config.limit) < _this.dataStateService.itemCount
                    && !_this.dataStateService.dataLoading) {
                    _this.dataStateService.dataLoading = true;
                    _this.zone.run((/**
                     * @return {?}
                     */
                    function () {
                        _this.offset = _this.config.offset + _this.config.limit;
                    }));
                }
            }));
        }
    };
    /**
     * Reset column sort and filter state
     */
    /**
     * Reset column sort and filter state
     * @private
     * @return {?}
     */
    DataTableComponent.prototype.clearColumnState = /**
     * Reset column sort and filter state
     * @private
     * @return {?}
     */
    function () {
        this.columns.forEach((/**
         * @param {?} column
         * @return {?}
         */
        function (column) {
            column.resetSortOrder();
            column.filter = undefined;
        }));
    };
    /**
     * Clear selected data row status
     */
    /**
     * Clear selected data row status
     * @private
     * @return {?}
     */
    DataTableComponent.prototype.clearRowSelectState = /**
     * Clear selected data row status
     * @private
     * @return {?}
     */
    function () {
        this.dataStateService.selectedRow = undefined;
        this.dataStateService.selectedRows = [];
        this.dataStateService.allRowSelected = false;
    };
    /**
     * Fetch data from data source
     * @param fetchMode Data fetch mode
     */
    /**
     * Fetch data from data source
     * @param {?=} fetchMode Data fetch mode
     * @return {?}
     */
    DataTableComponent.prototype.fetchData = /**
     * Fetch data from data source
     * @param {?=} fetchMode Data fetch mode
     * @return {?}
     */
    function (fetchMode) {
        if (fetchMode === void 0) { fetchMode = DataFetchMode.SOFT_RELOAD; }
        this.eventStateService.dataFetchStream.next(fetchMode);
    };
    /**
     * Initialize data source
     * @param source Data source stream
     */
    /**
     * Initialize data source
     * @param {?} source Data source stream
     * @return {?}
     */
    DataTableComponent.prototype.initDataSource = /**
     * Initialize data source
     * @param {?} source Data source stream
     * @return {?}
     */
    function (source) {
        var _this = this;
        this.dataTableResourceService.setDataSource(source);
        this.onDataBind = (/**
         * @param {?} params
         * @return {?}
         */
        function (params) {
            if (params.loadData) {
                _this.dataTableResourceService.setDataSource(source);
            }
            return _this.dataTableResourceService.query(params);
        });
        this.onFilterValueExtract = (/**
         * @param {?} column
         * @return {?}
         */
        function (column) {
            return _this.dataTableResourceService.extractFilterOptions(column);
        });
    };
    /**
     * Component value write event handler; Form control support implementation
     */
    /**
     * Component value write event handler; Form control support implementation
     * @param {?} value
     * @return {?}
     */
    DataTableComponent.prototype.writeValue = /**
     * Component value write event handler; Form control support implementation
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.config.selectMode === 'multi') {
            this.selectedRows = value;
        }
        else {
            this.selectedRow = value;
        }
    };
    /**
     * Register select change event handler; Form control support implementation
     * @param onSelectChange Select change event handler callback
     */
    /**
     * Register select change event handler; Form control support implementation
     * @param {?} onSelectChange Select change event handler callback
     * @return {?}
     */
    DataTableComponent.prototype.registerOnChange = /**
     * Register select change event handler; Form control support implementation
     * @param {?} onSelectChange Select change event handler callback
     * @return {?}
     */
    function (onSelectChange) {
        this.rowSelectChangeSubscription = this.eventStateService.rowSelectChangeStream.subscribe((/**
         * @param {?} selectedIds
         * @return {?}
         */
        function (selectedIds) {
            onSelectChange(selectedIds);
        }));
    };
    /**
     * Register on touch event handler; Form control support implementation
     * @param fn Touch event callback handler
     */
    /**
     * Register on touch event handler; Form control support implementation
     * @param {?} fn Touch event callback handler
     * @return {?}
     */
    DataTableComponent.prototype.registerOnTouched = /**
     * Register on touch event handler; Form control support implementation
     * @param {?} fn Touch event callback handler
     * @return {?}
     */
    function (fn) { };
    Object.defineProperty(DataTableComponent.prototype, "tableWidth", {
        /**
         * Get table width in pixels
         */
        get: /**
         * Get table width in pixels
         * @return {?}
         */
        function () {
            return this.config.width || this.dataStateService.tableWidth;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DataTableComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.dataStateService.id) {
            throw Error('Missing required parameter value for [id] input.');
        }
        if (this.config.loadOnScroll) {
            if (!this.config.minContentHeight) {
                throw Error('[minContentHeight] is required when [infiniteScrollable] is enabled.');
            }
            if (this.config.pageable) {
                throw Error('[pageable] and [infiniteScrollable] cannot be enabled at the same time.');
            }
        }
    };
    /**
     * Component destroy lifecycle event handler
     */
    /**
     * Component destroy lifecycle event handler
     * @return {?}
     */
    DataTableComponent.prototype.ngOnDestroy = /**
     * Component destroy lifecycle event handler
     * @return {?}
     */
    function () {
        if (this.dataFetchStreamSubscription) {
            this.dataFetchStreamSubscription.unsubscribe();
        }
        if (this.rowSelectChangeSubscription) {
            this.rowSelectChangeSubscription.unsubscribe();
        }
        if (this.scrollPositionSubscription) {
            this.scrollPositionSubscription.unsubscribe();
        }
        this.dataTableResourceService.dispose();
    };
    DataTableComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-data-table',
                    template: "<div class=\"ng-data-table-container\" #dataTableElement [class.scrollable]=\"config.contentHeight\">\n  <ng-data-table-header *ngIf=\"config.showHeader\" [columns]=\"columns\"> </ng-data-table-header>\n  <div class=\"ng-data-table-content-container\" [ngScrollPosition]=\"scrollPositionService.scrollPositionStream\">\n    <div class=\"ng-data-table-content-header-container\" [style.paddingRight]=\"headerPadding | ngPx\">\n      <div class=\"ng-data-table-content-header\" [ngScrollElement]=\"scrollPositionService.scrollPositionStream\">\n        <table class=\"ng-data-table\" [style.width]=\"tableWidth | ngPx\" [style.minWidth]=\"config.minContentWidth | ngPx\">\n          <colgroup ngDataTableColGroup [columns]=\"columns\"></colgroup>\n          <thead ngDataTableHead [columns]=\"columns\"></thead>\n        </table>\n      </div>\n    </div>\n    <div class=\"ng-data-table-content\" [ngScrollPosition]=\"scrollPositionService.scrollPositionStream\"\n         [style.height]=\"config.contentHeight | ngPx\" [style.minHeight]=\"config.minContentHeight | ngPx\">\n      <ng-data-table-loading-spinner [loadingSpinnerTemplate]=\"loadingSpinnerTemplate\" [isLoading]=\"isLoading\" [showOverlay]=\"true\">\n      </ng-data-table-loading-spinner>\n      <table class=\"ng-data-table\" [style.width]=\"tableWidth | ngPx\" [style.minWidth]=\"config.minContentWidth | ngPx\" [hidden]=\"isLoading\">\n        <colgroup ngDataTableColGroup [columns]=\"columns\"></colgroup>\n        <tbody\n          ngDataTableBody\n          [columns]=\"columns\"\n          [rowExpandTemplate]=\"rowExpandTemplate\"\n          [rowExpandLoadingSpinnerTemplate]=\"rowExpandLoadingSpinnerTemplate\"\n        ></tbody>\n      </table>\n      <ng-data-table-no-data-body\n        class=\"ng-data-table-no-records-message-container\"\n        *ngIf=\"dataStateService.showNoDataOverlay\"\n        [noRecordsTemplate]=\"noRecordsTemplate\"\n      >\n      </ng-data-table-no-data-body>\n    </div>\n  </div>\n  <ng-data-table-pagination *ngIf=\"config.pageable && dataStateService.itemCount\"></ng-data-table-pagination>\n</div>\n",
                    providers: [
                        DataTableConfigService,
                        DataTableEventStateService,
                        DataTablePersistenceService,
                        DataTableDataStateService,
                        DataTableScrollPositionService,
                        DataTableResourceService,
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return DataTableComponent; })),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    DataTableComponent.ctorParameters = function () { return [
        { type: DragAndDropService },
        { type: DataTablePersistenceService },
        { type: GlobalRefService },
        { type: DataTableEventStateService },
        { type: DataTableResourceService },
        { type: NgZone },
        { type: DataTableDataStateService },
        { type: DataTableScrollPositionService },
        { type: DataTableConfigService }
    ]; };
    DataTableComponent.propDecorators = {
        columns: [{ type: ContentChildren, args: [DataTableColumnComponent,] }],
        rowExpandTemplate: [{ type: ContentChild, args: ['ngDataTableRowExpand', { static: true },] }],
        noRecordsTemplate: [{ type: ContentChild, args: ['ngDataTableNoRecords', { static: true },] }],
        loadingSpinnerTemplate: [{ type: ContentChild, args: ['ngDataTableLoadingSpinner', { static: true },] }],
        rowExpandLoadingSpinnerTemplate: [{ type: ContentChild, args: ['ngDataTableRowExpandLoadingSpinner', { static: true },] }],
        dataTableElement: [{ type: ViewChild, args: ['dataTableElement', { static: true },] }],
        init: [{ type: Output }],
        rowSelectChange: [{ type: Output }],
        rowClick: [{ type: Output }],
        rowDoubleClick: [{ type: Output }],
        headerClick: [{ type: Output }],
        allRowSelectChange: [{ type: Output }],
        cellClick: [{ type: Output }],
        dataBound: [{ type: Output }],
        rowBind: [{ type: Output }],
        columnBind: [{ type: Output }],
        cellBind: [{ type: Output }],
        onDataBind: [{ type: Input }],
        onFilterValueExtract: [{ type: Input }],
        onDynamicRowSpanExtract: [{ type: Input }],
        items: [{ type: Input }],
        dataSource: [{ type: Input }],
        id: [{ type: Input }],
        persistTableState: [{ type: Input }],
        storageMode: [{ type: Input }],
        multiColumnSortable: [{ type: Input }],
        showHeader: [{ type: Input }],
        title: [{ type: Input }],
        width: [{ type: Input }],
        minContentHeight: [{ type: Input }],
        minContentWidth: [{ type: Input }],
        contentHeight: [{ type: Input }],
        pageable: [{ type: Input }],
        loadOnScroll: [{ type: Input }],
        loadViewDistanceRatio: [{ type: Input }],
        showIndexColumn: [{ type: Input }],
        indexColumnTitle: [{ type: Input }],
        rowSelectable: [{ type: Input }],
        selectMode: [{ type: Input }],
        showRowSelectCheckbox: [{ type: Input }],
        showRowSelectAllCheckbox: [{ type: Input }],
        showSubstituteRows: [{ type: Input }],
        expandableRows: [{ type: Input }],
        selectOnRowClick: [{ type: Input }],
        expandOnRowClick: [{ type: Input }],
        autoFetch: [{ type: Input }],
        showLoadingSpinner: [{ type: Input }],
        selectTrackBy: [{ type: Input }],
        selectedRow: [{ type: Input }],
        selectedRows: [{ type: Input }],
        filterDebounceTime: [{ type: Input }],
        filterDebounce: [{ type: Input }],
        showRefreshButton: [{ type: Input }],
        showColumnSelector: [{ type: Input }],
        columnSelectorWidth: [{ type: Input }],
        expanderColumnWidth: [{ type: Input }],
        indexColumnWidth: [{ type: Input }],
        selectionColumnWidth: [{ type: Input }],
        translations: [{ type: Input }],
        showRowExpandLoadingSpinner: [{ type: Input }],
        offset: [{ type: Input }],
        limit: [{ type: Input }],
        page: [{ type: Input }]
    };
    return DataTableComponent;
}());
export { DataTableComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.rowSelectChangeSubscription;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.dataFetchStreamSubscription;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.scrollPositionSubscription;
    /**
     * Data table column component collection.
     * @type {?}
     */
    DataTableComponent.prototype.columns;
    /**
     * Template to display when data row is expanded for detail view.
     * @type {?}
     */
    DataTableComponent.prototype.rowExpandTemplate;
    /**
     * Template to display when data set is empty.
     * @type {?}
     */
    DataTableComponent.prototype.noRecordsTemplate;
    /**
     * Template to display while loading data.
     * @type {?}
     */
    DataTableComponent.prototype.loadingSpinnerTemplate;
    /**
     * Template to display while row is expanding to load detail view.
     * @type {?}
     */
    DataTableComponent.prototype.rowExpandLoadingSpinnerTemplate;
    /**
     * Data table self DOM element reference.
     * @type {?}
     */
    DataTableComponent.prototype.dataTableElement;
    /**
     * Data table initialize event handler.
     * Triggered after data table initialize.
     * @type {?}
     */
    DataTableComponent.prototype.init;
    /**
     * Row selected state change event handler.
     * Triggered when table row selected state change.
     * @type {?}
     */
    DataTableComponent.prototype.rowSelectChange;
    /**
     * Row click event handler.
     * Triggered when data row is clicked.
     * @type {?}
     */
    DataTableComponent.prototype.rowClick;
    /**
     * Row double click event handler.
     * Triggered when data row is double clicked.
     * @type {?}
     */
    DataTableComponent.prototype.rowDoubleClick;
    /**
     * Header click event handler.
     * Triggered when header column clicked.
     * @type {?}
     */
    DataTableComponent.prototype.headerClick;
    /**
     * All row select change event handler.
     * Triggered when all row select state changed.
     * @type {?}
     */
    DataTableComponent.prototype.allRowSelectChange;
    /**
     * Cell click event handler.
     * Triggered when clicked on a cell.
     * @type {?}
     */
    DataTableComponent.prototype.cellClick;
    /**
     * Data bound event handler.
     * Triggered after data bind.
     * @type {?}
     */
    DataTableComponent.prototype.dataBound;
    /**
     * Row bind event handler.
     * Trigger on each row data bind.
     * @type {?}
     */
    DataTableComponent.prototype.rowBind;
    /**
     * Column bind event handler.
     * Triggered after column data bind.
     * @type {?}
     */
    DataTableComponent.prototype.columnBind;
    /**
     * Cell bind event handler.
     * Triggered after data table cell data bind.
     * @type {?}
     */
    DataTableComponent.prototype.cellBind;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.dragAndDropService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.dataTableStateService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.globalRefService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.eventStateService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.dataTableResourceService;
    /**
     * @type {?}
     * @private
     */
    DataTableComponent.prototype.zone;
    /** @type {?} */
    DataTableComponent.prototype.dataStateService;
    /** @type {?} */
    DataTableComponent.prototype.scrollPositionService;
    /** @type {?} */
    DataTableComponent.prototype.config;
}
