// ornamentum/esm5/data-table/components/data-table-column-filter-header/data-table-column-filter-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input } from '@angular/core';
import { Subject } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
import { DataFetchMode } from '../../models/data-fetch-mode.enum';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableEventStateService } from '../../services/data-table-event.service';
/**
 * Data table column filter header component. Apply columns associated data filtering.
 */
var DataTableColumnFilterHeaderComponent = /** @class */ (function () {
    function DataTableColumnFilterHeaderComponent(config, eventStateService) {
        this.config = config;
        this.eventStateService = eventStateService;
        this.columnFilterStream = new Subject();
        this.customFilterStream = new EventEmitter();
    }
    /**
     * Component initialize lifecycle event handler.
     */
    /**
     * Component initialize lifecycle event handler.
     * @return {?}
     */
    DataTableColumnFilterHeaderComponent.prototype.ngOnInit = /**
     * Component initialize lifecycle event handler.
     * @return {?}
     */
    function () {
        this.initCustomFilterEvent();
        this.initDebounceDefaultFilterEvent();
    };
    /**
     * Component destroy lifecycle event handler.
     */
    /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    DataTableColumnFilterHeaderComponent.prototype.ngOnDestroy = /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    function () {
        if (this.customFilterSubscription) {
            this.customFilterSubscription.unsubscribe();
        }
        if (this.columnFilterSubscription) {
            this.columnFilterSubscription.unsubscribe();
        }
    };
    /**
     * Initialize custom filter event.
     */
    /**
     * Initialize custom filter event.
     * @private
     * @return {?}
     */
    DataTableColumnFilterHeaderComponent.prototype.initCustomFilterEvent = /**
     * Initialize custom filter event.
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.customFilterSubscription = this.customFilterStream.subscribe((/**
         * @param {?} filterEventArgs
         * @return {?}
         */
        function (filterEventArgs) {
            filterEventArgs.column.filter = filterEventArgs.filter;
            _this.onFilter();
        }));
    };
    /**
     * Initialize debounce default filtering logic.
     */
    /**
     * Initialize debounce default filtering logic.
     * @private
     * @return {?}
     */
    DataTableColumnFilterHeaderComponent.prototype.initDebounceDefaultFilterEvent = /**
     * Initialize debounce default filtering logic.
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.columnFilterSubscription = this.columnFilterStream.pipe(debounceTime(this.config.filterDebounceTime)).subscribe((/**
         * @return {?}
         */
        function () {
            _this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
        }));
    };
    /**
     * Filter event handler.
     */
    /**
     * Filter event handler.
     * @return {?}
     */
    DataTableColumnFilterHeaderComponent.prototype.onFilter = /**
     * Filter event handler.
     * @return {?}
     */
    function () {
        if (this.config.filterDebounce) {
            this.columnFilterStream.next();
        }
        else {
            this.eventStateService.dataFetchStream.next(DataFetchMode.SOFT_LOAD);
        }
    };
    DataTableColumnFilterHeaderComponent.decorators = [
        { type: Component, args: [{
                    exportAs: 'ngDataTableColumnFilterHeader',
                    // tslint:disable-next-line
                    selector: '[ngDataTableColumnFilterHeader]',
                    template: "<th *ngIf=\"config.expandableRows\" class=\"ng-data-table-expand-column-header\"></th>\n<th *ngIf=\"config.showIndexColumn\" class=\"ng-data-table-index-column-header\"></th>\n<th *ngIf=\"config.showRowSelectCheckboxColumn\" class=\"ng-data-table-select-column-header\"></th>\n<ng-container *ngFor=\"let column of columns; index as i;\">\n  <th *ngIf=\"column.visible\">\n    <ng-data-table-column-filter-template [column]=\"column\"\n                                          [customFilterStream]=\"customFilterStream\"\n                                          [index]=\"i\"\n                                          (filter)=\"onFilter()\">\n    </ng-data-table-column-filter-template>\n  </th>\n</ng-container>\n"
                }] }
    ];
    /** @nocollapse */
    DataTableColumnFilterHeaderComponent.ctorParameters = function () { return [
        { type: DataTableConfigService },
        { type: DataTableEventStateService }
    ]; };
    DataTableColumnFilterHeaderComponent.propDecorators = {
        columns: [{ type: Input }]
    };
    return DataTableColumnFilterHeaderComponent;
}());
export { DataTableColumnFilterHeaderComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTableColumnFilterHeaderComponent.prototype.columnFilterStream;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnFilterHeaderComponent.prototype.customFilterSubscription;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnFilterHeaderComponent.prototype.columnFilterSubscription;
    /** @type {?} */
    DataTableColumnFilterHeaderComponent.prototype.customFilterStream;
    /** @type {?} */
    DataTableColumnFilterHeaderComponent.prototype.columns;
    /** @type {?} */
    DataTableColumnFilterHeaderComponent.prototype.config;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnFilterHeaderComponent.prototype.eventStateService;
}
