// ornamentum/esm2015/data-table/components/data-table-column/data-table-column.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, Input, TemplateRef } from '@angular/core';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableEventStateService } from '../../services/data-table-event.service';
import { DataTableDataStateService } from '../../services/data-table-data-state.service';
/**
 * Data table column component. Data table columns associated data is captured via this component.
 */
export class DataTableColumnComponent {
    /**
     * @param {?} dataTableConfigService
     * @param {?} eventStateService
     * @param {?} dataStateService
     */
    constructor(dataTableConfigService, eventStateService, dataStateService) {
        this.dataTableConfigService = dataTableConfigService;
        this.eventStateService = eventStateService;
        this.dataStateService = dataStateService;
        this.currentSortOrder = '';
        this.baseSortOrder = '';
        /**
         * Show filed in column selector popup if true.
         */
        this.showInColumnSelector = true; // TODO: move to base conf
        // Table column config
        this.sortable = dataTableConfigService.sortable;
        this.currentSortOrder = dataTableConfigService.sortOrder;
        this.filterable = dataTableConfigService.filterable;
        this.filterPlaceholder = dataTableConfigService.filterPlaceholder;
        this.resizable = dataTableConfigService.columnResizable;
        this.visible = dataTableConfigService.columnVisible;
        this.showDropdownFilter = dataTableConfigService.showDropdownFilter;
        this.showFilterClearButton = dataTableConfigService.showFilterClearButton;
        // Dropdown filter config
        this.dropdownFilterMenuPosition = dataTableConfigService.dropdownFilterMenuPosition;
        this.dropdownFilterSelectMode = dataTableConfigService.dropdownFilterSelectMode;
        this.dropdownFilterSearchable = dataTableConfigService.dropdownFilterSearchable;
        this.dropdownFilterSearchDebounceTime = dataTableConfigService.dropdownFilterSearchDebounceTime;
        this.dropdownFilterSearchDebounce = dataTableConfigService.dropdownFilterSearchDebounce;
        this.dropdownFilterWrapDisplaySelectLimit = dataTableConfigService.dropdownFilterWrapDisplaySelectLimit;
        this.dropdownFilterGroupByField = dataTableConfigService.dropdownFilterGroupByField;
        this.dropdownFilterShowSelectedOptionRemoveButton = dataTableConfigService.dropdownFilterShowSelectedOptionRemoveButton;
        this.dropdownFilterShowClearSelectionButton = dataTableConfigService.dropdownFilterShowClearSelectionButton;
        this.dropdownFilterMenuWidth = dataTableConfigService.dropdownFilterMenuWidth;
        this.dropdownFilterMenuHeight = dataTableConfigService.dropdownFilterMenuHeight;
        this.dropdownFilterMultiSelectOptionMaxWidth = dataTableConfigService.dropdownFilterMultiSelectOptionMaxWidth;
        this.dropdownFilterCloseMenuOnSelect = dataTableConfigService.dropdownFilterCloseMenuOnSelect;
        this.dropdownFilterDynamicDimensionCalculation = dataTableConfigService.dropdownFilterDynamicDimensionCalculation;
        this.dropdownFilterDynamicWidthRatio = dataTableConfigService.dropdownFilterDynamicWidthRatio;
        this.dropdownFilterDynamicHeightRatio = dataTableConfigService.dropdownFilterDynamicHeightRatio;
    }
    /**
     * Set initial column sort order.
     * @param {?} value
     * @return {?}
     */
    set sortOrder(value) {
        this.currentSortOrder = value;
        this.baseSortOrder = value;
    }
    /**
     * Get initial column sort order.
     * @return {?}
     */
    get sortOrder() {
        return this.currentSortOrder;
    }
    /**
     * Reset data sort order.
     * @return {?}
     */
    resetSortOrder() {
        this.currentSortOrder = this.baseSortOrder;
    }
    /**
     * Get dynamic cell color.
     * @param {?} row Data row object.
     * @return {?} Cell color string.
     */
    getCellColor(row) {
        if (this.onCellColorRender !== undefined) {
            return this.onCellColorRender(row, this);
        }
    }
    /**
     * Get new sort order upon sort click.
     * @return {?} New sort order enum value.
     */
    getNewSortOrder() {
        /** @type {?} */
        let newSortOrder;
        switch (this.sortOrder) {
            case 'asc':
                newSortOrder = 'desc';
                break;
            case 'desc':
                newSortOrder = '';
                break;
            case '':
                newSortOrder = 'asc';
                break;
        }
        return newSortOrder;
    }
    /**
     * Get current sort state icon css class enabled state.
     * @return {?} Sort order icon css class collection object.
     */
    getSortIconClass() {
        return {
            'sort-asc': this.sortOrder === 'asc',
            'sort-dsc': this.sortOrder === 'desc',
            'sort-reset': !this.sortOrder
        };
    }
    /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    ngOnDestroy() {
        if (this.filterValueExtractorSubscription) {
            this.filterValueExtractorSubscription.unsubscribe();
        }
    }
    /**
     * Component initialize lifecycle event handler.
     * @return {?}
     */
    ngOnInit() {
        if (!this.cssClass && this.field) {
            if (/^[a-zA-Z0-9_]+$/.test(this.field)) {
                this.cssClass = 'column-' + this.field;
            }
            else {
                this.cssClass = 'column-' + this.field.replace(/[^a-zA-Z0-9_]/g, '');
            }
        }
        this.eventStateService.columnBind.emit(this);
        if (this.dataTableConfigService.multiColumnSortable && this.sortable) {
            if (this.sortOrder === '') {
                if (this.sortPriority !== undefined) {
                    throw Error('[sortPriority] should be ignored when multi column sorting is enabled with natural sort order.');
                }
            }
            else {
                if (this.sortPriority === undefined) {
                    throw Error('[sortPriority] is required when multi column sorting is enabled with an explicit sort order.');
                }
            }
            if (this.sortPriority < 1) {
                throw Error('[sortPriority] must be greater than 1.');
            }
            if (this.dataStateService.currentSortPriority < this.sortPriority) {
                this.dataStateService.currentSortPriority = this.sortPriority;
            }
        }
    }
}
DataTableColumnComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-data-table-column',
                template: ''
            }] }
];
/** @nocollapse */
DataTableColumnComponent.ctorParameters = () => [
    { type: DataTableConfigService },
    { type: DataTableEventStateService },
    { type: DataTableDataStateService }
];
DataTableColumnComponent.propDecorators = {
    cellTemplate: [{ type: ContentChild, args: ['ngDataTableCell', { static: true },] }],
    headerTemplate: [{ type: ContentChild, args: ['ngDataTableHeader', { static: true },] }],
    filterTemplate: [{ type: ContentChild, args: ['ngDataTableFilter', { static: true },] }],
    dropdownFilterLoadingSpinnerTemplate: [{ type: ContentChild, args: ['ngFilterDropdownLoadingSpinner', { static: true },] }],
    dropdownFilterOptionTemplate: [{ type: ContentChild, args: ['ngFilterDropdownOption', { static: true },] }],
    dropdownFilterOptionGroupHeaderTemplate: [{ type: ContentChild, args: ['ngFilterDropdownOptionGroupHeader', { static: true },] }],
    filterExpression: [{ type: Input }],
    filterFieldMapper: [{ type: Input }],
    onCellColorRender: [{ type: Input }],
    title: [{ type: Input }],
    sortable: [{ type: Input }],
    sortPriority: [{ type: Input }],
    sortOrder: [{ type: Input }],
    filterable: [{ type: Input }],
    resizable: [{ type: Input }],
    field: [{ type: Input }],
    filterField: [{ type: Input }],
    sortField: [{ type: Input }],
    cssClass: [{ type: Input }],
    width: [{ type: Input }],
    visible: [{ type: Input }],
    showInColumnSelector: [{ type: Input }],
    filterPlaceholder: [{ type: Input }],
    filter: [{ type: Input }],
    showFilterClearButton: [{ type: Input }],
    resizeMinLimit: [{ type: Input }],
    showDropdownFilter: [{ type: Input }],
    dropdownFilterMenuPosition: [{ type: Input }],
    dropdownFilterSelectMode: [{ type: Input }],
    dropdownFilterSearchable: [{ type: Input }],
    dropdownFilterSearchDebounceTime: [{ type: Input }],
    dropdownFilterSearchDebounce: [{ type: Input }],
    dropDownFilterShowOptionSelectCheckbox: [{ type: Input }],
    dropdownFilterWrapDisplaySelectLimit: [{ type: Input }],
    dropdownFilterGroupByField: [{ type: Input }],
    dropdownFilterShowSelectedOptionRemoveButton: [{ type: Input }],
    dropdownFilterShowClearSelectionButton: [{ type: Input }],
    dropdownFilterMenuWidth: [{ type: Input }],
    dropdownFilterMenuHeight: [{ type: Input }],
    dropdownFilterMultiSelectOptionMaxWidth: [{ type: Input }],
    dropdownFilterCloseMenuOnSelect: [{ type: Input }],
    dropdownFilterDynamicDimensionCalculation: [{ type: Input }],
    dropdownFilterDynamicWidthRatio: [{ type: Input }],
    dropdownFilterDynamicHeightRatio: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTableColumnComponent.prototype.filterValueExtractorSubscription;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnComponent.prototype.currentSortOrder;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnComponent.prototype.baseSortOrder;
    /** @type {?} */
    DataTableColumnComponent.prototype.actualWidth;
    /** @type {?} */
    DataTableColumnComponent.prototype.cellTemplate;
    /** @type {?} */
    DataTableColumnComponent.prototype.headerTemplate;
    /** @type {?} */
    DataTableColumnComponent.prototype.filterTemplate;
    /** @type {?} */
    DataTableColumnComponent.prototype.dropdownFilterLoadingSpinnerTemplate;
    /** @type {?} */
    DataTableColumnComponent.prototype.dropdownFilterOptionTemplate;
    /** @type {?} */
    DataTableColumnComponent.prototype.dropdownFilterOptionGroupHeaderTemplate;
    /**
     * Filter expression event handler callback. Used to apply custom data filter expression logic.
     * @type {?}
     */
    DataTableColumnComponent.prototype.filterExpression;
    /**
     * Custom filter field map event handler callback. Used to extract filter field when showDropdownFilter option is true.
     * @type {?}
     */
    DataTableColumnComponent.prototype.filterFieldMapper;
    /**
     * Cell color render event handler callback.
     * @type {?}
     */
    DataTableColumnComponent.prototype.onCellColorRender;
    /**
     * Column display title.
     * @type {?}
     */
    DataTableColumnComponent.prototype.title;
    /**
     * Columns sortable if true. Show sort indicator on column title.
     * @type {?}
     */
    DataTableColumnComponent.prototype.sortable;
    /**
     * Multi column sort priority. Lowest number will get the height precedence. Usage of same precedence number in
     * multiple columns may lead to unexpected behaviors. This priority number will be displayed in the column header
     * when multi column sorting is enabled hence, consider indexing accordingly.
     * @type {?}
     */
    DataTableColumnComponent.prototype.sortPriority;
    /**
     * Column filterable if true. Show filter options on filter header row when enabled.
     * @type {?}
     */
    DataTableColumnComponent.prototype.filterable;
    /**
     * Column resizeable if true. Show column resize indicator on column right corner.
     * @type {?}
     */
    DataTableColumnComponent.prototype.resizable;
    /**
     * Data item mapping field name.
     * @type {?}
     */
    DataTableColumnComponent.prototype.field;
    /**
     * Filter field identifier. Fallback to field if not provided.
     * @type {?}
     */
    DataTableColumnComponent.prototype.filterField;
    /**
     * Sort field identifier. Fallback to field if not provided.
     * @type {?}
     */
    DataTableColumnComponent.prototype.sortField;
    /**
     * Column title CSS class names. Use space delimiter to separate class names.
     * @type {?}
     */
    DataTableColumnComponent.prototype.cssClass;
    /**
     * Static column width in pixels or percentage.
     * @type {?}
     */
    DataTableColumnComponent.prototype.width;
    /**
     * Render column if true. Else include in column selector but not rendered.
     * @type {?}
     */
    DataTableColumnComponent.prototype.visible;
    /**
     * Show filed in column selector popup if true.
     * @type {?}
     */
    DataTableColumnComponent.prototype.showInColumnSelector;
    /**
     * Filter placeholder value. Placeholder text to show on filter text box. Applicable only for none dropdown filter mode.
     * @type {?}
     */
    DataTableColumnComponent.prototype.filterPlaceholder;
    /**
     * Applied filter value on initialize.
     * @type {?}
     */
    DataTableColumnComponent.prototype.filter;
    /**
     * Show filter clear button if true. Applicable only for none dropdown filter mode.
     * @type {?}
     */
    DataTableColumnComponent.prototype.showFilterClearButton;
    /**
     * Resize minimum limit. Column cannot be resized to fit less than the number of pixels specified here.
     * @type {?}
     */
    DataTableColumnComponent.prototype.resizeMinLimit;
    /**
     * Show dropdown filter if true. Filter data using dropdown filter.
     * @type {?}
     */
    DataTableColumnComponent.prototype.showDropdownFilter;
    /**
     * Dropdown filter menu position. Placement of filter popup menu.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterMenuPosition;
    /**
     * Dropdown select mode. Filter option select mode.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterSelectMode;
    /**
     * Dropdown filter searchable if true. Display search box within filter option menu.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterSearchable;
    /**
     * Dropdown filter search debounce time in milliseconds. Applicable only when dropdownFilterSearchDebounce is true.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterSearchDebounceTime;
    /**
     * Enable dropdown filter data search debounce with provided dropdownFilterSearchDebounceTime if true.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterSearchDebounce;
    /**
     * Dropdown filter show option select checkbox.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropDownFilterShowOptionSelectCheckbox;
    /**
     * Dropdown filter selected items display limit.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterWrapDisplaySelectLimit;
    /**
     * Dropdown filter group by field name in item schema.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterGroupByField;
    /**
     * Dropdown filter show selected option remove button if true.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterShowSelectedOptionRemoveButton;
    /**
     * Dropdown filter show all select options clear button if true.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterShowClearSelectionButton;
    /**
     * Dropdown filter menu width in pixels.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterMenuWidth;
    /**
     * Dropdown filter menu height in pixels.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterMenuHeight;
    /**
     * Dropdown filter multi select option max width.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterMultiSelectOptionMaxWidth;
    /**
     * Dropdown filter close menu on select if true.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterCloseMenuOnSelect;
    /**
     * Dynamically calculate Dropdown filter menu dimensions relative to column width; dropdownFilterMenuWidth and
     * dropdownFilterMenuHeight configuration are ignored when true.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterDynamicDimensionCalculation;
    /**
     * Dynamic dropdown view width ratio. Used for dynamic dimension calculation.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterDynamicWidthRatio;
    /**
     * Dynamic dropdown view height ratio. Used for dynamic dimension calculation.
     * @type {?}
     */
    DataTableColumnComponent.prototype.dropdownFilterDynamicHeightRatio;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnComponent.prototype.dataTableConfigService;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnComponent.prototype.eventStateService;
    /**
     * @type {?}
     * @private
     */
    DataTableColumnComponent.prototype.dataStateService;
}
