// ornamentum/esm5/dropdown-resource/dropdown-resource.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ResourceUtilityModule } from '../resource-utility/resource-utility.module';
import { DropdownResourceConfigService } from './services/dropdown-resource-config.service';
import { DropdownHttpResourceFactoryService } from './services/dropdown-http-resource-factory.service';
import { DropdownWebsocketResourceFactoryService } from './services/dropdown-websocket-resource-factory.service';
/** @type {?} */
var PROVIDERS = [DropdownHttpResourceFactoryService, DropdownWebsocketResourceFactoryService];
/**
 * Dropdown data source module.
 */
var DropdownResourceModule = /** @class */ (function () {
    function DropdownResourceModule() {
    }
    /**
     * Set module root configuration overrides.
     * @return Module with custom providers.
     */
    /**
     * Set module root configuration overrides.
     * @return {?} Module with custom providers.
     */
    DropdownResourceModule.forRoot = /**
     * Set module root configuration overrides.
     * @return {?} Module with custom providers.
     */
    function () {
        return {
            ngModule: DropdownResourceModule,
            providers: [DropdownResourceConfigService]
        };
    };
    DropdownResourceModule.decorators = [
        { type: NgModule, args: [{
                    imports: [ResourceUtilityModule],
                    providers: tslib_1.__spread(PROVIDERS)
                },] }
    ];
    return DropdownResourceModule;
}());
export { DropdownResourceModule };
export { DropdownHttpDataFetchService } from './services/dropdown-http-data-fetch.service';
export { DropdownHttpResourceFactoryService };
export { DropdownWebsocketDataFetchService } from './services/dropdown-websocket-data-fetch.service';
export { DropdownWebsocketResourceFactoryService };
