// ornamentum/esm5/data-table/directives/scroll-position.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, NgZone, Input } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { map, pairwise } from 'rxjs/operators';
/**
 * Scroll position directive.
 * Track current scroll position of target element.
 */
var ScrollPositionDirective = /** @class */ (function () {
    function ScrollPositionDirective(el, zone) {
        this.el = el;
        this.zone = zone;
    }
    /**
     * After component initialize lifecycle event handler.
     */
    /**
     * After component initialize lifecycle event handler.
     * @return {?}
     */
    ScrollPositionDirective.prototype.ngAfterViewInit = /**
     * After component initialize lifecycle event handler.
     * @return {?}
     */
    function () {
        var _this = this;
        this.zone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            _this.scrollEventSubscription = fromEvent(_this.el.nativeElement, 'scroll')
                .pipe(map((/**
             * @return {?}
             */
            function () {
                return {
                    scrollLeft: _this.el.nativeElement.scrollLeft,
                    scrollTop: _this.el.nativeElement.scrollTop,
                    scrollHeight: _this.el.nativeElement.scrollHeight,
                    scrollWidth: _this.el.nativeElement.scrollWidth,
                    clientHeight: _this.el.nativeElement.clientHeight,
                    clientWidth: _this.el.nativeElement.clientWidth,
                };
            })), pairwise(), map((/**
             * @param {?} pair
             * @return {?}
             */
            function (pair) {
                var _a = tslib_1.__read(pair, 2), previous = _a[0], current = _a[1];
                return tslib_1.__assign({}, current, { isHorizontal: previous.scrollLeft !== current.scrollLeft, isVertical: previous.scrollTop !== current.scrollTop });
            })))
                .subscribe((/**
             * @param {?} pos
             * @return {?}
             */
            function (pos) {
                _this.ngScrollPosition.next(pos);
            }));
        }));
    };
    /**
     * Component destroy lifecycle event handler.
     */
    /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    ScrollPositionDirective.prototype.ngOnDestroy = /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    function () {
        if (this.scrollEventSubscription) {
            this.scrollEventSubscription.unsubscribe();
        }
        this.ngScrollPosition.complete();
    };
    ScrollPositionDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngScrollPosition]'
                },] }
    ];
    /** @nocollapse */
    ScrollPositionDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgZone }
    ]; };
    ScrollPositionDirective.propDecorators = {
        ngScrollPosition: [{ type: Input }]
    };
    return ScrollPositionDirective;
}());
export { ScrollPositionDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScrollPositionDirective.prototype.scrollEventSubscription;
    /** @type {?} */
    ScrollPositionDirective.prototype.ngScrollPosition;
    /**
     * @type {?}
     * @private
     */
    ScrollPositionDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    ScrollPositionDirective.prototype.zone;
}
