// ornamentum/esm5/data-table/directives/scroll-element.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * Scroll element directive.
 * Used to set scroll position to target element.
 */
var ScrollElementDirective = /** @class */ (function () {
    function ScrollElementDirective(el) {
        this.el = el;
    }
    /**
     * After component initialize lifecycle event handler.
     */
    /**
     * After component initialize lifecycle event handler.
     * @return {?}
     */
    ScrollElementDirective.prototype.ngAfterViewInit = /**
     * After component initialize lifecycle event handler.
     * @return {?}
     */
    function () {
        var _this = this;
        this.scrollPositionSubscription = this.ngScrollElement.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.el.nativeElement.scrollLeft = value.scrollLeft;
        }));
    };
    /**
     * Component destroy lifecycle event handler.
     */
    /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    ScrollElementDirective.prototype.ngOnDestroy = /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    function () {
        if (this.scrollPositionSubscription) {
            this.scrollPositionSubscription.unsubscribe();
        }
    };
    ScrollElementDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngScrollElement]'
                },] }
    ];
    /** @nocollapse */
    ScrollElementDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    ScrollElementDirective.propDecorators = {
        ngScrollElement: [{ type: Input }]
    };
    return ScrollElementDirective;
}());
export { ScrollElementDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScrollElementDirective.prototype.scrollPositionSubscription;
    /** @type {?} */
    ScrollElementDirective.prototype.ngScrollElement;
    /**
     * @type {?}
     * @private
     */
    ScrollElementDirective.prototype.el;
}
