// ornamentum/esm5/dropdown/components/dropdown/dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, EventEmitter, forwardRef, Injector, Input, Output, Renderer2, TemplateRef, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { Observable, of } from 'rxjs';
import { catchError, debounceTime, switchMap } from 'rxjs/operators';
import { get } from '../../../utility/services/object-utility.class';
import { DropdownViewComponent } from '../dropdown-view/dropdown-view.component';
import { PopoverComponentLoaderFactoryService } from '../../../utility/utility.module';
import { DropdownConfigService } from '../../services/dropdown-config.service';
import { DropdownDataStateService } from '../../services/dropdown-data-state.service';
import { DropdownEventStateService } from '../../services/dropdown-event-state.service';
import { DropdownResourceService } from '../../services/dropdown-resource.service';
import { ValidatorService } from '../../../utility/services/validator.service';
/**
 * Dropdown main component.
 */
var DropdownComponent = /** @class */ (function () {
    function DropdownComponent(componentLoaderFactory, injector, eventStateService, dropdownResourceService, renderer, dataStateService, config) {
        this.componentLoaderFactory = componentLoaderFactory;
        this.injector = injector;
        this.eventStateService = eventStateService;
        this.dropdownResourceService = dropdownResourceService;
        this.renderer = renderer;
        this.dataStateService = dataStateService;
        this.config = config;
        this.dataStateService.componentLoaderRef = this.componentLoaderFactory.createLoader(this.renderer);
        this.dataBound = this.eventStateService.dataBoundStream;
        this.selectChange = this.eventStateService.selectChangeStream;
        this.init = this.eventStateService.initStream;
    }
    Object.defineProperty(DropdownComponent.prototype, "optionTemplate", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.dropdownOptionTemplate = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "optionGroupHeaderTemplate", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.dropdownOptionGroupHeaderTemplate = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "onDataBind", {
        // Input - Event handlers
        /**
         * Set data bind callback. This handler is fired on each data fetch request.
         */
        set: 
        // Input - Event handlers
        /**
         * Set data bind callback. This handler is fired on each data fetch request.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.onDataBind = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "loadingSpinnerTemplateRef", {
        // Inputs
        /**
         * Set dropdown loading spinner template reference object. Used by data table component to explicitly pass the template reference.
         */
        set: 
        // Inputs
        /**
         * Set dropdown loading spinner template reference object. Used by data table component to explicitly pass the template reference.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.loadingSpinnerTemplate = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "optionTemplateRef", {
        /**
         * Set dropdown option template reference. Used by data table component to explicitly pass the template reference.
         */
        set: /**
         * Set dropdown option template reference. Used by data table component to explicitly pass the template reference.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.optionTemplate = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "optionGroupHeaderTemplateRef", {
        /**
         * Set dropdown options group header template reference. Used by data table component to explicitly pass the template reference.
         */
        set: /**
         * Set dropdown options group header template reference. Used by data table component to explicitly pass the template reference.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.optionGroupHeaderTemplate = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "options", {
        /**
         * Set static dropdown options collection. No need to set data source when static options collection is provided.
         */
        set: /**
         * Set static dropdown options collection. No need to set data source when static options collection is provided.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!value) {
                return;
            }
            this.eventStateService.staticDataSourceStream.next(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "dataSource", {
        /**
         * Set data source observable. This observable is used to retrieve dropdown options for data binding.
         */
        set: /**
         * Set data source observable. This observable is used to retrieve dropdown options for data binding.
         * @param {?} source
         * @return {?}
         */
        function (source) {
            this.initDataSource(source);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "id", {
        /**
         * Set dropdown unique identifier.
         */
        set: /**
         * Set dropdown unique identifier.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!ValidatorService.idPatternValidatorExpression.test(value)) {
                throw Error('Invalid [id] input value. Unique identifier parameter only accept string begin with a letter ([A-Za-z]) and may be followed by any number of letters, digits ([0-9]), hyphens ("-"), underscores ("_").');
            }
            this.dataStateService.id = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "translations", {
        /**
         * Set translation data object. Used to localize table static label text.
         */
        set: /**
         * Set translation data object. Used to localize table static label text.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.translations = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "selectTrackBy", {
        /**
         * Set select option track by field path which is used to uniquely identify options for selection tracking.
         * This field support object paths expressions 'root[0].nest'.
         */
        set: /**
         * Set select option track by field path which is used to uniquely identify options for selection tracking.
         * This field support object paths expressions 'root[0].nest'.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.selectTrackBy = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "displayTrackBy", {
        /**
         * Set display value track by field path which is used to extract dropdown option display value.
         * This field support object paths expressions 'root[0].nest'.
         */
        set: /**
         * Set display value track by field path which is used to extract dropdown option display value.
         * This field support object paths expressions 'root[0].nest'.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.displayTrackBy = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "groupByField", {
        /**
         * Set options group field path which is used to group the dropdown options.
         * This field support object paths expressions 'root[0].nest'.
         */
        set: /**
         * Set options group field path which is used to group the dropdown options.
         * This field support object paths expressions 'root[0].nest'.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.groupByField = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "disabledTrackBy", {
        /**
         * Set dropdown option disable state field path which is used to disabled state dropdown options.
         * This field support object paths expressions 'root[0].nest'.
         */
        set: /**
         * Set dropdown option disable state field path which is used to disabled state dropdown options.
         * This field support object paths expressions 'root[0].nest'.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.disabledTrackBy = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "selectedOptions", {
        /**
         * Set selected options collection. These options will be set selected by default on initial load.
         * Applicable only when multi select mode is enabled.
         */
        set: /**
         * Set selected options collection. These options will be set selected by default on initial load.
         * Applicable only when multi select mode is enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.selectedOptions = value || [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "selectedOption", {
        /**
         * Set selected option. This option is selected by default on initial load.
         * Applicable only when single select mode is enabled.
         */
        set: /**
         * Set selected option. This option is selected by default on initial load.
         * Applicable only when single select mode is enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.selectedOption = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "limit", {
        /**
         * Set number of options to fetch on scroll to bottom action when load on scroll mode is enabled.
         */
        set: /**
         * Set number of options to fetch on scroll to bottom action when load on scroll mode is enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.limit = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "wrapDisplaySelectLimit", {
        /**
         * Set wrap selected options in dropdown view and show the number of options selected instead when
         * limit is met or exceeded. Applicable only when multi select mode is enabled.
         */
        set: /**
         * Set wrap selected options in dropdown view and show the number of options selected instead when
         * limit is met or exceeded. Applicable only when multi select mode is enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.wrapDisplaySelectLimit = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "loadOnScroll", {
        /**
         * Set infinite scrollable state to load data on demand with scroll motion. Dropdown data fetch call is
         * initiated with limit and offset when user scroll to bottom hence loading the full data set on init.
         */
        set: /**
         * Set infinite scrollable state to load data on demand with scroll motion. Dropdown data fetch call is
         * initiated with limit and offset when user scroll to bottom hence loading the full data set on init.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.loadOnScroll = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "loadViewDistanceRatio", {
        /**
         * Set view height ratio to trigger data fetch with infinite scrollable mode.
         * Higher ratio will will increase the scroll sensitivity.
         */
        set: /**
         * Set view height ratio to trigger data fetch with infinite scrollable mode.
         * Higher ratio will will increase the scroll sensitivity.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.loadViewDistanceRatio = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "selectMode", {
        /**
         * Set option select mode.
         * - 'multi' : Support selecting multiple options.
         * - 'single' : Support selecting a single option from options collection.
         * - 'single-toggle' : Support selecting a single option from options collection. Selection can not be removed but
         * only toggled by tapping on another option.
         */
        set: /**
         * Set option select mode.
         * - 'multi' : Support selecting multiple options.
         * - 'single' : Support selecting a single option from options collection.
         * - 'single-toggle' : Support selecting a single option from options collection. Selection can not be removed but
         * only toggled by tapping on another option.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.selectMode = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "filterable", {
        /**
         * Show dropdown option search filter text-box if true.
         */
        set: /**
         * Show dropdown option search filter text-box if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.filterable = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "filterText", {
        /**
         * Set default filter value to be applied on initial load. All options are displayed when filter text value is
         * empty string. Applicable only when dropdown is filterable.
         */
        set: /**
         * Set default filter value to be applied on initial load. All options are displayed when filter text value is
         * empty string. Applicable only when dropdown is filterable.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.filterText = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "filterDebounce", {
        /**
         * Set time based filter debounce to optimize performance and avoid request flooding by reducing the filter
         * request frequency if true. Applicable only when dropdown filterable state is enabled.
         */
        set: /**
         * Set time based filter debounce to optimize performance and avoid request flooding by reducing the filter
         * request frequency if true. Applicable only when dropdown filterable state is enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.filterDebounce = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "filterDebounceTime", {
        /**
         * Set filter debounce time in milliseconds. Applicable only when searchDebounce is true.
         */
        set: /**
         * Set filter debounce time in milliseconds. Applicable only when searchDebounce is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.filterDebounceTime = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "loadDataOnInit", {
        /**
         * Set load data on component initialize if true.
         */
        set: /**
         * Set load data on component initialize if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.loadDataOnInit = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "showSelectedOptionRemoveButton", {
        /**
         * Show selected option remove button if true.
         * Applicable only when multi select mode ios enabled.
         */
        set: /**
         * Show selected option remove button if true.
         * Applicable only when multi select mode ios enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showSelectedOptionRemoveButton = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "showClearSelectionButton", {
        /**
         * Set show all select options clear button if true.
         * Applicable only when multi select mode ios enabled.
         */
        set: /**
         * Set show all select options clear button if true.
         * Applicable only when multi select mode ios enabled.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showClearSelectionButton = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "menuWidth", {
        /**
         * Set options menu width in pixels.
         */
        set: /**
         * Set options menu width in pixels.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.menuWidth = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "menuHeight", {
        /**
         * Set options menu height in pixels.
         */
        set: /**
         * Set options menu height in pixels.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.menuHeight = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "menuPosition", {
        /**
         * Set popup options menu display position relative to dropdown component.
         * Support 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right' values.
         */
        set: /**
         * Set popup options menu display position relative to dropdown component.
         * Support 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right' values.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.menuPosition = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "disabled", {
        /**
         * Set dropdown view disabled state.
         */
        set: /**
         * Set dropdown view disabled state.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.dataStateService.disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "closeMenuOnSelect", {
        /**
         * Set Close dropdown menu on option select if true.
         */
        set: /**
         * Set Close dropdown menu on option select if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.closeMenuOnSelect = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "showOptionSelectCheckbox", {
        /**
         * Set show dropdown option select checkbox if true.
         */
        set: /**
         * Set show dropdown option select checkbox if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showOptionSelectCheckbox = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "showOptionIndex", {
        /**
         * Set show dropdown option index checkbox if true.
         */
        set: /**
         * Set show dropdown option index checkbox if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showOptionIndex = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "showOptionTrackBy", {
        /**
         * Set show dropdown option TrackBy id checkbox if true.
         */
        set: /**
         * Set show dropdown option TrackBy id checkbox if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.showOptionTrackBy = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "multiSelectOptionMaxWidth", {
        /**
         * Multi select option selected item maximum width. Apply ellipsis when selected option display text
         * exceed the max width.
         */
        set: /**
         * Multi select option selected item maximum width. Apply ellipsis when selected option display text
         * exceed the max width.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.multiSelectOptionMaxWidth = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "setFirstOptionSelected", {
        /**
         * Set first dropdown option selected on data fetch if true.
         */
        set: /**
         * Set first dropdown option selected on data fetch if true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.setFirstOptionSelected = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "triggerSelectChangeOnInit", {
        /**
         * Trigger select change event on init if true.
         * Can be used to enable selectedOptions or selectedOption associated change trigger.
         */
        set: /**
         * Trigger select change event on init if true.
         * Can be used to enable selectedOptions or selectedOption associated change trigger.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.triggerSelectChangeOnInit = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "triggerSelectChangeOnModelUpdate", {
        /**
         * Set trigger select change on explicit model update if true.
         * Applicable only when form binding is used.
         */
        set: /**
         * Set trigger select change on explicit model update if true.
         * Applicable only when form binding is used.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.triggerSelectChangeOnModelUpdate = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "triggerSelectChangeOnFirstOptionSelect", {
        /**
         * Set trigger select change on first option select change if true.
         * Applicable only when setFirstOptionSelected is true.
         */
        set: /**
         * Set trigger select change on first option select change if true.
         * Applicable only when setFirstOptionSelected is true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.triggerSelectChangeOnFirstOptionSelect = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "dynamicDimensionCalculation", {
        /**
         * Set dynamically calculate dropdown view dimensions relative to dropdown button width.
         * MenuWith and menuHeight values are ignored when true.
         */
        set: /**
         * Set dynamically calculate dropdown view dimensions relative to dropdown button width.
         * MenuWith and menuHeight values are ignored when true.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.dynamicDimensionCalculation = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "dynamicWidthRatio", {
        /**
         * Set dynamic dropdown options view dimensions calculation width ratio relative to dropdown selector.
         */
        set: /**
         * Set dynamic dropdown options view dimensions calculation width ratio relative to dropdown selector.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.dynamicWidthRatio = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "dynamicHeightRatio", {
        /**
         * Set dynamic dropdown options view dimensions calculation height ratio relative to dropdown selector.
         */
        set: /**
         * Set dynamic dropdown options view dimensions calculation height ratio relative to dropdown selector.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.dynamicHeightRatio = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "relativeParentElement", {
        /**
         * Set relative parent element to render dropdown view container.
         */
        set: /**
         * Set relative parent element to render dropdown view container.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.config.relativeParentElement = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Initialize data source.
     * @param source Data source stream.
     */
    /**
     * Initialize data source.
     * @private
     * @param {?} source Data source stream.
     * @return {?}
     */
    DropdownComponent.prototype.initDataSource = /**
     * Initialize data source.
     * @private
     * @param {?} source Data source stream.
     * @return {?}
     */
    function (source) {
        var _this = this;
        this.dropdownResourceService.setDataSource(source);
        this.dataStateService.onDataBind = (/**
         * @param {?} params
         * @return {?}
         */
        function (params) {
            if (params.hardReload) {
                _this.dropdownResourceService.setDataSource(source);
            }
            return _this.dropdownResourceService.query(params);
        });
    };
    /**
     * Performs dropdown toggle event.
     * @param event Mouse click event args.
     * @param element Dropdown button element.
     */
    /**
     * Performs dropdown toggle event.
     * @param {?} event Mouse click event args.
     * @param {?} element Dropdown button element.
     * @return {?}
     */
    DropdownComponent.prototype.toggleDropdown = /**
     * Performs dropdown toggle event.
     * @param {?} event Mouse click event args.
     * @param {?} element Dropdown button element.
     * @return {?}
     */
    function (event, element) {
        /** @type {?} */
        var target = (/** @type {?} */ (event.target));
        if (target && target.classList && target.classList.contains('ng-ignore-propagation')) {
            return;
        }
        this.dataStateService.componentLoaderRef.toggle(DropdownViewComponent, element, this.injector, {
            relativeParentElement: this.config.relativeParentElement,
            position: this.config.menuPosition
        });
        if (this.config.dynamicDimensionCalculation) {
            this.config.menuWidth = element.offsetWidth * this.config.dynamicWidthRatio;
            this.config.menuHeight = element.offsetWidth * this.config.dynamicHeightRatio;
        }
    };
    Object.defineProperty(DropdownComponent.prototype, "wrapSelectedOptions", {
        /**
         * Get options wrapped state.
         */
        get: /**
         * Get options wrapped state.
         * @return {?}
         */
        function () {
            if (this.config.wrapDisplaySelectLimit !== undefined) {
                return this.dataStateService.selectedOptions.length > this.config.wrapDisplaySelectLimit;
            }
            return false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "wrappedOptionDisplayText", {
        /**
         * Get wrapped option display text.
         */
        get: /**
         * Get wrapped option display text.
         * @return {?}
         */
        function () {
            return "(" + this.dataStateService.selectedOptions.length + ") " + this.config.translations.selectedOptionWrapPlaceholder;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Lifecycle hook that is called when component is destroyed.
     */
    /**
     * Lifecycle hook that is called when component is destroyed.
     * @return {?}
     */
    DropdownComponent.prototype.ngOnDestroy = /**
     * Lifecycle hook that is called when component is destroyed.
     * @return {?}
     */
    function () {
        if (this.onSelectChangeSubscription) {
            this.onSelectChangeSubscription.unsubscribe();
        }
        this.dataStateService.componentLoaderRef.dispose();
    };
    Object.defineProperty(DropdownComponent.prototype, "hasSelectedOptions", {
        /**
         * Get selected options availability state.
         */
        get: /**
         * Get selected options availability state.
         * @return {?}
         */
        function () {
            if (this.config.selectMode === 'multi') {
                return !!this.dataStateService.selectedOptions.length;
            }
            return !!this.dataStateService.selectedOption;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Trigger select change.
     */
    /**
     * Trigger select change.
     * @return {?}
     */
    DropdownComponent.prototype.triggerSelectChange = /**
     * Trigger select change.
     * @return {?}
     */
    function () {
        if (this.config.selectMode === 'multi') {
            this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOptions);
        }
        else {
            this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOption);
        }
    };
    /**
     * Clear selected options.
     */
    /**
     * Clear selected options.
     * @return {?}
     */
    DropdownComponent.prototype.clearSelectedOptions = /**
     * Clear selected options.
     * @return {?}
     */
    function () {
        if (this.config.selectMode === 'multi') {
            this.dataStateService.selectedOptions = [];
            this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOptions);
        }
        else {
            this.dataStateService.selectedOption = undefined;
            this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOption);
        }
    };
    /**
     * Set disabled state.
     * ControlValueAccessor implementation.
     * @param isDisabled True if disabled.
     */
    /**
     * Set disabled state.
     * ControlValueAccessor implementation.
     * @param {?} isDisabled True if disabled.
     * @return {?}
     */
    DropdownComponent.prototype.setDisabledState = /**
     * Set disabled state.
     * ControlValueAccessor implementation.
     * @param {?} isDisabled True if disabled.
     * @return {?}
     */
    function (isDisabled) {
        this.dataStateService.disabled = isDisabled;
    };
    /**
     * Called when value selected value gets updated.
     * ControlValueAccessor implementation.
     * @param value Selected value.
     */
    /**
     * Called when value selected value gets updated.
     * ControlValueAccessor implementation.
     * @param {?} value Selected value.
     * @return {?}
     */
    DropdownComponent.prototype.writeValue = /**
     * Called when value selected value gets updated.
     * ControlValueAccessor implementation.
     * @param {?} value Selected value.
     * @return {?}
     */
    function (value) {
        if (this.config.selectMode === 'multi') {
            this.dataStateService.selectedOptions = value || [];
        }
        else {
            this.dataStateService.selectedOption = value;
        }
        if (this.config.triggerSelectChangeOnModelUpdate) {
            this.triggerSelectChange();
        }
    };
    /**
     * Register on change event.
     * ControlValueAccessor implementation.
     * @param onSelectChange On select change callback function.
     */
    /**
     * Register on change event.
     * ControlValueAccessor implementation.
     * @param {?} onSelectChange On select change callback function.
     * @return {?}
     */
    DropdownComponent.prototype.registerOnChange = /**
     * Register on change event.
     * ControlValueAccessor implementation.
     * @param {?} onSelectChange On select change callback function.
     * @return {?}
     */
    function (onSelectChange) {
        this.onSelectChangeSubscription = this.selectChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            onSelectChange(value);
        }));
    };
    /**
     * Register on touched event.
     * ControlValueAccessor implementation.
     * @param fn Function reference.
     */
    /**
     * Register on touched event.
     * ControlValueAccessor implementation.
     * @param {?} fn Function reference.
     * @return {?}
     */
    DropdownComponent.prototype.registerOnTouched = /**
     * Register on touched event.
     * ControlValueAccessor implementation.
     * @param {?} fn Function reference.
     * @return {?}
     */
    function (fn) {
        // TODO: Implement touch event handler
    };
    /**
     * Lifecycle hook that is called when component is initialized.
     */
    /**
     * Lifecycle hook that is called when component is initialized.
     * @return {?}
     */
    DropdownComponent.prototype.ngOnInit = /**
     * Lifecycle hook that is called when component is initialized.
     * @return {?}
     */
    function () {
        if (!this.dataStateService.id) {
            throw Error('Missing required parameter value for [id] input.');
        }
        if (!this.dataStateService.onDataBind) {
            this.dataSource = this.eventStateService.staticDataSourceStream;
        }
        this.initDataFetchEvent();
        if (this.config.loadDataOnInit) {
            this.eventStateService.dataFetchStream.emit(false);
        }
        if (this.config.triggerSelectChangeOnInit) {
            this.triggerSelectChange();
        }
        this.eventStateService.initStream.emit(this);
    };
    /**
     * Map source data object to dropdown option model.
     * @param option Source dropdown option.
     * @param index Current index.
     */
    /**
     * Map source data object to dropdown option model.
     * @private
     * @param {?} option Source dropdown option.
     * @param {?} index Current index.
     * @return {?}
     */
    DropdownComponent.prototype.mapDropdownOption = /**
     * Map source data object to dropdown option model.
     * @private
     * @param {?} option Source dropdown option.
     * @param {?} index Current index.
     * @return {?}
     */
    function (option, index) {
        /** @type {?} */
        var id = get(option, this.config.selectTrackBy);
        return {
            disabled: get(option, this.config.disabledTrackBy),
            id: id,
            index: index + this.dataStateService.offset + 1,
            option: option,
            text: get(option, this.config.displayTrackBy)
        };
    };
    /**
     *
     * Set dropdown options data.
     * @param queryResult Query result object reference.
     */
    /**
     *
     * Set dropdown options data.
     * @private
     * @param {?} queryResult Query result object reference.
     * @return {?}
     */
    DropdownComponent.prototype.setDropdownOptions = /**
     *
     * Set dropdown options data.
     * @private
     * @param {?} queryResult Query result object reference.
     * @return {?}
     */
    function (queryResult) {
        var _this = this;
        if (this.config.groupByField) {
            this.dataStateService.dropdownOptionGroups = queryResult.options.reduce((/**
             * @param {?} accumulator
             * @param {?} option
             * @param {?} index
             * @return {?}
             */
            function (accumulator, option, index) {
                /** @type {?} */
                var groupFieldValue = get(option, _this.config.groupByField);
                /** @type {?} */
                var currentGroup = accumulator.find((/**
                 * @param {?} group
                 * @return {?}
                 */
                function (group) { return group.groupName === groupFieldValue; }));
                if (currentGroup) {
                    currentGroup.options.push(_this.mapDropdownOption(option, index));
                }
                else {
                    accumulator.push({
                        groupName: groupFieldValue,
                        options: [_this.mapDropdownOption(option, index)]
                    });
                }
                return accumulator;
            }), this.config.loadOnScroll && this.dataStateService.offset > 0 ? this.dataStateService.dropdownOptionGroups : []);
        }
        else {
            this.dataStateService.dropdownOptions = queryResult.options.reduce((/**
             * @param {?} accumulator
             * @param {?} option
             * @param {?} index
             * @return {?}
             */
            function (accumulator, option, index) {
                accumulator.push(_this.mapDropdownOption(option, index));
                return accumulator;
            }), this.config.loadOnScroll && this.dataStateService.offset > 0 ? this.dataStateService.dropdownOptions : []);
        }
        if (this.config.setFirstOptionSelected && queryResult.options.length) {
            if (this.config.selectMode === 'multi') {
                this.dataStateService.selectedOptions = [queryResult.options[0]];
            }
            else {
                this.dataStateService.selectedOption = queryResult.options[0];
            }
            if (this.config.triggerSelectChangeOnFirstOptionSelect) {
                this.triggerSelectChange();
            }
        }
        this.dataStateService.totalOptionCount = queryResult.count;
        this.dataStateService.currentOptionCount += queryResult.options.length;
    };
    /**
     * On after data bind event handler.
     * @param queryResult Query result object reference.
     */
    /**
     * On after data bind event handler.
     * @private
     * @param {?} queryResult Query result object reference.
     * @return {?}
     */
    DropdownComponent.prototype.onAfterDataBind = /**
     * On after data bind event handler.
     * @private
     * @param {?} queryResult Query result object reference.
     * @return {?}
     */
    function (queryResult) {
        this.setDropdownOptions(queryResult);
        this.dataStateService.dataLoading = false;
        this.eventStateService.dataBoundStream.emit();
    };
    /**
     * Fetch query results.
     * @param hardReload Hard reload state.
     */
    /**
     * Fetch query results.
     * @private
     * @param {?} hardReload Hard reload state.
     * @return {?}
     */
    DropdownComponent.prototype.fetchQueryResults = /**
     * Fetch query results.
     * @private
     * @param {?} hardReload Hard reload state.
     * @return {?}
     */
    function (hardReload) {
        this.dataStateService.dataLoading = true;
        if (hardReload) {
            this.dataStateService.offset = 0;
            this.dataStateService.filterText = '';
        }
        /** @type {?} */
        var requestParams = {
            hardReload: hardReload
        };
        if (this.config.loadOnScroll) {
            requestParams.limit = this.config.limit;
            requestParams.offset = this.dataStateService.offset;
        }
        if (this.config.filterable) {
            requestParams.filter = {
                key: this.config.displayTrackBy,
                value: this.dataStateService.filterText
            };
        }
        return this.dataStateService.onDataBind(requestParams);
    };
    /**
     * Initialize data fetch event.
     */
    /**
     * Initialize data fetch event.
     * @private
     * @return {?}
     */
    DropdownComponent.prototype.initDataFetchEvent = /**
     * Initialize data fetch event.
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var noop = {
            options: [],
            count: 0
        };
        this.eventStateService.dataFetchStream
            .pipe(debounceTime(20), switchMap((/**
         * @param {?} hardReload
         * @return {?}
         */
        function (hardReload) { return _this.fetchQueryResults(hardReload); })), catchError((/**
         * @return {?}
         */
        function () {
            return of(noop);
        })))
            .subscribe((/**
         * @param {?} queryResult
         * @return {?}
         */
        function (queryResult) {
            _this.onAfterDataBind(queryResult);
        }), (/**
         * @return {?}
         */
        function () {
            _this.onAfterDataBind(noop);
        }));
    };
    /**
     * Trigger explicit data fetch.
     * @param hardReload Hard reload state.
     */
    /**
     * Trigger explicit data fetch.
     * @param {?=} hardReload Hard reload state.
     * @return {?}
     */
    DropdownComponent.prototype.fetchData = /**
     * Trigger explicit data fetch.
     * @param {?=} hardReload Hard reload state.
     * @return {?}
     */
    function (hardReload) {
        if (hardReload === void 0) { hardReload = false; }
        this.eventStateService.dataFetchStream.emit(hardReload);
    };
    /**
     * On select option remove event handler.
     * @param index Selected option index.
     */
    /**
     * On select option remove event handler.
     * @param {?} index Selected option index.
     * @return {?}
     */
    DropdownComponent.prototype.onSelectOptionRemove = /**
     * On select option remove event handler.
     * @param {?} index Selected option index.
     * @return {?}
     */
    function (index) {
        this.dataStateService.selectedOptions.splice(index, 1);
        this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOptions);
    };
    /**
     * Close dropdown options menu.
     */
    /**
     * Close dropdown options menu.
     * @return {?}
     */
    DropdownComponent.prototype.close = /**
     * Close dropdown options menu.
     * @return {?}
     */
    function () {
        this.dataStateService.componentLoaderRef.hide();
    };
    DropdownComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-dropdown',
                    template: "<div #dropdownElement class=\"ng-dropdown\">\n  <div class=\"ng-dropdown-toggle\" (click)=\"toggleDropdown($event, dropdownElement)\" [class.disabled]=\"dataStateService.disabled\">\n    <ng-container *ngIf=\"config.selectMode === 'multi'\">\n      <ng-container *ngIf=\"!wrapSelectedOptions\">\n        <span class=\"ng-dropdown-selected-value\"\n          *ngFor=\"let option of dataStateService.selectedOptions; let index = index\"\n          [style.max-width]=\"config.multiSelectOptionMaxWidth | ngPx\"\n          [class.ng-dropdown-option-close-enabled]=\"config.showSelectedOptionRemoveButton\">\n          {{ config.getDisplayText(option) }}\n          <i class=\"ng-dropdown-remove-selected-value ng-ignore-propagation\"\n            *ngIf=\"config.showSelectedOptionRemoveButton\"\n            (click)=\"onSelectOptionRemove(index)\"></i>\n        </span>\n      </ng-container>\n      <span class=\"ng-dropdown-selected-option\" *ngIf=\"wrapSelectedOptions\">\n        {{ wrappedOptionDisplayText }}\n      </span>\n      <span class=\"ng-dropdown-selected-placeholder\" *ngIf=\"!dataStateService.selectedOptions.length\">\n        {{ config.translations.selectPlaceholder }}\n      </span>\n    </ng-container>\n    <ng-container *ngIf=\"config.selectMode !== 'multi'\">\n      <span class=\"ng-dropdown-selected-option\" *ngIf=\"dataStateService.selectedOption\">\n        {{ config.getDisplayText(dataStateService.selectedOption) }}\n      </span>\n      <span class=\"ng-dropdown-selected-placeholder\" *ngIf=\"!dataStateService.selectedOption\">\n        {{ config.translations.selectPlaceholder }}\n      </span>\n    </ng-container>\n    <span class=\"ng-dropdown-input-group-btn\" [hidden]=\"!hasSelectedOptions\" *ngIf=\"config.showClearSelectionButton\">\n      <div class=\"ng-dropdown-delete-button ng-ignore-propagation\" (click)=\"clearSelectedOptions()\"></div>\n    </span>\n    <span class=\"ng-dropdown-down-arrow\" *ngIf=\"!dataStateService.dataLoading\"></span>\n    <span class=\"ng-dropdown-loading\" [hidden]=\"!dataStateService.dataLoading\">\n      <div class=\"ng-dropdown-loading-animation\" *ngIf=\"!loadingSpinnerTemplate\"></div>\n      <div *ngIf=\"loadingSpinnerTemplate\" [ngTemplateOutlet]=\"loadingSpinnerTemplate\"></div>\n    </span>\n  </div>\n</div>\n",
                    providers: [
                        DropdownConfigService,
                        DropdownDataStateService,
                        DropdownEventStateService,
                        DropdownResourceService,
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return DropdownComponent; })),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    DropdownComponent.ctorParameters = function () { return [
        { type: PopoverComponentLoaderFactoryService },
        { type: Injector },
        { type: DropdownEventStateService },
        { type: DropdownResourceService },
        { type: Renderer2 },
        { type: DropdownDataStateService },
        { type: DropdownConfigService }
    ]; };
    DropdownComponent.propDecorators = {
        loadingSpinnerTemplate: [{ type: ContentChild, args: ['ngDropdownLoadingSpinner', { static: true },] }],
        optionTemplate: [{ type: ContentChild, args: ['ngDropdownOption', { static: true },] }],
        optionGroupHeaderTemplate: [{ type: ContentChild, args: ['ngDropdownOptionGroupHeader', { static: true },] }],
        init: [{ type: Output }],
        selectChange: [{ type: Output }],
        dataBound: [{ type: Output }],
        onDataBind: [{ type: Input }],
        loadingSpinnerTemplateRef: [{ type: Input }],
        optionTemplateRef: [{ type: Input }],
        optionGroupHeaderTemplateRef: [{ type: Input }],
        options: [{ type: Input }],
        dataSource: [{ type: Input }],
        id: [{ type: Input }],
        translations: [{ type: Input }],
        selectTrackBy: [{ type: Input }],
        displayTrackBy: [{ type: Input }],
        groupByField: [{ type: Input }],
        disabledTrackBy: [{ type: Input }],
        selectedOptions: [{ type: Input }],
        selectedOption: [{ type: Input }],
        limit: [{ type: Input }],
        wrapDisplaySelectLimit: [{ type: Input }],
        loadOnScroll: [{ type: Input }],
        loadViewDistanceRatio: [{ type: Input }],
        selectMode: [{ type: Input }],
        filterable: [{ type: Input }],
        filterText: [{ type: Input }],
        filterDebounce: [{ type: Input }],
        filterDebounceTime: [{ type: Input }],
        loadDataOnInit: [{ type: Input }],
        showSelectedOptionRemoveButton: [{ type: Input }],
        showClearSelectionButton: [{ type: Input }],
        menuWidth: [{ type: Input }],
        menuHeight: [{ type: Input }],
        menuPosition: [{ type: Input }],
        disabled: [{ type: Input }],
        closeMenuOnSelect: [{ type: Input }],
        showOptionSelectCheckbox: [{ type: Input }],
        showOptionIndex: [{ type: Input }],
        showOptionTrackBy: [{ type: Input }],
        multiSelectOptionMaxWidth: [{ type: Input }],
        setFirstOptionSelected: [{ type: Input }],
        triggerSelectChangeOnInit: [{ type: Input }],
        triggerSelectChangeOnModelUpdate: [{ type: Input }],
        triggerSelectChangeOnFirstOptionSelect: [{ type: Input }],
        dynamicDimensionCalculation: [{ type: Input }],
        dynamicWidthRatio: [{ type: Input }],
        dynamicHeightRatio: [{ type: Input }],
        relativeParentElement: [{ type: Input }]
    };
    return DropdownComponent;
}());
export { DropdownComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.onSelectChangeSubscription;
    /** @type {?} */
    DropdownComponent.prototype.loadingSpinnerTemplate;
    /**
     * Dropdown initialize event handler
     * @type {?}
     */
    DropdownComponent.prototype.init;
    /**
     * Dropdown option select change event handler
     * @type {?}
     */
    DropdownComponent.prototype.selectChange;
    /**
     * Dropdown data bind event handler
     * @type {?}
     */
    DropdownComponent.prototype.dataBound;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.componentLoaderFactory;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.eventStateService;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.dropdownResourceService;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.renderer;
    /** @type {?} */
    DropdownComponent.prototype.dataStateService;
    /** @type {?} */
    DropdownComponent.prototype.config;
}
