// ornamentum/esm2015/dropdown/dropdown.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { DropdownComponent } from './components/dropdown/dropdown.component';
import { DropdownViewComponent } from './components/dropdown-view/dropdown-view.component';
import { DropdownFilterComponent } from './components/dropdown-filter/dropdown-filter.component';
import { DropdownOptionsComponent } from './components/dropdown-options/dropdown-options.component';
import { DROPDOWN_CONFIG } from './services/dropdown-config.service';
import { UtilityModule } from '../utility/utility.module';
/** @type {?} */
const COMPONENTS = [DropdownComponent, DropdownViewComponent, DropdownFilterComponent, DropdownOptionsComponent];
/**
 * Module representing dropdown component.
 */
export class DropdownModule {
    /**
     * Set module root configuration overrides.
     * @param {?=} dropdownConfig Dropdown configuration object.
     * @return {?} Module with custom providers.
     */
    static forRoot(dropdownConfig) {
        return {
            ngModule: DropdownModule,
            providers: [
                {
                    provide: DROPDOWN_CONFIG,
                    useValue: dropdownConfig
                }
            ]
        };
    }
}
DropdownModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, FormsModule, UtilityModule.forRoot()],
                declarations: [...COMPONENTS],
                exports: [DropdownComponent],
                entryComponents: [DropdownViewComponent]
            },] }
];
export { DropdownComponent } from './components/dropdown/dropdown.component';
