// ornamentum/esm5/dropdown/components/dropdown-view/dropdown-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { Subject } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
import { DropdownConfigService } from '../../services/dropdown-config.service';
import { DropdownDataStateService } from '../../services/dropdown-data-state.service';
import { DropdownEventStateService } from '../../services/dropdown-event-state.service';
/**
 * Dropdown options list view component
 */
var DropdownViewComponent = /** @class */ (function () {
    function DropdownViewComponent(config, dataStateService, eventStateService) {
        this.config = config;
        this.dataStateService = dataStateService;
        this.eventStateService = eventStateService;
        this.scrollEvent = new Subject();
    }
    /**
     * Lifecycle hook that is called when component is initialized.
     */
    /**
     * Lifecycle hook that is called when component is initialized.
     * @return {?}
     */
    DropdownViewComponent.prototype.ngOnInit = /**
     * Lifecycle hook that is called when component is initialized.
     * @return {?}
     */
    function () {
        var _this = this;
        this.scrollEventSubscription = this.scrollEvent.pipe(debounceTime(100)).subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.checkScrollPosition(event);
        }));
    };
    /**
     * Lifecycle hook that is called when component is destroyed.
     */
    /**
     * Lifecycle hook that is called when component is destroyed.
     * @return {?}
     */
    DropdownViewComponent.prototype.ngOnDestroy = /**
     * Lifecycle hook that is called when component is destroyed.
     * @return {?}
     */
    function () {
        if (this.scrollEventSubscription) {
            this.scrollEventSubscription.unsubscribe();
        }
    };
    /**
     * Performs data loading when scrolling when scroll threshold is met. This will be trigger only when load on scroll is enabled.
     * @param event Target event arguments reference.
     */
    /**
     * Performs data loading when scrolling when scroll threshold is met. This will be trigger only when load on scroll is enabled.
     * @param {?} event Target event arguments reference.
     * @return {?}
     */
    DropdownViewComponent.prototype.checkScrollPosition = /**
     * Performs data loading when scrolling when scroll threshold is met. This will be trigger only when load on scroll is enabled.
     * @param {?} event Target event arguments reference.
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var scrollTop = event.target.scrollTop;
        /** @type {?} */
        var scrollHeight = event.target.scrollHeight;
        /** @type {?} */
        var scrollElementHeight = event.target.clientHeight;
        /** @type {?} */
        var roundingPixel = 1;
        /** @type {?} */
        var gutterPixel = 1;
        if (scrollTop >= scrollHeight - (1 + this.config.loadViewDistanceRatio) * scrollElementHeight - roundingPixel - gutterPixel &&
            this.dataStateService.currentOptionCount < this.dataStateService.totalOptionCount &&
            !this.dataStateService.dataLoading) {
            this.dataStateService.dataLoading = true;
            this.dataStateService.offset = this.dataStateService.offset + this.config.limit;
            this.eventStateService.dataFetchStream.emit(false);
        }
    };
    DropdownViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-dropdown-view',
                    template: "<div class=\"ng-dropdown-menu-wrapper\"\n  [class.ng-dropdown-option-list]=\"!this.config.filterable\"\n  [style.width]=\"config.menuWidth | ngPx\"\n  [style.height]=\"config.menuHeight | ngPx\">\n  <div class=\"ng-dropdown-filter-wrapper\" *ngIf=\"config.filterable\">\n    <ng-dropdown-filter></ng-dropdown-filter>\n  </div>\n  <div class=\"ng-dropdown-option-container-wrapper\" (scroll)=\"config.loadOnScroll ? scrollEvent.next($event) : null\">\n    <ng-dropdown-options></ng-dropdown-options>\n  </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    DropdownViewComponent.ctorParameters = function () { return [
        { type: DropdownConfigService },
        { type: DropdownDataStateService },
        { type: DropdownEventStateService }
    ]; };
    return DropdownViewComponent;
}());
export { DropdownViewComponent };
if (false) {
    /** @type {?} */
    DropdownViewComponent.prototype.scrollEvent;
    /**
     * @type {?}
     * @private
     */
    DropdownViewComponent.prototype.scrollEventSubscription;
    /** @type {?} */
    DropdownViewComponent.prototype.config;
    /** @type {?} */
    DropdownViewComponent.prototype.dataStateService;
    /**
     * @type {?}
     * @private
     */
    DropdownViewComponent.prototype.eventStateService;
}
