// ornamentum/esm5/data-table/components/data-table-pagination/data-table-pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewChild } from '@angular/core';
import { debounceTime } from 'rxjs/operators';
import { DataFetchMode } from '../../models/data-fetch-mode.enum';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableDataStateService } from '../../services/data-table-data-state.service';
import { DataTableEventStateService } from '../../services/data-table-event.service';
import { ResizeService } from '../../../utility/services/resize.service';
/**
 * Data table pagination component. Render data table paginator widget.
 */
var DataTablePaginationComponent = /** @class */ (function () {
    function DataTablePaginationComponent(config, dataStateService, eventStateService, resizeService) {
        this.config = config;
        this.dataStateService = dataStateService;
        this.eventStateService = eventStateService;
        this.resizeService = resizeService;
        this.isMobile = false;
    }
    /**
     * Set mobile mode state.
     */
    /**
     * Set mobile mode state.
     * @private
     * @return {?}
     */
    DataTablePaginationComponent.prototype.setMobileModeState = /**
     * Set mobile mode state.
     * @private
     * @return {?}
     */
    function () {
        this.isMobile = this.paginationContainer.nativeElement.clientWidth < 767;
    };
    /**
     * First page click event handler.
     */
    /**
     * First page click event handler.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.firstPageClick = /**
     * First page click event handler.
     * @return {?}
     */
    function () {
        this.config.offset = 0;
        this.eventStateService.dataFetchStream.emit(DataFetchMode.SOFT_LOAD);
    };
    /**
     * Previous page click event handler.
     */
    /**
     * Previous page click event handler.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.previousPageClick = /**
     * Previous page click event handler.
     * @return {?}
     */
    function () {
        this.config.offset = this.config.offset - Math.min(this.config.limit, this.config.offset);
        this.eventStateService.dataFetchStream.emit(DataFetchMode.SOFT_LOAD);
    };
    /**
     * Next page click event handler.
     */
    /**
     * Next page click event handler.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.nextPageClick = /**
     * Next page click event handler.
     * @return {?}
     */
    function () {
        this.config.offset = this.config.offset + this.config.limit;
        this.eventStateService.dataFetchStream.emit(DataFetchMode.SOFT_LOAD);
    };
    /**
     * Last page click event handler.
     */
    /**
     * Last page click event handler.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.lastPageClick = /**
     * Last page click event handler.
     * @return {?}
     */
    function () {
        this.config.offset = (this.maxPage - 1) * this.config.limit;
        this.eventStateService.dataFetchStream.emit(DataFetchMode.SOFT_LOAD);
    };
    Object.defineProperty(DataTablePaginationComponent.prototype, "maxPage", {
        /**
         * Get maximum page count.
         */
        get: /**
         * Get maximum page count.
         * @return {?}
         */
        function () {
            return Math.ceil(this.dataStateService.itemCount / this.config.limit);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTablePaginationComponent.prototype, "page", {
        /**
         * Get page number.
         */
        get: /**
         * Get page number.
         * @return {?}
         */
        function () {
            return Math.floor(this.config.offset / this.config.limit) + 1;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Check limit invalid status. True if limit is invalid.
     * @param element Limit input DOM element.
     * @return Invalid status.
     */
    /**
     * Check limit invalid status. True if limit is invalid.
     * @param {?} element Limit input DOM element.
     * @return {?} Invalid status.
     */
    DataTablePaginationComponent.prototype.isInvalidLimit = /**
     * Check limit invalid status. True if limit is invalid.
     * @param {?} element Limit input DOM element.
     * @return {?} Invalid status.
     */
    function (element) {
        /** @type {?} */
        var limit = Number(element.value);
        return element.value === '' || limit > this.config.maxLimit || limit < 1;
    };
    /**
     * On page size change event handler.
     * @param element HTML input element.
     */
    /**
     * On page size change event handler.
     * @param {?} element HTML input element.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.onPageSizeChange = /**
     * On page size change event handler.
     * @param {?} element HTML input element.
     * @return {?}
     */
    function (element) {
        if (this.isInvalidLimit(element)) {
            element.value = String(this.config.limit);
            return;
        }
        /** @type {?} */
        var limit = Number(element.value);
        if (this.config.limit !== limit) {
            this.config.offset = 0;
            this.config.limit = limit;
            this.eventStateService.dataFetchStream.emit(DataFetchMode.SOFT_LOAD);
        }
    };
    /**
     * On page size revert event handler.
     * @param element HTML input element.
     */
    /**
     * On page size revert event handler.
     * @param {?} element HTML input element.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.onPageSizeRevert = /**
     * On page size revert event handler.
     * @param {?} element HTML input element.
     * @return {?}
     */
    function (element) {
        element.value = String(this.config.limit);
    };
    /**
     * Is invalid page index; True if page index is invalid.
     * @param element Page index DOM element.
     * @return Invalid status.
     */
    /**
     * Is invalid page index; True if page index is invalid.
     * @param {?} element Page index DOM element.
     * @return {?} Invalid status.
     */
    DataTablePaginationComponent.prototype.isInvalidPageIndex = /**
     * Is invalid page index; True if page index is invalid.
     * @param {?} element Page index DOM element.
     * @return {?} Invalid status.
     */
    function (element) {
        /** @type {?} */
        var page = Number(element.value);
        return element.value === '' || page > this.maxPage || page < 1;
    };
    /**
     * On page size change event handler.
     * @param element HTML input element.
     */
    /**
     * On page size change event handler.
     * @param {?} element HTML input element.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.onPageIndexChange = /**
     * On page size change event handler.
     * @param {?} element HTML input element.
     * @return {?}
     */
    function (element) {
        if (this.isInvalidPageIndex(element)) {
            element.value = String(this.page);
            return;
        }
        /** @type {?} */
        var page = Number(element.value);
        if (this.page !== page) {
            this.config.offset = (page - 1) * this.config.limit;
            this.eventStateService.dataFetchStream.emit(DataFetchMode.SOFT_LOAD);
        }
    };
    /**
     * On page size revert event handler.
     * @param element HTML input element.
     */
    /**
     * On page size revert event handler.
     * @param {?} element HTML input element.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.onPageIndexRevert = /**
     * On page size revert event handler.
     * @param {?} element HTML input element.
     * @return {?}
     */
    function (element) {
        element.value = String(this.page);
    };
    Object.defineProperty(DataTablePaginationComponent.prototype, "hasPrevious", {
        /**
         * Get previous page availability status.
         */
        get: /**
         * Get previous page availability status.
         * @return {?}
         */
        function () {
            return this.config.offset <= 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTablePaginationComponent.prototype, "hasNext", {
        /**
         * Get next page availability status.
         */
        get: /**
         * Get next page availability status.
         * @return {?}
         */
        function () {
            return this.config.offset + this.config.limit >= this.dataStateService.itemCount;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTablePaginationComponent.prototype, "startRowIndex", {
        /**
         * Get start row index.
         */
        get: /**
         * Get start row index.
         * @return {?}
         */
        function () {
            return this.config.offset + 1;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTablePaginationComponent.prototype, "endRowIndex", {
        /**
         * Get end row index.
         */
        get: /**
         * Get end row index.
         * @return {?}
         */
        function () {
            return Math.min(this.config.offset + this.config.limit, this.dataStateService.itemCount);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Prevent invalid key press.
     * @param event Keyboard event argument object.
     */
    /**
     * Prevent invalid key press.
     * @param {?} event Keyboard event argument object.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.preventInvalidKeyPress = /**
     * Prevent invalid key press.
     * @param {?} event Keyboard event argument object.
     * @return {?}
     */
    function (event) {
        if ((event.key >= '0' && event.key <= '9') ||
            event.key === 'ArrowLeft' ||
            event.key === 'ArrowRight' ||
            event.key === 'Delete' ||
            event.key === 'Backspace' ||
            event.key === 'Escape' ||
            event.key === 'Enter') {
            return;
        }
        else {
            event.preventDefault();
        }
    };
    /**
     * component init lifecycle event handler.
     */
    /**
     * component init lifecycle event handler.
     * @return {?}
     */
    DataTablePaginationComponent.prototype.ngOnInit = /**
     * component init lifecycle event handler.
     * @return {?}
     */
    function () {
        var _this = this;
        this.setMobileModeState();
        this.resizeService.resize.pipe(debounceTime(200)).subscribe((/**
         * @return {?}
         */
        function () {
            _this.setMobileModeState();
        }));
    };
    DataTablePaginationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-data-table-pagination',
                    template: "<div class=\"ng-data-table-pagination-container\" #paginationContainer>\n  <div class=\"ng-data-table-pagination-range\" [hidden]=\"isMobile\">\n    {{ config.translations.pagination.rangeKey }} <span [textContent]=\"startRowIndex\"></span>\n    -<span [textContent]=\"endRowIndex\"></span>\n    {{ config.translations.pagination.rangeSeparator }}\n    <span [textContent]=\"dataStateService.itemCount\"></span>\n  </div>\n  <div class=\"ng-data-table-pagination-controllers\" [class.sm-controllers]=\"isMobile\">\n    <div class=\"ng-data-table-pagination-limit\">\n      <div class=\"ng-data-table-pagination-input-container\">\n        <label class=\"ng-data-table-pagination-input-label\">{{ config.translations.pagination.limit }}</label>\n        <input\n          #limitInput\n          type=\"number\"\n          class=\"ng-data-table-pagination-input\"\n          min=\"1\"\n          step=\"1\"\n          [attr.max]=\"config.maxLimit\"\n          [class.ng-data-table-input-error]=\"isInvalidLimit(limitInput)\"\n          [ngModel]=\"config.limit\"\n          required\n          (keypress)=\"preventInvalidKeyPress($event)\"\n          (keyup.enter)=\"onPageSizeChange(limitInput)\"\n          (keyup.esc)=\"onPageSizeRevert(limitInput)\"\n        />\n      </div>\n    </div>\n    <div class=\"ng-data-table-pagination-pages\">\n      <button\n        [disabled]=\"hasPrevious\"\n        (click)=\"firstPageClick()\"\n        [title]=\"config.translations.pagination.firstTooltip\"\n        class=\"ng-data-table-action-button ng-data-table-pagination-firstpage\"\n      ></button>\n      <button\n        [disabled]=\"hasPrevious\"\n        (click)=\"previousPageClick()\"\n        [title]=\"config.translations.pagination.previousTooltip\"\n        class=\"ng-data-table-action-button ng-data-table-pagination-prevpage\"\n      ></button>\n      <div class=\"ng-data-table-pagination-page\">\n        <div class=\"ng-data-table-pagination-input-container\">\n          <input\n            #pageInput\n            type=\"number\"\n            class=\"ng-data-table-pagination-input\"\n            min=\"1\"\n            step=\"1\"\n            [attr.max]=\"maxPage\"\n            [class.ng-data-table-input-error]=\"isInvalidPageIndex(pageInput)\"\n            [ngModel]=\"page\"\n            required\n            (keypress)=\"preventInvalidKeyPress($event)\"\n            (keyup.enter)=\"onPageIndexChange(pageInput)\"\n            (keyup.esc)=\"onPageIndexRevert(pageInput)\"\n          />\n          <div class=\"ng-data-table-pagination-input-label\">\n            <span>/ </span>\n            <span [textContent]=\"maxPage\"></span>\n          </div>\n        </div>\n      </div>\n      <button\n        [disabled]=\"hasNext\"\n        (click)=\"nextPageClick()\"\n        [title]=\"config.translations.pagination.nextTooltip\"\n        class=\"ng-data-table-action-button ng-data-table-pagination-nextpage\"\n      ></button>\n      <button\n        [disabled]=\"hasNext\"\n        (click)=\"lastPageClick()\"\n        [title]=\"config.translations.pagination.lastTooltip\"\n        class=\"ng-data-table-action-button ng-data-table-pagination-lastpage\"\n      ></button>\n    </div>\n  </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    DataTablePaginationComponent.ctorParameters = function () { return [
        { type: DataTableConfigService },
        { type: DataTableDataStateService },
        { type: DataTableEventStateService },
        { type: ResizeService }
    ]; };
    DataTablePaginationComponent.propDecorators = {
        paginationContainer: [{ type: ViewChild, args: ['paginationContainer', { static: true },] }]
    };
    return DataTablePaginationComponent;
}());
export { DataTablePaginationComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTablePaginationComponent.prototype.paginationContainer;
    /** @type {?} */
    DataTablePaginationComponent.prototype.isMobile;
    /** @type {?} */
    DataTablePaginationComponent.prototype.config;
    /** @type {?} */
    DataTablePaginationComponent.prototype.dataStateService;
    /**
     * @type {?}
     * @private
     */
    DataTablePaginationComponent.prototype.eventStateService;
    /**
     * @type {?}
     * @private
     */
    DataTablePaginationComponent.prototype.resizeService;
}
