// ornamentum/esm2015/data-table/directives/scroll-position.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, NgZone, Input } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { map, pairwise } from 'rxjs/operators';
/**
 * Scroll position directive.
 * Track current scroll position of target element.
 */
export class ScrollPositionDirective {
    /**
     * @param {?} el
     * @param {?} zone
     */
    constructor(el, zone) {
        this.el = el;
        this.zone = zone;
    }
    /**
     * After component initialize lifecycle event handler.
     * @return {?}
     */
    ngAfterViewInit() {
        this.zone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            this.scrollEventSubscription = fromEvent(this.el.nativeElement, 'scroll')
                .pipe(map((/**
             * @return {?}
             */
            () => {
                return {
                    scrollLeft: this.el.nativeElement.scrollLeft,
                    scrollTop: this.el.nativeElement.scrollTop,
                    scrollHeight: this.el.nativeElement.scrollHeight,
                    scrollWidth: this.el.nativeElement.scrollWidth,
                    clientHeight: this.el.nativeElement.clientHeight,
                    clientWidth: this.el.nativeElement.clientWidth,
                };
            })), pairwise(), map((/**
             * @param {?} pair
             * @return {?}
             */
            (pair) => {
                const [previous, current] = pair;
                return Object.assign({}, current, { isHorizontal: previous.scrollLeft !== current.scrollLeft, isVertical: previous.scrollTop !== current.scrollTop });
            })))
                .subscribe((/**
             * @param {?} pos
             * @return {?}
             */
            (pos) => {
                this.ngScrollPosition.next(pos);
            }));
        }));
    }
    /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    ngOnDestroy() {
        if (this.scrollEventSubscription) {
            this.scrollEventSubscription.unsubscribe();
        }
        this.ngScrollPosition.complete();
    }
}
ScrollPositionDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngScrollPosition]'
            },] }
];
/** @nocollapse */
ScrollPositionDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgZone }
];
ScrollPositionDirective.propDecorators = {
    ngScrollPosition: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScrollPositionDirective.prototype.scrollEventSubscription;
    /** @type {?} */
    ScrollPositionDirective.prototype.ngScrollPosition;
    /**
     * @type {?}
     * @private
     */
    ScrollPositionDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    ScrollPositionDirective.prototype.zone;
}
