// ornamentum/esm2015/data-table/directives/scroll-element.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * Scroll element directive.
 * Used to set scroll position to target element.
 */
export class ScrollElementDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
    }
    /**
     * After component initialize lifecycle event handler.
     * @return {?}
     */
    ngAfterViewInit() {
        this.scrollPositionSubscription = this.ngScrollElement.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.el.nativeElement.scrollLeft = value.scrollLeft;
        }));
    }
    /**
     * Component destroy lifecycle event handler.
     * @return {?}
     */
    ngOnDestroy() {
        if (this.scrollPositionSubscription) {
            this.scrollPositionSubscription.unsubscribe();
        }
    }
}
ScrollElementDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngScrollElement]'
            },] }
];
/** @nocollapse */
ScrollElementDirective.ctorParameters = () => [
    { type: ElementRef }
];
ScrollElementDirective.propDecorators = {
    ngScrollElement: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScrollElementDirective.prototype.scrollPositionSubscription;
    /** @type {?} */
    ScrollElementDirective.prototype.ngScrollElement;
    /**
     * @type {?}
     * @private
     */
    ScrollElementDirective.prototype.el;
}
