// ornamentum/esm5/dropdown/services/dropdown-resource.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ReplaySubject, of } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { get } from '../../utility/services/object-utility.class';
/**
 * Dropdown resource service. Dropdown client side data query is handled via this service.
 * @template T
 */
var DropdownResourceService = /** @class */ (function () {
    function DropdownResourceService() {
    }
    /**
     * Set source data source.
     * @param dataSource Data source observable.
     */
    /**
     * Set source data source.
     * @param {?} dataSource Data source observable.
     * @return {?}
     */
    DropdownResourceService.prototype.setDataSource = /**
     * Set source data source.
     * @param {?} dataSource Data source observable.
     * @return {?}
     */
    function (dataSource) {
        var _this = this;
        this.dispose();
        if (this.optionDataStream && !this.optionDataStream.closed) {
            this.optionDataStream.complete();
        }
        this.optionDataStream = new ReplaySubject(1);
        this.dataSourceSubscription = dataSource.subscribe((/**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            _this.optionDataStream.next(options);
        }));
    };
    /**
     * Query data.
     * @param params Dropdown request parameters.
     */
    /**
     * Query data.
     * @param {?} params Dropdown request parameters.
     * @return {?}
     */
    DropdownResourceService.prototype.query = /**
     * Query data.
     * @param {?} params Dropdown request parameters.
     * @return {?}
     */
    function (params) {
        return this.optionDataStream.pipe(switchMap((/**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            /** @type {?} */
            var result = options.slice();
            if (params.filter && params.filter.value) {
                /** @type {?} */
                var value_1 = String(params.filter.value).toLowerCase();
                result = result.filter((/**
                 * @param {?} option
                 * @return {?}
                 */
                function (option) {
                    /** @type {?} */
                    var key = String(get(option, params.filter.key)).toLowerCase();
                    return key.includes(value_1);
                }));
            }
            if (params.offset !== undefined) {
                /** @type {?} */
                var offset = params.offset + 1 > result.length ? 0 : params.offset;
                if (params.limit === undefined) {
                    result = result.slice(offset, result.length);
                }
                else {
                    result = result.slice(offset, offset + params.limit);
                }
            }
            return of({
                options: result,
                count: options.length
            });
        })));
    };
    /**
     * Dispose data source.
     */
    /**
     * Dispose data source.
     * @return {?}
     */
    DropdownResourceService.prototype.dispose = /**
     * Dispose data source.
     * @return {?}
     */
    function () {
        if (this.dataSourceSubscription) {
            this.dataSourceSubscription.unsubscribe();
            this.dataSourceSubscription = null;
        }
        if (this.optionDataStream && !this.optionDataStream.closed) {
            this.optionDataStream.complete();
        }
    };
    DropdownResourceService.decorators = [
        { type: Injectable }
    ];
    return DropdownResourceService;
}());
export { DropdownResourceService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DropdownResourceService.prototype.optionDataStream;
    /**
     * @type {?}
     * @private
     */
    DropdownResourceService.prototype.dataSourceSubscription;
}
