// ornamentum/esm5/dropdown/services/dropdown-event-state.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
/**
 * Dropdown event state service; Manage dropdown event handler state.
 */
var DropdownEventStateService = /** @class */ (function () {
    function DropdownEventStateService() {
        this.dataFetchStream = new EventEmitter();
        this.dataBoundStream = new EventEmitter();
        this.selectChangeStream = new EventEmitter();
        this.initStream = new EventEmitter();
        this.staticDataSourceStream = new ReplaySubject(1);
    }
    DropdownEventStateService.decorators = [
        { type: Injectable }
    ];
    return DropdownEventStateService;
}());
export { DropdownEventStateService };
if (false) {
    /** @type {?} */
    DropdownEventStateService.prototype.dataFetchStream;
    /** @type {?} */
    DropdownEventStateService.prototype.dataBoundStream;
    /** @type {?} */
    DropdownEventStateService.prototype.selectChangeStream;
    /** @type {?} */
    DropdownEventStateService.prototype.initStream;
    /** @type {?} */
    DropdownEventStateService.prototype.staticDataSourceStream;
}
