// ornamentum/esm2015/data-table/services/data-table-persistence.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { GlobalRefService } from '../../utility/utility.module';
import { DataTableConfigService } from './data-table-config.service';
/**
 * Data table persistence service; Manage data table state when persist state options is enabled.
 */
export class DataTablePersistenceService {
    /**
     * @param {?} globalRefService
     * @param {?} config
     */
    constructor(globalRefService, config) {
        this.globalRefService = globalRefService;
        this.config = config;
    }
    /**
     * Set table state storage mode.
     * @param {?} value Storage mode.
     * @return {?}
     */
    set storageMode(value) {
        if (this.globalRefService.isBrowser) {
            if (value === 'local') {
                this.storage = this.globalRefService.window.localStorage;
            }
            else {
                this.storage = this.globalRefService.window.sessionStorage;
            }
        }
    }
    /**
     * Set table state by identifier.
     * @param {?} id Table identifier.
     * @param {?} value Data table request parameters object.
     * @param {?=} version Data version.
     * @return {?}
     */
    setState(id, value, version = 'v1') {
        if (this.globalRefService.isBrowser) {
            /** @type {?} */
            const data = {
                ver: version,
                val: value
            };
            this.storage.setItem(`${this.config.stateKeyPrefix}${id}`, JSON.stringify(data));
        }
    }
    /**
     * Get table state by identifier.
     * @param {?} id Table identifier.
     * @param {?=} version Data version.
     * @return {?} Data table request params object.
     */
    getState(id, version = 'v1') {
        if (this.globalRefService.isBrowser) {
            /** @type {?} */
            const data = JSON.parse(this.storage.getItem(`${this.config.stateKeyPrefix}${id}`));
            if (data && data.ver === version) {
                return data.val;
            }
        }
        return null;
    }
}
DataTablePersistenceService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DataTablePersistenceService.ctorParameters = () => [
    { type: GlobalRefService },
    { type: DataTableConfigService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTablePersistenceService.prototype.storage;
    /**
     * @type {?}
     * @private
     */
    DataTablePersistenceService.prototype.globalRefService;
    /**
     * @type {?}
     * @private
     */
    DataTablePersistenceService.prototype.config;
}
