// ornamentum/esm2015/utility/services/component-loader.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Dynamic component loader; Responsible of dynamically rendering angular components.
 * @record
 * @template T
 */
export function ComponentLoader() { }
if (false) {
    /**
     * Hide component if visible.
     * @return {?} Rendered component reference.
     */
    ComponentLoader.prototype.hide = function () { };
    /**
     * Render component if not available, else display hidden component.
     * @param {?} component Component class type.
     * @param {?} parentElement Parent element to append the target component.
     * @param {?} injector Component injector reference.
     * @param {?=} options Component loader options object.
     * @return {?} Rendered component reference.
     */
    ComponentLoader.prototype.show = function (component, parentElement, injector, options) { };
    /**
     * Toggle component display state or render if not available.
     * @param {?} component Component class type.
     * @param {?} parentElement Parent element to append the target component.
     * @param {?} injector Component injector reference.
     * @param {?=} options Component loader options object.
     * @return {?} Rendered component reference.
     */
    ComponentLoader.prototype.toggle = function (component, parentElement, injector, options) { };
    /**
     * Dispose rendered component reference and bindings.
     * @return {?}
     */
    ComponentLoader.prototype.dispose = function () { };
}
