// ornamentum/esm2015/utility/services/global-ref.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * Global reference service; List all global javascript references here.
 */
export class GlobalRefService {
    constructor() {
        this.setScrollbarWidth();
    }
    /**
     * Set scrollbar width of current browser environment; Ony set on browser environment to support SSR.
     * @return {?}
     */
    setScrollbarWidth() {
        if (this.scrollbarWidthValue !== undefined) {
            return;
        }
        if (this.isBrowser) {
            /** @type {?} */
            const outer = document.createElement('div');
            outer.style.visibility = 'hidden';
            outer.style.width = '100px';
            outer.style.msOverflowStyle = 'scrollbar'; // needed for WinJS apps
            document.body.appendChild(outer);
            /** @type {?} */
            const widthNoScroll = outer.offsetWidth;
            // force scrollbars
            outer.style.overflow = 'scroll';
            // add inner div
            /** @type {?} */
            const inner = document.createElement('div');
            inner.style.width = '100%';
            outer.appendChild(inner);
            /** @type {?} */
            const widthWithScroll = inner.offsetWidth;
            // remove divs
            outer.parentNode.removeChild(outer);
            this.scrollbarWidthValue = widthNoScroll - widthWithScroll;
        }
    }
    /**
     * Get scrollbar width.
     * @return {?} Scroll bar width.
     */
    get scrollbarWidth() {
        return this.scrollbarWidthValue;
    }
    /**
     * Get window reference.
     * @return {?} Window reference.
     */
    get window() {
        return window;
    }
    /**
     * Is browser environment.
     * @return {?} True if current environment is browser.
     */
    get isBrowser() {
        return typeof window !== 'undefined';
    }
}
GlobalRefService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
GlobalRefService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    GlobalRefService.prototype.scrollbarWidthValue;
}
