// ornamentum/esm2015/utility/utility.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { InitDirective } from './directives/init.directive';
import { FocusDirective } from './directives/focus.directive';
import { PixelConverterPipe } from './pipes/pixel-converter.pipe';
import { UtilityConfigService } from './services/utility-config.service';
import { DragAndDropService } from './services/drag-and-drop.service';
import { GlobalRefService } from './services/global-ref.service';
import { PopoverComponentLoaderFactoryService } from './services/popover-component-loader-factory.service';
import { ResizeService } from './services/resize.service';
import { ValidatorService } from './services/validator.service';
/** @type {?} */
const DIRECTIVES = [InitDirective, FocusDirective];
/** @type {?} */
const PIPES = [PixelConverterPipe];
/** @type {?} */
const PROVIDERS = [DragAndDropService, GlobalRefService, PopoverComponentLoaderFactoryService, ResizeService, ValidatorService];
/**
 * Module representing utility providers, directives and pipes.
 */
export class UtilityModule {
    /**
     * Set module root configuration overrides.
     * @return {?} Module with custom providers.
     */
    static forRoot() {
        return {
            ngModule: UtilityModule,
            providers: [UtilityConfigService]
        };
    }
}
UtilityModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [...DIRECTIVES, ...PIPES],
                providers: [...PROVIDERS],
                exports: [...DIRECTIVES, ...PIPES]
            },] }
];
export { DragAndDropService } from './services/drag-and-drop.service';
export { GlobalRefService } from './services/global-ref.service';
export { UtilityConfigService } from './services/utility-config.service';
export { PopoverComponentLoaderFactoryService } from './services/popover-component-loader-factory.service';
export { PopoverComponentLoader } from './services/popover-component-loader.class';
