// ornamentum/esm5/utility/services/popover-component-loader-factory.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApplicationRef, ComponentFactoryResolver, Injectable } from '@angular/core';
import { GlobalRefService } from './global-ref.service';
import { ResizeService } from './resize.service';
import { PopoverComponentLoader } from './popover-component-loader.class';
/**
 * Popover component loader factory service.
 */
var PopoverComponentLoaderFactoryService = /** @class */ (function () {
    function PopoverComponentLoaderFactoryService(componentFactoryResolver, appRef, globalRefService, resizeService) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.appRef = appRef;
        this.globalRefService = globalRefService;
        this.resizeService = resizeService;
    }
    /**
     * Create new component loader.
     * @param renderer Angular renderer reference.
     * @return Popover component loader reference.
     */
    /**
     * Create new component loader.
     * @template T
     * @param {?} renderer Angular renderer reference.
     * @return {?} Popover component loader reference.
     */
    PopoverComponentLoaderFactoryService.prototype.createLoader = /**
     * Create new component loader.
     * @template T
     * @param {?} renderer Angular renderer reference.
     * @return {?} Popover component loader reference.
     */
    function (renderer) {
        return new PopoverComponentLoader(this.componentFactoryResolver, this.appRef, this.globalRefService, renderer, this.resizeService);
    };
    PopoverComponentLoaderFactoryService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PopoverComponentLoaderFactoryService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ApplicationRef },
        { type: GlobalRefService },
        { type: ResizeService }
    ]; };
    return PopoverComponentLoaderFactoryService;
}());
export { PopoverComponentLoaderFactoryService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PopoverComponentLoaderFactoryService.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    PopoverComponentLoaderFactoryService.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    PopoverComponentLoaderFactoryService.prototype.globalRefService;
    /**
     * @type {?}
     * @private
     */
    PopoverComponentLoaderFactoryService.prototype.resizeService;
}
