// ornamentum/esm2015/data-table/components/data-table-col-group/data-table-col-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { DataTableColumnComponent } from '../data-table-column/data-table-column.component';
import { DataTableConfigService } from '../../services/data-table-config.service';
/**
 * Data table column group component. This component wraps data table column groups.
 */
export class DataTableColGroupComponent {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
    }
}
DataTableColGroupComponent.decorators = [
    { type: Component, args: [{
                exportAs: 'ngDataTableColGroup',
                // tslint:disable-next-line
                selector: '[ngDataTableColGroup]',
                template: "<col *ngIf=\"config.expandableRows\" [style.width]=\"config.expanderColumnWidth | ngPx\" />\n<col *ngIf=\"config.showIndexColumn\" [style.width]=\"config.indexColumnWidth | ngPx\" />\n<col *ngIf=\"config.showRowSelectCheckboxColumn\" [style.width]=\"config.selectionColumnWidth | ngPx\" />\n<ng-container *ngFor=\"let column of columns\">\n  <col [style.width]=\"column.width | ngPx\" *ngIf=\"column.visible\" />\n</ng-container>\n"
            }] }
];
/** @nocollapse */
DataTableColGroupComponent.ctorParameters = () => [
    { type: DataTableConfigService }
];
DataTableColGroupComponent.propDecorators = {
    columns: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DataTableColGroupComponent.prototype.columns;
    /** @type {?} */
    DataTableColGroupComponent.prototype.config;
}
