// ornamentum/esm5/data-table/services/data-table-resource.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ReplaySubject, of } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { orderBy, get } from '../../utility/services/object-utility.class';
/**
 * Data table resource service; Manage data table client side data querying.
 * @template T
 */
var DataTableResourceService = /** @class */ (function () {
    function DataTableResourceService() {
    }
    /**
     * Set data source stream to query.
     * @param dataSource Data source stream.
     */
    /**
     * Set data source stream to query.
     * @param {?} dataSource Data source stream.
     * @return {?}
     */
    DataTableResourceService.prototype.setDataSource = /**
     * Set data source stream to query.
     * @param {?} dataSource Data source stream.
     * @return {?}
     */
    function (dataSource) {
        var _this = this;
        this.dispose();
        if (this.itemDataStream && !this.itemDataStream.closed) {
            this.itemDataStream.complete();
        }
        this.itemDataStream = new ReplaySubject(1);
        this.dataSourceSubscription = dataSource.subscribe((/**
         * @param {?} items
         * @return {?}
         */
        function (items) {
            _this.itemDataStream.next(items);
        }));
    };
    /**
     * Query items by data table request params.
     * @param params Data table parameters object.
     * @return Query result stream.
     */
    /**
     * Query items by data table request params.
     * @param {?} params Data table parameters object.
     * @return {?} Query result stream.
     */
    DataTableResourceService.prototype.query = /**
     * Query items by data table request params.
     * @param {?} params Data table parameters object.
     * @return {?} Query result stream.
     */
    function (params) {
        return this.itemDataStream.pipe(switchMap((/**
         * @param {?} items
         * @return {?}
         */
        function (items) {
            /** @type {?} */
            var itemCount = items.length;
            /** @type {?} */
            var result = items.slice();
            if (params.fields.length) {
                /** @type {?} */
                var filterFields_1 = params.fields.filter((/**
                 * @param {?} field
                 * @return {?}
                 */
                function (field) { return field.filterable; }));
                if (filterFields_1.length) {
                    result = items.filter((/**
                     * @param {?} item
                     * @return {?}
                     */
                    function (item) {
                        return filterFields_1.every((/**
                         * @param {?} filterColumn
                         * @return {?}
                         */
                        function (filterColumn) {
                            if (filterColumn.filterExpression) {
                                return filterColumn.filterExpression(item, filterColumn.field, filterColumn.filterValue);
                            }
                            if (filterColumn.filterValue === undefined || filterColumn.filterValue === '') {
                                return true;
                            }
                            /** @type {?} */
                            var fieldValue = get(item, filterColumn.field);
                            if (fieldValue === undefined) {
                                return true;
                            }
                            if (Array.isArray(filterColumn.filterValue)) {
                                return filterColumn.filterValue.length === 0 || filterColumn.filterValue.includes(fieldValue);
                            }
                            /** @type {?} */
                            var value = String(fieldValue).toLowerCase();
                            /** @type {?} */
                            var filterValue = String(filterColumn.filterValue).toLowerCase();
                            return value.includes(filterValue);
                        }));
                    }));
                    itemCount = result.length;
                }
                /** @type {?} */
                var sortColumns = params.fields.filter((/**
                 * @param {?} column
                 * @return {?}
                 */
                function (column) {
                    return column.sortable && column.sortOrder !== '';
                }));
                if (sortColumns.length) {
                    /** @type {?} */
                    var orderedSortColumns = sortColumns;
                    if (sortColumns.length > 1) {
                        orderedSortColumns = sortColumns.concat().sort((/**
                         * @param {?} a
                         * @param {?} b
                         * @return {?}
                         */
                        function (a, b) {
                            return a.sortPriority - b.sortPriority;
                        }));
                    }
                    /** @type {?} */
                    var orderParams = orderedSortColumns.reduce((/**
                     * @param {?} accumulator
                     * @param {?} column
                     * @return {?}
                     */
                    function (accumulator, column) {
                        if (accumulator) {
                            accumulator.fields.push(column.field);
                            accumulator.orders.push(column.sortOrder);
                        }
                        return accumulator;
                    }), {
                        fields: [],
                        orders: []
                    });
                    result = orderBy(result, orderParams.fields, orderParams.orders);
                }
            }
            if (params.offset !== undefined) {
                /** @type {?} */
                var offset = params.offset + 1 > result.length ? 0 : params.offset;
                if (params.limit === undefined) {
                    result = result.slice(offset, result.length);
                }
                else {
                    result = result.slice(offset, offset + params.limit);
                }
            }
            return of({
                items: result,
                count: itemCount
            });
        })));
    };
    /**
     * Extract data table filter options.
     * @param filterColumn Data table column component.
     * @return Filter options collection stream.
     */
    /**
     * Extract data table filter options.
     * @param {?} filterColumn Data table column component.
     * @return {?} Filter options collection stream.
     */
    DataTableResourceService.prototype.extractFilterOptions = /**
     * Extract data table filter options.
     * @param {?} filterColumn Data table column component.
     * @return {?} Filter options collection stream.
     */
    function (filterColumn) {
        return this.itemDataStream.pipe(switchMap((/**
         * @param {?} items
         * @return {?}
         */
        function (items) {
            /** @type {?} */
            var filteredItems = items
                .reduce((/**
             * @param {?} acc
             * @param {?} item
             * @param {?} index
             * @return {?}
             */
            function (acc, item, index) {
                if (filterColumn.filterFieldMapper) {
                    return acc.concat(filterColumn.filterFieldMapper(item, index));
                }
                /** @type {?} */
                var filterField = filterColumn.filterField || filterColumn.field;
                /** @type {?} */
                var filterValue = get(item, filterField);
                acc.push({
                    key: filterValue,
                    value: filterValue
                });
                return acc;
            }), [])
                .filter((/**
             * @param {?} value
             * @param {?} index
             * @param {?} self
             * @return {?}
             */
            function (value, index, self) {
                return self.findIndex((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item.key === value.key; })) === index;
            }));
            return of(filteredItems);
        })));
    };
    /**
     * Dispose client data source streams.
     */
    /**
     * Dispose client data source streams.
     * @return {?}
     */
    DataTableResourceService.prototype.dispose = /**
     * Dispose client data source streams.
     * @return {?}
     */
    function () {
        if (this.dataSourceSubscription) {
            this.dataSourceSubscription.unsubscribe();
            this.dataSourceSubscription = null;
        }
        if (this.itemDataStream && !this.itemDataStream.closed) {
            this.itemDataStream.complete();
        }
    };
    DataTableResourceService.decorators = [
        { type: Injectable }
    ];
    return DataTableResourceService;
}());
export { DataTableResourceService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTableResourceService.prototype.itemDataStream;
    /**
     * @type {?}
     * @private
     */
    DataTableResourceService.prototype.dataSourceSubscription;
}
