// ornamentum/esm5/data-table/services/data-table-data-state.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * Data table state manager service; Manage current data table state snapshot.
 */
var DataTableDataStateService = /** @class */ (function () {
    function DataTableDataStateService() {
        this.allRowSelected = false;
        this.selectedRows = [];
        this.dataRows = [];
        this.dataLoading = true;
        this.substituteRows = [];
        this.heardReload = false;
        this.currentSortPriority = 0;
        this.onDynamicRowSpanExtract = (/**
         * @return {?}
         */
        function () { return 1; });
    }
    Object.defineProperty(DataTableDataStateService.prototype, "showNoDataOverlay", {
        /**
         * Get show no data overlay status.
         * @return True if no data overlay should be shown.
         */
        get: /**
         * Get show no data overlay status.
         * @return {?} True if no data overlay should be shown.
         */
        function () {
            return !this.dataRows.length && !this.dataLoading;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Get data table row unique id.
     * @param append Target identifier.
     * @param index Target index.
     */
    /**
     * Get data table row unique id.
     * @param {?} append Target identifier.
     * @param {?} index Target index.
     * @return {?}
     */
    DataTableDataStateService.prototype.getUniqueId = /**
     * Get data table row unique id.
     * @param {?} append Target identifier.
     * @param {?} index Target index.
     * @return {?}
     */
    function (append, index) {
        return this.id + "-dt-" + append + "-" + index;
    };
    DataTableDataStateService.decorators = [
        { type: Injectable }
    ];
    return DataTableDataStateService;
}());
export { DataTableDataStateService };
if (false) {
    /** @type {?} */
    DataTableDataStateService.prototype.id;
    /** @type {?} */
    DataTableDataStateService.prototype.allRowSelected;
    /** @type {?} */
    DataTableDataStateService.prototype.selectedRow;
    /** @type {?} */
    DataTableDataStateService.prototype.selectedRows;
    /** @type {?} */
    DataTableDataStateService.prototype.dataRows;
    /** @type {?} */
    DataTableDataStateService.prototype.itemCount;
    /** @type {?} */
    DataTableDataStateService.prototype.tableWidth;
    /** @type {?} */
    DataTableDataStateService.prototype.dataLoading;
    /** @type {?} */
    DataTableDataStateService.prototype.substituteRows;
    /** @type {?} */
    DataTableDataStateService.prototype.heardReload;
    /** @type {?} */
    DataTableDataStateService.prototype.currentSortPriority;
    /** @type {?} */
    DataTableDataStateService.prototype.relativeParentElement;
    /** @type {?} */
    DataTableDataStateService.prototype.onFilterValueExtract;
    /** @type {?} */
    DataTableDataStateService.prototype.onDataBind;
    /** @type {?} */
    DataTableDataStateService.prototype.onDynamicRowSpanExtract;
}
