// ornamentum/esm5/data-table/services/data-table-event.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
/**
 * Data table event state service; Manage all internal data tale event streams.
 */
var DataTableEventStateService = /** @class */ (function () {
    function DataTableEventStateService() {
        this.allRowSelectChangeStream = new EventEmitter();
        this.dataFetchStream = new EventEmitter();
        this.headerClickStream = new EventEmitter();
        this.rowBindStream = new EventEmitter();
        this.rowClickStream = new EventEmitter();
        this.rowDoubleClickStream = new EventEmitter();
        this.rowSelectChangeStream = new EventEmitter();
        this.cellBindStream = new EventEmitter();
        this.cellClickStream = new EventEmitter();
        this.initStream = new EventEmitter();
        this.dataBoundStream = new EventEmitter();
        this.columnBind = new EventEmitter();
        this.fetchFilterOptionsStream = new ReplaySubject(1);
        this.staticDataSourceStream = new ReplaySubject(1);
    }
    DataTableEventStateService.decorators = [
        { type: Injectable }
    ];
    return DataTableEventStateService;
}());
export { DataTableEventStateService };
if (false) {
    /** @type {?} */
    DataTableEventStateService.prototype.allRowSelectChangeStream;
    /** @type {?} */
    DataTableEventStateService.prototype.dataFetchStream;
    /** @type {?} */
    DataTableEventStateService.prototype.headerClickStream;
    /** @type {?} */
    DataTableEventStateService.prototype.rowBindStream;
    /** @type {?} */
    DataTableEventStateService.prototype.rowClickStream;
    /** @type {?} */
    DataTableEventStateService.prototype.rowDoubleClickStream;
    /** @type {?} */
    DataTableEventStateService.prototype.rowSelectChangeStream;
    /** @type {?} */
    DataTableEventStateService.prototype.cellBindStream;
    /** @type {?} */
    DataTableEventStateService.prototype.cellClickStream;
    /** @type {?} */
    DataTableEventStateService.prototype.initStream;
    /** @type {?} */
    DataTableEventStateService.prototype.dataBoundStream;
    /** @type {?} */
    DataTableEventStateService.prototype.columnBind;
    /** @type {?} */
    DataTableEventStateService.prototype.fetchFilterOptionsStream;
    /** @type {?} */
    DataTableEventStateService.prototype.staticDataSourceStream;
}
