// ornamentum/esm5/data-table/services/data-table-persistence.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { GlobalRefService } from '../../utility/utility.module';
import { DataTableConfigService } from './data-table-config.service';
/**
 * Data table persistence service; Manage data table state when persist state options is enabled.
 */
var DataTablePersistenceService = /** @class */ (function () {
    function DataTablePersistenceService(globalRefService, config) {
        this.globalRefService = globalRefService;
        this.config = config;
    }
    Object.defineProperty(DataTablePersistenceService.prototype, "storageMode", {
        /**
         * Set table state storage mode.
         * @param value Storage mode.
         */
        set: /**
         * Set table state storage mode.
         * @param {?} value Storage mode.
         * @return {?}
         */
        function (value) {
            if (this.globalRefService.isBrowser) {
                if (value === 'local') {
                    this.storage = this.globalRefService.window.localStorage;
                }
                else {
                    this.storage = this.globalRefService.window.sessionStorage;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Set table state by identifier.
     * @param id Table identifier.
     * @param version Data version.
     * @param value Data table request parameters object.
     */
    /**
     * Set table state by identifier.
     * @param {?} id Table identifier.
     * @param {?} value Data table request parameters object.
     * @param {?=} version Data version.
     * @return {?}
     */
    DataTablePersistenceService.prototype.setState = /**
     * Set table state by identifier.
     * @param {?} id Table identifier.
     * @param {?} value Data table request parameters object.
     * @param {?=} version Data version.
     * @return {?}
     */
    function (id, value, version) {
        if (version === void 0) { version = 'v1'; }
        if (this.globalRefService.isBrowser) {
            /** @type {?} */
            var data = {
                ver: version,
                val: value
            };
            this.storage.setItem("" + this.config.stateKeyPrefix + id, JSON.stringify(data));
        }
    };
    /**
     * Get table state by identifier.
     * @param id Table identifier.
     * @param version Data version.
     * @return Data table request params object.
     */
    /**
     * Get table state by identifier.
     * @param {?} id Table identifier.
     * @param {?=} version Data version.
     * @return {?} Data table request params object.
     */
    DataTablePersistenceService.prototype.getState = /**
     * Get table state by identifier.
     * @param {?} id Table identifier.
     * @param {?=} version Data version.
     * @return {?} Data table request params object.
     */
    function (id, version) {
        if (version === void 0) { version = 'v1'; }
        if (this.globalRefService.isBrowser) {
            /** @type {?} */
            var data = JSON.parse(this.storage.getItem("" + this.config.stateKeyPrefix + id));
            if (data && data.ver === version) {
                return data.val;
            }
        }
        return null;
    };
    DataTablePersistenceService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DataTablePersistenceService.ctorParameters = function () { return [
        { type: GlobalRefService },
        { type: DataTableConfigService }
    ]; };
    return DataTablePersistenceService;
}());
export { DataTablePersistenceService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTablePersistenceService.prototype.storage;
    /**
     * @type {?}
     * @private
     */
    DataTablePersistenceService.prototype.globalRefService;
    /**
     * @type {?}
     * @private
     */
    DataTablePersistenceService.prototype.config;
}
