// ornamentum/esm2015/data-table/components/data-table-head/data-table-head.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
/**
 * Data table header component. Render data table column title and filter header rows.
 */
export class DataTableHeadComponent {
    /**
     * Get filter column availability status.
     * @return {?} True if there is at least one filter column.
     */
    get hasFilterColumns() {
        return this.columns.some((/**
         * @param {?} column
         * @return {?}
         */
        (column) => column.filterable));
    }
}
DataTableHeadComponent.decorators = [
    { type: Component, args: [{
                exportAs: 'ngDataTableHead',
                // tslint:disable-next-line
                selector: '[ngDataTableHead]',
                template: "<tr ngDataTableColumnTitleHeader [columns]=\"columns\"></tr>\n<tr *ngIf=\"hasFilterColumns\" ngDataTableColumnFilterHeader [columns]=\"columns\"></tr>\n"
            }] }
];
DataTableHeadComponent.propDecorators = {
    columns: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DataTableHeadComponent.prototype.columns;
}
