// ngx-float-button/esm5/lib/ngx-float-item-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, ChangeDetectionStrategy } from '@angular/core';
var NgxFloatItemButtonComponent = /** @class */ (function () {
    function NgxFloatItemButtonComponent() {
        this.color = 'white';
        this.disabled = false;
        this.clicked = new EventEmitter();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    NgxFloatItemButtonComponent.prototype.emitClickEvent = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.disabled) {
            return this.disabled;
        }
        this.clicked.emit($event);
    };
    NgxFloatItemButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-float-item-button',
                    template: "\n    <div #elementref class=\"item {{ disabled ? 'disabled' : ''}}\" (click)=\"emitClickEvent($event)\">\n        <div class=\"content-wrapper\" #contentref>\n          <div class=\"content\" [style.display]=\"content ? 'block' : 'none'\">{{content}}</div>\n        </div>\n        <a class=\"fab-item\" [style.backgroundColor]=\"color\">\n           <mat-icon> {{ icon }} </mat-icon>\n        </a>\n    </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n      .item {\n        width: 250px;\n        height: 40px;\n        left: -203px;\n        transform: translate3d(0, 0, 0);\n        transition: transform, opacity ease-out 200ms;\n        transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1);\n        transition-duration: 180ms;\n        position: absolute;\n        cursor: pointer;\n        top: 5px;\n        display: flex;\n        justify-content: flex-end;\n        align-items: center;\n      }\n\n      .item.disabled {\n        pointer-events: none;\n      }\n\n      .item.disabled .fab-item {\n        background-color: lightgray;\n      }\n\n      .content {\n        background: #333333;\n        margin-right: 50px;\n        line-height: 25px;\n        color: white;\n        padding: 2px 7px;\n        border-radius: 3px;\n        display: none;\n        font-size: 12px;\n        height: 25px;\n        margin-top: 4px;\n        box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.26);\n      }\n\n      .fab-item {\n        right: 0;\n        background: white;\n        border-radius: 100%;\n        width: 40px;\n        height: 40px;\n        position: absolute;\n        color: #797979;\n        text-align: center;\n        cursor: pointer;\n        line-height: 50px;\n        box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.26);\n      }\n    "]
                }] }
    ];
    NgxFloatItemButtonComponent.propDecorators = {
        icon: [{ type: Input }],
        content: [{ type: Input }],
        color: [{ type: Input }],
        disabled: [{ type: Input }],
        clicked: [{ type: Output }],
        elementref: [{ type: ViewChild, args: ['elementref',] }],
        contentref: [{ type: ViewChild, args: ['contentref',] }]
    };
    return NgxFloatItemButtonComponent;
}());
export { NgxFloatItemButtonComponent };
if (false) {
    /** @type {?} */
    NgxFloatItemButtonComponent.prototype.icon;
    /** @type {?} */
    NgxFloatItemButtonComponent.prototype.content;
    /** @type {?} */
    NgxFloatItemButtonComponent.prototype.color;
    /** @type {?} */
    NgxFloatItemButtonComponent.prototype.disabled;
    /** @type {?} */
    NgxFloatItemButtonComponent.prototype.clicked;
    /** @type {?} */
    NgxFloatItemButtonComponent.prototype.elementref;
    /** @type {?} */
    NgxFloatItemButtonComponent.prototype.contentref;
}
