// ornamentum/esm2015/data-table/components/data-table-no-data-body/data-table-no-data-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef } from '@angular/core';
import { DataFetchMode } from '../../models/data-fetch-mode.enum';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableEventStateService } from '../../services/data-table-event.service';
/**
 * No data body component. Used to display no data message when data not available.
 */
export class DataTableNoDataBodyComponent {
    /**
     * @param {?} config
     * @param {?} eventStateService
     */
    constructor(config, eventStateService) {
        this.config = config;
        this.eventStateService = eventStateService;
    }
    /**
     * Reset all applied filters.
     * @return {?}
     */
    resetFilters() {
        this.eventStateService.dataFetchStream.emit(DataFetchMode.HARD_RELOAD);
    }
    /**
     * Get default no data template display status.
     * @return {?} True if default no data template should be displayed.
     */
    get showDefaultNoDataTemplate() {
        return !!(!this.noRecordsTemplate && this.config.translations.noDataMessage);
    }
}
DataTableNoDataBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-data-table-no-data-body',
                template: "<div class=\"ng-data-table-no-records-message\" *ngIf=\"showDefaultNoDataTemplate\">\n  <h1 class=\"ng-data-table-no-records-main-heading-message\">{{ config.translations.noDataMessage.header }}</h1>\n  <h2 class=\"ng-data-table-no-records-sub-heading-message\">{{ config.translations.noDataMessage.body }}</h2>\n  <button (click)=\"resetFilters()\" class=\"ng-data-table-no-records-message-button\">Reload Data</button>\n</div>\n<ng-template *ngIf=\"noRecordsTemplate\" [ngTemplateOutlet]=\"noRecordsTemplate\"></ng-template>\n"
            }] }
];
/** @nocollapse */
DataTableNoDataBodyComponent.ctorParameters = () => [
    { type: DataTableConfigService },
    { type: DataTableEventStateService }
];
DataTableNoDataBodyComponent.propDecorators = {
    noRecordsTemplate: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DataTableNoDataBodyComponent.prototype.noRecordsTemplate;
    /** @type {?} */
    DataTableNoDataBodyComponent.prototype.config;
    /** @type {?} */
    DataTableNoDataBodyComponent.prototype.eventStateService;
}
