// ornamentum/esm5/data-table/components/data-table-loading-spinner/data-table-loading-spinner.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef } from '@angular/core';
import { DataTableConfigService } from '../../services/data-table-config.service';
import { DataTableDataStateService } from '../../services/data-table-data-state.service';
/**
 * Data loading spinner component. Displayed within table body while fetching data.
 */
var DataTableLoadingSpinnerComponent = /** @class */ (function () {
    function DataTableLoadingSpinnerComponent(config, dataStateService) {
        this.config = config;
        this.dataStateService = dataStateService;
    }
    DataTableLoadingSpinnerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-data-table-loading-spinner',
                    template: "<div class=\"ng-data-table-loading-spinner\" [hidden]=\"!isLoading\">\n  <div *ngIf=\"loadingSpinnerTemplate\" class=\"ng-data-table-loading-cover\">\n    <div [ngTemplateOutlet]=\"loadingSpinnerTemplate\"></div>\n  </div>\n  <div class=\"ng-data-table-loading-cover\" [class.ng-data-table-show-overlay]=\"showOverlay\" *ngIf=\"!loadingSpinnerTemplate\">\n    <div class=\"ng-data-table-loading-animation\"></div>\n  </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    DataTableLoadingSpinnerComponent.ctorParameters = function () { return [
        { type: DataTableConfigService },
        { type: DataTableDataStateService }
    ]; };
    DataTableLoadingSpinnerComponent.propDecorators = {
        loadingSpinnerTemplate: [{ type: Input }],
        isLoading: [{ type: Input }],
        showOverlay: [{ type: Input }]
    };
    return DataTableLoadingSpinnerComponent;
}());
export { DataTableLoadingSpinnerComponent };
if (false) {
    /** @type {?} */
    DataTableLoadingSpinnerComponent.prototype.loadingSpinnerTemplate;
    /** @type {?} */
    DataTableLoadingSpinnerComponent.prototype.isLoading;
    /** @type {?} */
    DataTableLoadingSpinnerComponent.prototype.showOverlay;
    /** @type {?} */
    DataTableLoadingSpinnerComponent.prototype.config;
    /** @type {?} */
    DataTableLoadingSpinnerComponent.prototype.dataStateService;
}
