// ornamentum/esm5/data-table/components/data-table-column-selector/data-table-column-selector.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { DataTableDataStateService } from '../../services/data-table-data-state.service';
import { DataTableConfigService } from '../../services/data-table-config.service';
/**
 * Data table column selector component. Toggle display state of columns via this popup component.
 */
var DataTableColumnSelectorComponent = /** @class */ (function () {
    function DataTableColumnSelectorComponent(dataStateService, config) {
        this.dataStateService = dataStateService;
        this.config = config;
    }
    DataTableColumnSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-data-table-column-selector',
                    template: "<div class=\"ng-data-table-column-selector-wrapper\" [style.width]=\"width | ngPx\">\n  <div class=\"ng-data-table-column-selector-box-heading\">{{ config.translations.columnSelector.header }}</div>\n  <div class=\"ng-data-table-column-selector-box\">\n    <ng-container *ngFor=\"let column of columns; index as i;\">\n      <div class=\"ng-data-table-column-selector-column ng-data-table-column-selector-checkbox\" *ngIf=\"column.showInColumnSelector\">\n        <div class=\"ng-data-table-checkbox-container\">\n          <input class=\"ng-data-table-checkbox-input\" type=\"checkbox\"\n                 [id]=\"dataStateService.getUniqueId('cs', i)\" [(ngModel)]=\"column.visible\"/>\n          <label [for]=\"dataStateService.getUniqueId('cs', i)\"><span [textContent]=\"column.title\"></span></label>\n        </div>\n      </div>\n    </ng-container>\n  </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    DataTableColumnSelectorComponent.ctorParameters = function () { return [
        { type: DataTableDataStateService },
        { type: DataTableConfigService }
    ]; };
    DataTableColumnSelectorComponent.propDecorators = {
        columns: [{ type: Input }],
        width: [{ type: Input }]
    };
    return DataTableColumnSelectorComponent;
}());
export { DataTableColumnSelectorComponent };
if (false) {
    /** @type {?} */
    DataTableColumnSelectorComponent.prototype.columns;
    /** @type {?} */
    DataTableColumnSelectorComponent.prototype.width;
    /** @type {?} */
    DataTableColumnSelectorComponent.prototype.dataStateService;
    /** @type {?} */
    DataTableColumnSelectorComponent.prototype.config;
}
