// ornamentum/esm2015/dropdown/components/dropdown/dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, EventEmitter, forwardRef, Injector, Input, Output, Renderer2, TemplateRef, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { Observable, of } from 'rxjs';
import { catchError, debounceTime, switchMap } from 'rxjs/operators';
import { get } from '../../../utility/services/object-utility.class';
import { DropdownViewComponent } from '../dropdown-view/dropdown-view.component';
import { PopoverComponentLoaderFactoryService } from '../../../utility/utility.module';
import { DropdownConfigService } from '../../services/dropdown-config.service';
import { DropdownDataStateService } from '../../services/dropdown-data-state.service';
import { DropdownEventStateService } from '../../services/dropdown-event-state.service';
import { DropdownResourceService } from '../../services/dropdown-resource.service';
import { ValidatorService } from '../../../utility/services/validator.service';
/**
 * Dropdown main component.
 */
export class DropdownComponent {
    /**
     * @param {?} componentLoaderFactory
     * @param {?} injector
     * @param {?} eventStateService
     * @param {?} dropdownResourceService
     * @param {?} renderer
     * @param {?} dataStateService
     * @param {?} config
     */
    constructor(componentLoaderFactory, injector, eventStateService, dropdownResourceService, renderer, dataStateService, config) {
        this.componentLoaderFactory = componentLoaderFactory;
        this.injector = injector;
        this.eventStateService = eventStateService;
        this.dropdownResourceService = dropdownResourceService;
        this.renderer = renderer;
        this.dataStateService = dataStateService;
        this.config = config;
        this.dataStateService.componentLoaderRef = this.componentLoaderFactory.createLoader(this.renderer);
        this.dataBound = this.eventStateService.dataBoundStream;
        this.selectChange = this.eventStateService.selectChangeStream;
        this.init = this.eventStateService.initStream;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set optionTemplate(value) {
        this.dataStateService.dropdownOptionTemplate = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set optionGroupHeaderTemplate(value) {
        this.dataStateService.dropdownOptionGroupHeaderTemplate = value;
    }
    // Input - Event handlers
    /**
     * Set data bind callback. This handler is fired on each data fetch request.
     * @param {?} value
     * @return {?}
     */
    set onDataBind(value) {
        this.dataStateService.onDataBind = value;
    }
    // Inputs
    /**
     * Set dropdown loading spinner template reference object. Used by data table component to explicitly pass the template reference.
     * @param {?} value
     * @return {?}
     */
    set loadingSpinnerTemplateRef(value) {
        this.loadingSpinnerTemplate = value;
    }
    /**
     * Set dropdown option template reference. Used by data table component to explicitly pass the template reference.
     * @param {?} value
     * @return {?}
     */
    set optionTemplateRef(value) {
        this.optionTemplate = value;
    }
    /**
     * Set dropdown options group header template reference. Used by data table component to explicitly pass the template reference.
     * @param {?} value
     * @return {?}
     */
    set optionGroupHeaderTemplateRef(value) {
        this.optionGroupHeaderTemplate = value;
    }
    /**
     * Set static dropdown options collection. No need to set data source when static options collection is provided.
     * @param {?} value
     * @return {?}
     */
    set options(value) {
        if (!value) {
            return;
        }
        this.eventStateService.staticDataSourceStream.next(value);
    }
    /**
     * Set data source observable. This observable is used to retrieve dropdown options for data binding.
     * @param {?} source
     * @return {?}
     */
    set dataSource(source) {
        this.initDataSource(source);
    }
    /**
     * Set dropdown unique identifier.
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        if (!ValidatorService.idPatternValidatorExpression.test(value)) {
            throw Error('Invalid [id] input value. Unique identifier parameter only accept string begin with a letter ([A-Za-z]) and may be followed by any number of letters, digits ([0-9]), hyphens ("-"), underscores ("_").');
        }
        this.dataStateService.id = value;
    }
    /**
     * Set translation data object. Used to localize table static label text.
     * @param {?} value
     * @return {?}
     */
    set translations(value) {
        this.config.translations = value;
    }
    /**
     * Set select option track by field path which is used to uniquely identify options for selection tracking.
     * This field support object paths expressions 'root[0].nest'.
     * @param {?} value
     * @return {?}
     */
    set selectTrackBy(value) {
        this.config.selectTrackBy = value;
    }
    /**
     * Set display value track by field path which is used to extract dropdown option display value.
     * This field support object paths expressions 'root[0].nest'.
     * @param {?} value
     * @return {?}
     */
    set displayTrackBy(value) {
        this.config.displayTrackBy = value;
    }
    /**
     * Set options group field path which is used to group the dropdown options.
     * This field support object paths expressions 'root[0].nest'.
     * @param {?} value
     * @return {?}
     */
    set groupByField(value) {
        this.config.groupByField = value;
    }
    /**
     * Set dropdown option disable state field path which is used to disabled state dropdown options.
     * This field support object paths expressions 'root[0].nest'.
     * @param {?} value
     * @return {?}
     */
    set disabledTrackBy(value) {
        this.config.disabledTrackBy = value;
    }
    /**
     * Set selected options collection. These options will be set selected by default on initial load.
     * Applicable only when multi select mode is enabled.
     * @param {?} value
     * @return {?}
     */
    set selectedOptions(value) {
        this.dataStateService.selectedOptions = value || [];
    }
    /**
     * Set selected option. This option is selected by default on initial load.
     * Applicable only when single select mode is enabled.
     * @param {?} value
     * @return {?}
     */
    set selectedOption(value) {
        this.dataStateService.selectedOption = value;
    }
    /**
     * Set number of options to fetch on scroll to bottom action when load on scroll mode is enabled.
     * @param {?} value
     * @return {?}
     */
    set limit(value) {
        this.config.limit = value;
    }
    /**
     * Set wrap selected options in dropdown view and show the number of options selected instead when
     * limit is met or exceeded. Applicable only when multi select mode is enabled.
     * @param {?} value
     * @return {?}
     */
    set wrapDisplaySelectLimit(value) {
        this.config.wrapDisplaySelectLimit = value;
    }
    /**
     * Set infinite scrollable state to load data on demand with scroll motion. Dropdown data fetch call is
     * initiated with limit and offset when user scroll to bottom hence loading the full data set on init.
     * @param {?} value
     * @return {?}
     */
    set loadOnScroll(value) {
        this.config.loadOnScroll = value;
    }
    /**
     * Set view height ratio to trigger data fetch with infinite scrollable mode.
     * Higher ratio will will increase the scroll sensitivity.
     * @param {?} value
     * @return {?}
     */
    set loadViewDistanceRatio(value) {
        this.config.loadViewDistanceRatio = value;
    }
    /**
     * Set option select mode.
     * - 'multi' : Support selecting multiple options.
     * - 'single' : Support selecting a single option from options collection.
     * - 'single-toggle' : Support selecting a single option from options collection. Selection can not be removed but
     * only toggled by tapping on another option.
     * @param {?} value
     * @return {?}
     */
    set selectMode(value) {
        this.config.selectMode = value;
    }
    /**
     * Show dropdown option search filter text-box if true.
     * @param {?} value
     * @return {?}
     */
    set filterable(value) {
        this.config.filterable = value;
    }
    /**
     * Set default filter value to be applied on initial load. All options are displayed when filter text value is
     * empty string. Applicable only when dropdown is filterable.
     * @param {?} value
     * @return {?}
     */
    set filterText(value) {
        this.dataStateService.filterText = value;
    }
    /**
     * Set time based filter debounce to optimize performance and avoid request flooding by reducing the filter
     * request frequency if true. Applicable only when dropdown filterable state is enabled.
     * @param {?} value
     * @return {?}
     */
    set filterDebounce(value) {
        this.config.filterDebounce = value;
    }
    /**
     * Set filter debounce time in milliseconds. Applicable only when searchDebounce is true.
     * @param {?} value
     * @return {?}
     */
    set filterDebounceTime(value) {
        this.config.filterDebounceTime = value;
    }
    /**
     * Set load data on component initialize if true.
     * @param {?} value
     * @return {?}
     */
    set loadDataOnInit(value) {
        this.config.loadDataOnInit = value;
    }
    /**
     * Show selected option remove button if true.
     * Applicable only when multi select mode ios enabled.
     * @param {?} value
     * @return {?}
     */
    set showSelectedOptionRemoveButton(value) {
        this.config.showSelectedOptionRemoveButton = value;
    }
    /**
     * Set show all select options clear button if true.
     * Applicable only when multi select mode ios enabled.
     * @param {?} value
     * @return {?}
     */
    set showClearSelectionButton(value) {
        this.config.showClearSelectionButton = value;
    }
    /**
     * Set options menu width in pixels.
     * @param {?} value
     * @return {?}
     */
    set menuWidth(value) {
        this.config.menuWidth = value;
    }
    /**
     * Set options menu height in pixels.
     * @param {?} value
     * @return {?}
     */
    set menuHeight(value) {
        this.config.menuHeight = value;
    }
    /**
     * Set popup options menu display position relative to dropdown component.
     * Support 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right' values.
     * @param {?} value
     * @return {?}
     */
    set menuPosition(value) {
        this.config.menuPosition = value;
    }
    /**
     * Set dropdown view disabled state.
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this.dataStateService.disabled = value;
    }
    /**
     * Set Close dropdown menu on option select if true.
     * @param {?} value
     * @return {?}
     */
    set closeMenuOnSelect(value) {
        this.config.closeMenuOnSelect = value;
    }
    /**
     * Set show dropdown option select checkbox if true.
     * @param {?} value
     * @return {?}
     */
    set showOptionSelectCheckbox(value) {
        this.config.showOptionSelectCheckbox = value;
    }
    /**
     * Set show dropdown option index checkbox if true.
     * @param {?} value
     * @return {?}
     */
    set showOptionIndex(value) {
        this.config.showOptionIndex = value;
    }
    /**
     * Set show dropdown option TrackBy id checkbox if true.
     * @param {?} value
     * @return {?}
     */
    set showOptionTrackBy(value) {
        this.config.showOptionTrackBy = value;
    }
    /**
     * Multi select option selected item maximum width. Apply ellipsis when selected option display text
     * exceed the max width.
     * @param {?} value
     * @return {?}
     */
    set multiSelectOptionMaxWidth(value) {
        this.config.multiSelectOptionMaxWidth = value;
    }
    /**
     * Set first dropdown option selected on data fetch if true.
     * @param {?} value
     * @return {?}
     */
    set setFirstOptionSelected(value) {
        this.config.setFirstOptionSelected = value;
    }
    /**
     * Trigger select change event on init if true.
     * Can be used to enable selectedOptions or selectedOption associated change trigger.
     * @param {?} value
     * @return {?}
     */
    set triggerSelectChangeOnInit(value) {
        this.config.triggerSelectChangeOnInit = value;
    }
    /**
     * Set trigger select change on explicit model update if true.
     * Applicable only when form binding is used.
     * @param {?} value
     * @return {?}
     */
    set triggerSelectChangeOnModelUpdate(value) {
        this.config.triggerSelectChangeOnModelUpdate = value;
    }
    /**
     * Set trigger select change on first option select change if true.
     * Applicable only when setFirstOptionSelected is true.
     * @param {?} value
     * @return {?}
     */
    set triggerSelectChangeOnFirstOptionSelect(value) {
        this.config.triggerSelectChangeOnFirstOptionSelect = value;
    }
    /**
     * Set dynamically calculate dropdown view dimensions relative to dropdown button width.
     * MenuWith and menuHeight values are ignored when true.
     * @param {?} value
     * @return {?}
     */
    set dynamicDimensionCalculation(value) {
        this.config.dynamicDimensionCalculation = value;
    }
    /**
     * Set dynamic dropdown options view dimensions calculation width ratio relative to dropdown selector.
     * @param {?} value
     * @return {?}
     */
    set dynamicWidthRatio(value) {
        this.config.dynamicWidthRatio = value;
    }
    /**
     * Set dynamic dropdown options view dimensions calculation height ratio relative to dropdown selector.
     * @param {?} value
     * @return {?}
     */
    set dynamicHeightRatio(value) {
        this.config.dynamicHeightRatio = value;
    }
    /**
     * Set relative parent element to render dropdown view container.
     * @param {?} value
     * @return {?}
     */
    set relativeParentElement(value) {
        this.config.relativeParentElement = value;
    }
    /**
     * Initialize data source.
     * @private
     * @param {?} source Data source stream.
     * @return {?}
     */
    initDataSource(source) {
        this.dropdownResourceService.setDataSource(source);
        this.dataStateService.onDataBind = (/**
         * @param {?} params
         * @return {?}
         */
        (params) => {
            if (params.hardReload) {
                this.dropdownResourceService.setDataSource(source);
            }
            return this.dropdownResourceService.query(params);
        });
    }
    /**
     * Performs dropdown toggle event.
     * @param {?} event Mouse click event args.
     * @param {?} element Dropdown button element.
     * @return {?}
     */
    toggleDropdown(event, element) {
        /** @type {?} */
        const target = (/** @type {?} */ (event.target));
        if (target && target.classList && target.classList.contains('ng-ignore-propagation')) {
            return;
        }
        this.dataStateService.componentLoaderRef.toggle(DropdownViewComponent, element, this.injector, {
            relativeParentElement: this.config.relativeParentElement,
            position: this.config.menuPosition
        });
        if (this.config.dynamicDimensionCalculation) {
            this.config.menuWidth = element.offsetWidth * this.config.dynamicWidthRatio;
            this.config.menuHeight = element.offsetWidth * this.config.dynamicHeightRatio;
        }
    }
    /**
     * Get options wrapped state.
     * @return {?}
     */
    get wrapSelectedOptions() {
        if (this.config.wrapDisplaySelectLimit !== undefined) {
            return this.dataStateService.selectedOptions.length > this.config.wrapDisplaySelectLimit;
        }
        return false;
    }
    /**
     * Get wrapped option display text.
     * @return {?}
     */
    get wrappedOptionDisplayText() {
        return `(${this.dataStateService.selectedOptions.length}) ${this.config.translations.selectedOptionWrapPlaceholder}`;
    }
    /**
     * Lifecycle hook that is called when component is destroyed.
     * @return {?}
     */
    ngOnDestroy() {
        if (this.onSelectChangeSubscription) {
            this.onSelectChangeSubscription.unsubscribe();
        }
        this.dataStateService.componentLoaderRef.dispose();
    }
    /**
     * Get selected options availability state.
     * @return {?}
     */
    get hasSelectedOptions() {
        if (this.config.selectMode === 'multi') {
            return !!this.dataStateService.selectedOptions.length;
        }
        return !!this.dataStateService.selectedOption;
    }
    /**
     * Trigger select change.
     * @return {?}
     */
    triggerSelectChange() {
        if (this.config.selectMode === 'multi') {
            this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOptions);
        }
        else {
            this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOption);
        }
    }
    /**
     * Clear selected options.
     * @return {?}
     */
    clearSelectedOptions() {
        if (this.config.selectMode === 'multi') {
            this.dataStateService.selectedOptions = [];
            this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOptions);
        }
        else {
            this.dataStateService.selectedOption = undefined;
            this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOption);
        }
    }
    /**
     * Set disabled state.
     * ControlValueAccessor implementation.
     * @param {?} isDisabled True if disabled.
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.dataStateService.disabled = isDisabled;
    }
    /**
     * Called when value selected value gets updated.
     * ControlValueAccessor implementation.
     * @param {?} value Selected value.
     * @return {?}
     */
    writeValue(value) {
        if (this.config.selectMode === 'multi') {
            this.dataStateService.selectedOptions = value || [];
        }
        else {
            this.dataStateService.selectedOption = value;
        }
        if (this.config.triggerSelectChangeOnModelUpdate) {
            this.triggerSelectChange();
        }
    }
    /**
     * Register on change event.
     * ControlValueAccessor implementation.
     * @param {?} onSelectChange On select change callback function.
     * @return {?}
     */
    registerOnChange(onSelectChange) {
        this.onSelectChangeSubscription = this.selectChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            onSelectChange(value);
        }));
    }
    /**
     * Register on touched event.
     * ControlValueAccessor implementation.
     * @param {?} fn Function reference.
     * @return {?}
     */
    registerOnTouched(fn) {
        // TODO: Implement touch event handler
    }
    /**
     * Lifecycle hook that is called when component is initialized.
     * @return {?}
     */
    ngOnInit() {
        if (!this.dataStateService.id) {
            throw Error('Missing required parameter value for [id] input.');
        }
        if (!this.dataStateService.onDataBind) {
            this.dataSource = this.eventStateService.staticDataSourceStream;
        }
        this.initDataFetchEvent();
        if (this.config.loadDataOnInit) {
            this.eventStateService.dataFetchStream.emit(false);
        }
        if (this.config.triggerSelectChangeOnInit) {
            this.triggerSelectChange();
        }
        this.eventStateService.initStream.emit(this);
    }
    /**
     * Map source data object to dropdown option model.
     * @private
     * @param {?} option Source dropdown option.
     * @param {?} index Current index.
     * @return {?}
     */
    mapDropdownOption(option, index) {
        /** @type {?} */
        const id = get(option, this.config.selectTrackBy);
        return {
            disabled: get(option, this.config.disabledTrackBy),
            id,
            index: index + this.dataStateService.offset + 1,
            option,
            text: get(option, this.config.displayTrackBy)
        };
    }
    /**
     *
     * Set dropdown options data.
     * @private
     * @param {?} queryResult Query result object reference.
     * @return {?}
     */
    setDropdownOptions(queryResult) {
        if (this.config.groupByField) {
            this.dataStateService.dropdownOptionGroups = queryResult.options.reduce((/**
             * @param {?} accumulator
             * @param {?} option
             * @param {?} index
             * @return {?}
             */
            (accumulator, option, index) => {
                /** @type {?} */
                const groupFieldValue = get(option, this.config.groupByField);
                /** @type {?} */
                const currentGroup = accumulator.find((/**
                 * @param {?} group
                 * @return {?}
                 */
                (group) => group.groupName === groupFieldValue));
                if (currentGroup) {
                    currentGroup.options.push(this.mapDropdownOption(option, index));
                }
                else {
                    accumulator.push({
                        groupName: groupFieldValue,
                        options: [this.mapDropdownOption(option, index)]
                    });
                }
                return accumulator;
            }), this.config.loadOnScroll && this.dataStateService.offset > 0 ? this.dataStateService.dropdownOptionGroups : []);
        }
        else {
            this.dataStateService.dropdownOptions = queryResult.options.reduce((/**
             * @param {?} accumulator
             * @param {?} option
             * @param {?} index
             * @return {?}
             */
            (accumulator, option, index) => {
                accumulator.push(this.mapDropdownOption(option, index));
                return accumulator;
            }), this.config.loadOnScroll && this.dataStateService.offset > 0 ? this.dataStateService.dropdownOptions : []);
        }
        if (this.config.setFirstOptionSelected && queryResult.options.length) {
            if (this.config.selectMode === 'multi') {
                this.dataStateService.selectedOptions = [queryResult.options[0]];
            }
            else {
                this.dataStateService.selectedOption = queryResult.options[0];
            }
            if (this.config.triggerSelectChangeOnFirstOptionSelect) {
                this.triggerSelectChange();
            }
        }
        this.dataStateService.totalOptionCount = queryResult.count;
        this.dataStateService.currentOptionCount += queryResult.options.length;
    }
    /**
     * On after data bind event handler.
     * @private
     * @param {?} queryResult Query result object reference.
     * @return {?}
     */
    onAfterDataBind(queryResult) {
        this.setDropdownOptions(queryResult);
        this.dataStateService.dataLoading = false;
        this.eventStateService.dataBoundStream.emit();
    }
    /**
     * Fetch query results.
     * @private
     * @param {?} hardReload Hard reload state.
     * @return {?}
     */
    fetchQueryResults(hardReload) {
        this.dataStateService.dataLoading = true;
        if (hardReload) {
            this.dataStateService.offset = 0;
            this.dataStateService.filterText = '';
        }
        /** @type {?} */
        const requestParams = {
            hardReload
        };
        if (this.config.loadOnScroll) {
            requestParams.limit = this.config.limit;
            requestParams.offset = this.dataStateService.offset;
        }
        if (this.config.filterable) {
            requestParams.filter = {
                key: this.config.displayTrackBy,
                value: this.dataStateService.filterText
            };
        }
        return this.dataStateService.onDataBind(requestParams);
    }
    /**
     * Initialize data fetch event.
     * @private
     * @return {?}
     */
    initDataFetchEvent() {
        /** @type {?} */
        const noop = {
            options: [],
            count: 0
        };
        this.eventStateService.dataFetchStream
            .pipe(debounceTime(20), switchMap((/**
         * @param {?} hardReload
         * @return {?}
         */
        (hardReload) => this.fetchQueryResults(hardReload))), catchError((/**
         * @return {?}
         */
        () => {
            return of(noop);
        })))
            .subscribe((/**
         * @param {?} queryResult
         * @return {?}
         */
        (queryResult) => {
            this.onAfterDataBind(queryResult);
        }), (/**
         * @return {?}
         */
        () => {
            this.onAfterDataBind(noop);
        }));
    }
    /**
     * Trigger explicit data fetch.
     * @param {?=} hardReload Hard reload state.
     * @return {?}
     */
    fetchData(hardReload = false) {
        this.eventStateService.dataFetchStream.emit(hardReload);
    }
    /**
     * On select option remove event handler.
     * @param {?} index Selected option index.
     * @return {?}
     */
    onSelectOptionRemove(index) {
        this.dataStateService.selectedOptions.splice(index, 1);
        this.eventStateService.selectChangeStream.emit(this.dataStateService.selectedOptions);
    }
    /**
     * Close dropdown options menu.
     * @return {?}
     */
    close() {
        this.dataStateService.componentLoaderRef.hide();
    }
}
DropdownComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-dropdown',
                template: "<div #dropdownElement class=\"ng-dropdown\">\n  <div class=\"ng-dropdown-toggle\" (click)=\"toggleDropdown($event, dropdownElement)\" [class.disabled]=\"dataStateService.disabled\">\n    <ng-container *ngIf=\"config.selectMode === 'multi'\">\n      <ng-container *ngIf=\"!wrapSelectedOptions\">\n        <span class=\"ng-dropdown-selected-value\"\n          *ngFor=\"let option of dataStateService.selectedOptions; let index = index\"\n          [style.max-width]=\"config.multiSelectOptionMaxWidth | ngPx\"\n          [class.ng-dropdown-option-close-enabled]=\"config.showSelectedOptionRemoveButton\">\n          {{ config.getDisplayText(option) }}\n          <i class=\"ng-dropdown-remove-selected-value ng-ignore-propagation\"\n            *ngIf=\"config.showSelectedOptionRemoveButton\"\n            (click)=\"onSelectOptionRemove(index)\"></i>\n        </span>\n      </ng-container>\n      <span class=\"ng-dropdown-selected-option\" *ngIf=\"wrapSelectedOptions\">\n        {{ wrappedOptionDisplayText }}\n      </span>\n      <span class=\"ng-dropdown-selected-placeholder\" *ngIf=\"!dataStateService.selectedOptions.length\">\n        {{ config.translations.selectPlaceholder }}\n      </span>\n    </ng-container>\n    <ng-container *ngIf=\"config.selectMode !== 'multi'\">\n      <span class=\"ng-dropdown-selected-option\" *ngIf=\"dataStateService.selectedOption\">\n        {{ config.getDisplayText(dataStateService.selectedOption) }}\n      </span>\n      <span class=\"ng-dropdown-selected-placeholder\" *ngIf=\"!dataStateService.selectedOption\">\n        {{ config.translations.selectPlaceholder }}\n      </span>\n    </ng-container>\n    <span class=\"ng-dropdown-input-group-btn\" [hidden]=\"!hasSelectedOptions\" *ngIf=\"config.showClearSelectionButton\">\n      <div class=\"ng-dropdown-delete-button ng-ignore-propagation\" (click)=\"clearSelectedOptions()\"></div>\n    </span>\n    <span class=\"ng-dropdown-down-arrow\" *ngIf=\"!dataStateService.dataLoading\"></span>\n    <span class=\"ng-dropdown-loading\" [hidden]=\"!dataStateService.dataLoading\">\n      <div class=\"ng-dropdown-loading-animation\" *ngIf=\"!loadingSpinnerTemplate\"></div>\n      <div *ngIf=\"loadingSpinnerTemplate\" [ngTemplateOutlet]=\"loadingSpinnerTemplate\"></div>\n    </span>\n  </div>\n</div>\n",
                providers: [
                    DropdownConfigService,
                    DropdownDataStateService,
                    DropdownEventStateService,
                    DropdownResourceService,
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => DropdownComponent)),
                        multi: true
                    }
                ]
            }] }
];
/** @nocollapse */
DropdownComponent.ctorParameters = () => [
    { type: PopoverComponentLoaderFactoryService },
    { type: Injector },
    { type: DropdownEventStateService },
    { type: DropdownResourceService },
    { type: Renderer2 },
    { type: DropdownDataStateService },
    { type: DropdownConfigService }
];
DropdownComponent.propDecorators = {
    loadingSpinnerTemplate: [{ type: ContentChild, args: ['ngDropdownLoadingSpinner', { static: true },] }],
    optionTemplate: [{ type: ContentChild, args: ['ngDropdownOption', { static: true },] }],
    optionGroupHeaderTemplate: [{ type: ContentChild, args: ['ngDropdownOptionGroupHeader', { static: true },] }],
    init: [{ type: Output }],
    selectChange: [{ type: Output }],
    dataBound: [{ type: Output }],
    onDataBind: [{ type: Input }],
    loadingSpinnerTemplateRef: [{ type: Input }],
    optionTemplateRef: [{ type: Input }],
    optionGroupHeaderTemplateRef: [{ type: Input }],
    options: [{ type: Input }],
    dataSource: [{ type: Input }],
    id: [{ type: Input }],
    translations: [{ type: Input }],
    selectTrackBy: [{ type: Input }],
    displayTrackBy: [{ type: Input }],
    groupByField: [{ type: Input }],
    disabledTrackBy: [{ type: Input }],
    selectedOptions: [{ type: Input }],
    selectedOption: [{ type: Input }],
    limit: [{ type: Input }],
    wrapDisplaySelectLimit: [{ type: Input }],
    loadOnScroll: [{ type: Input }],
    loadViewDistanceRatio: [{ type: Input }],
    selectMode: [{ type: Input }],
    filterable: [{ type: Input }],
    filterText: [{ type: Input }],
    filterDebounce: [{ type: Input }],
    filterDebounceTime: [{ type: Input }],
    loadDataOnInit: [{ type: Input }],
    showSelectedOptionRemoveButton: [{ type: Input }],
    showClearSelectionButton: [{ type: Input }],
    menuWidth: [{ type: Input }],
    menuHeight: [{ type: Input }],
    menuPosition: [{ type: Input }],
    disabled: [{ type: Input }],
    closeMenuOnSelect: [{ type: Input }],
    showOptionSelectCheckbox: [{ type: Input }],
    showOptionIndex: [{ type: Input }],
    showOptionTrackBy: [{ type: Input }],
    multiSelectOptionMaxWidth: [{ type: Input }],
    setFirstOptionSelected: [{ type: Input }],
    triggerSelectChangeOnInit: [{ type: Input }],
    triggerSelectChangeOnModelUpdate: [{ type: Input }],
    triggerSelectChangeOnFirstOptionSelect: [{ type: Input }],
    dynamicDimensionCalculation: [{ type: Input }],
    dynamicWidthRatio: [{ type: Input }],
    dynamicHeightRatio: [{ type: Input }],
    relativeParentElement: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.onSelectChangeSubscription;
    /** @type {?} */
    DropdownComponent.prototype.loadingSpinnerTemplate;
    /**
     * Dropdown initialize event handler
     * @type {?}
     */
    DropdownComponent.prototype.init;
    /**
     * Dropdown option select change event handler
     * @type {?}
     */
    DropdownComponent.prototype.selectChange;
    /**
     * Dropdown data bind event handler
     * @type {?}
     */
    DropdownComponent.prototype.dataBound;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.componentLoaderFactory;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.eventStateService;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.dropdownResourceService;
    /**
     * @type {?}
     * @private
     */
    DropdownComponent.prototype.renderer;
    /** @type {?} */
    DropdownComponent.prototype.dataStateService;
    /** @type {?} */
    DropdownComponent.prototype.config;
}
